{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# This notebook was created to find all of the High Quality SNV calls for all of the isolates in rollingDB, changed from previous version by NOT using Lineage Calls from MegaPipe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }<|/style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }<|/style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "\n",
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from itertools import compress\n",
    "from pylab import MaxNLocator\n",
    "import seaborn as sns; sns.set()\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "import ast\n",
    "import itertools\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "\n",
    "import fastcluster\n",
    "from sklearn import cluster, datasets\n",
    "import scipy.cluster.hierarchy as hier\n",
    "from sklearn.cluster import KMeans\n",
    "import time\n",
    "import sys\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "import itertools\n",
    "import gzip\n",
    "\n",
    "import networkx as nx\n",
    "import scipy\n",
    "import pickle\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [1] This section of the notebook loads all of the Lineage Calls that were made on the isolates in rollingDB as well as the mean coverage / isolates for all of the isolates in rollingDB. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We run through the pilon-generated VCF for each isolate (N = 33,875) and scrapes the High Quality Single Nucleotide Variant calls made against H37Rv. The DataFrame containing these high quality SNV calls (with the Reference Position, Reference Allele and Alterante Allele) for each isolate is then stored in a directory containing all of the DFs for all isolates. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Side-Note: Some isolate have ~20x coverage and we require 20 reads at a given position to make a SNP call"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Scrape Luca's DB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0500090232801\n",
      "0.10001804656\n",
      "0.15002706984\n",
      "0.20003609312\n",
      "0.2500451164\n",
      "0.30005413968\n",
      "0.35006316296\n",
      "0.400072186241\n",
      "0.450081209521\n",
      "0.500090232801\n",
      "0.550099256081\n",
      "0.600108279361\n",
      "0.650117302641\n"
     ]
    }
   ],
   "source": [
    "#genomic data directory\n",
    "rolling_DB_dir = '/n/data1/hms/dbmi/farhat/rollingDB/genomic_data/'\n",
    "\n",
    "#get all folders (each folder corresponds to a different sequenced isolate)\n",
    "isolate_directories = os.listdir(rolling_DB_dir)\n",
    "\n",
    "#iterate through each sequenced isolate\n",
    "isolate_i = 0\n",
    "\n",
    "for isolate_ID in isolate_directories:\n",
    "\n",
    "    #directory that stores files for each sequenced isolate\n",
    "    directory_for_sequenced_isolate = rolling_DB_dir + isolate_ID\n",
    "\n",
    "    #check to see if megapipe successfully ran on sequenced isolate\n",
    "    try:\n",
    "        #existence of a PILON directory and corresponding VCF file\n",
    "        if ( 'pilon' in os.listdir(directory_for_sequenced_isolate) ):\n",
    "            \n",
    "            #existence of a REDUCED VCF and existence of FULL VCF file in relevent directories \n",
    "            if ( 'vcf' in list( itertools.chain.from_iterable( [filename.split('.') for filename in os.listdir(directory_for_sequenced_isolate + '/pilon/')] ) ) ) and ( isolate_ID + '_full' in list( itertools.chain.from_iterable( [filename.split('.') for filename in os.listdir(directory_for_sequenced_isolate + '/pilon/')] ) ) ):\n",
    "                \n",
    "                #we have a valid VCF file so megapipe ran successfully, let's keep the variant call information for this sequenced isolate\n",
    "                \n",
    "                #PILON-generated VCF DATA\n",
    "                ########################################################################################################################\n",
    "                \n",
    "                #dictionaries to construct DataFrame from later that hold information about each high-quality SNP found for a specific isolate from rollingDB\n",
    "                ref_dict = {} #reference allele\n",
    "                alt_dict = {} #alternate allele\n",
    "                alt_AF_dict = {} #frequency of alternate allele in population (0.75 - 1.0)\n",
    "                \n",
    "                #load in VCF file if one was found\n",
    "                vcf_reader = vcf.Reader( open(directory_for_sequenced_isolate + '/pilon/' + isolate_ID + '.vcf' , 'r') )\n",
    "\n",
    "                #iterate through each Variant Call \n",
    "                for record in vcf_reader:\n",
    "                    \n",
    "                    #check to see if the call is a PASS by Pilon\n",
    "                    if record.FILTER == []:\n",
    "\n",
    "                        #check to see if variant is SNP (length of Reference Allele is 1 base ; there's only 1 alternate allele ; length of alternate allele is 1 base)\n",
    "                        if (len(record.REF) == 1) and (len(record.ALT) == 1) and (len(str(record.ALT[0])) == 1):\n",
    "\n",
    "                            ##### Retrieve Relevant information for filtering quality of Base Call #####\n",
    "                            # Mean Base Quality @ locus\n",
    "                            BQ = record.INFO['BQ']\n",
    "                            # Mean Mapping Quality @ locus\n",
    "                            MQ = record.INFO['MQ']\n",
    "                            # Number of Reads w/ Deletion \n",
    "                            DC = record.INFO['DC']\n",
    "                            # Number of Reads w/ Insertion\n",
    "                            IC = record.INFO['IC']\n",
    "                            # Depth of Valid Reads in Pileup\n",
    "                            VD = record.INFO['DP']\n",
    "\n",
    "                            ### Filtering Criteria\n",
    "                            #---> Mean Base Quality > 20\n",
    "                            #---> Mean Mapping Quality > 30\n",
    "                            #---> No Reads Supporting Insertions\n",
    "                            #---> No Reads Supporting Deletions\n",
    "                            #---> Number of High Quality Reads >= 20\n",
    "                            if (BQ > 20) and (MQ > 30) and (DC == 0) and (IC == 0) and (VD >= 20): #SNP passed filtering criteria!\n",
    "\n",
    "                                # Filtering Criteria for mutant allele frequency calculation (range: 0.75 - 1.0)\n",
    "                                ref_allele = str( record.REF )\n",
    "                                alt_allele = str( record.ALT[0] )\n",
    "                                ref_pos = int( record.POS )\n",
    "                                \n",
    "                                # Calculate Alternate Allele Frequency\n",
    "                                # only stores SNV if 1 or 2 bases are supported by reads\n",
    "                                ########################################################################\n",
    "                                #position-base dictionaries (order of output in Pilon)\n",
    "                                base_order = {0:'A' , 1:'C' , 2:'G', 3:'T'}\n",
    "                                base_order_r = {'A':0 , 'C':1 , 'G':2 , 'T':3}\n",
    "\n",
    "                                # Quality-Percentage - Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\n",
    "                                QP = record.INFO['QP']\n",
    "\n",
    "                                if (sum(np.array(QP) == 0) == 3): #1 Allele supported by Reads (Alternate), store information for this Base Call (3 bases have NO reads supporting)\n",
    "\n",
    "                                    #create a Boolean list for the Bases that have at least 1% of support from reads\n",
    "                                    supported_bases_boolean = list( np.array( QP ) > 0 )\n",
    "\n",
    "                                    #positions of Bases [A,C,G,T] with at least 1% of support from reads\n",
    "                                    supported_bases_i = [i for i,x in enumerate(supported_bases_boolean) if x == True] #returns list of length \n",
    "                                    supported_base = base_order[supported_bases_i[0]]\n",
    "\n",
    "                                    #if the Reference Allele is supported\n",
    "                                    if supported_base == ref_allele:\n",
    "                                        alt_allele_frequency = 0.0 #no alternate allele\n",
    "\n",
    "                                    #if an Alternate Allele is supported:\n",
    "                                    if supported_base != ref_allele:\n",
    "                                        alt_allele_frequency = 1.0 #alternate allele\n",
    "\n",
    "                                elif (sum(np.array(QP) == 0) == 2): #2 Alleles supported, make sure 1 is the Reference & confirm the Alternate (2 bases have NO reads supporting)\n",
    "\n",
    "                                    #position of the Reference Allele in [A,C,G,T]\n",
    "                                    ref_allele_base_i = base_order_r[ref_allele] \n",
    "\n",
    "                                    #create a Boolean list for the Bases that have at least 1% of support from reads\n",
    "                                    supported_bases_boolean = list( np.array( QP ) > 0 )\n",
    "\n",
    "                                    #positions of Bases [A,C,G,T] with at least 1% of support from reads\n",
    "                                    supported_bases_i = [i for i,x in enumerate(supported_bases_boolean) if x == True] #returns list of length 2\n",
    "\n",
    "                                    #check that Reference Allele is supported by at least 1% of the reads\n",
    "                                    if ref_allele_base_i in supported_bases_i:\n",
    "\n",
    "                                        #find the Alternate Allele that was supported by at least 1% of the reads\n",
    "                                        for supported_base_i in supported_bases_i:\n",
    "\n",
    "                                            if supported_base_i != ref_allele_base_i: #then must be alternate allele\n",
    "\n",
    "                                                alt_allele = base_order[supported_base_i] #Alternate Allele\n",
    "                                                alt_allele_frequency = float(QP[supported_base_i]) / 100.0 #weighted percentage of reads supporting Alternate Allele\n",
    "                                                break\n",
    "                                ########################################################################\n",
    "                                \n",
    "                                # use only consensus SNPs (where alternate allele frequency >= 75%)\n",
    "                                if alt_allele_frequency >= 0.75:\n",
    "                                \n",
    "                                    # Store good-quality SNV in SNV dictionaries\n",
    "                                    ref_dict[ref_pos] = ref_allele\n",
    "                                    alt_dict[ref_pos] = alt_allele\n",
    "\n",
    "                #construct DataFrame from dictionaries\n",
    "                SNV_DF = pd.DataFrame( index = pd.Series( ref_dict.keys() ) )\n",
    "                SNV_DF['ref'] = pd.Series(ref_dict)\n",
    "                SNV_DF['alt'] = pd.Series(alt_dict)\n",
    "\n",
    "                #make reference-position a column and re-index, sort according to reference position\n",
    "                SNV_DF['pos'] = SNV_DF.index\n",
    "                SNV_DF.sort_values(by = 'pos' , ascending = True , inplace = True)\n",
    "                SNV_DF.reset_index(inplace = True , drop = True)\n",
    "\n",
    "                #pickle DataFrame holding High Quality SNVs from isolate\n",
    "                SNV_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB/' + isolate_ID + '_SNVs.pkl')\n",
    "                ########################################################################################################################\n",
    "                \n",
    "                #keep track of progress   \n",
    "                isolate_i += 1\n",
    "                if isolate_i % np.ceil(0.05*len(isolate_directories)) == 0:\n",
    "                    print float(isolate_i) / float(len(isolate_directories))\n",
    "        \n",
    "    except OSError: #hit some file that is not another directory with genomic data \n",
    "        continue"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [2] This section of the notebook constructs a script that iterates over *subsets* of the HQ SNV DataFrames for each isolate and creates a 'catalogue' of all of the HQ SNVs found in rollingDB. All of the relevent information is stored in dictionaries before being transformed into a DataFrame with all of the unique SNVs (indexed by Reference Position) and the number of isolates that contain a SNV at that position."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "import vcf\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "from pylab import MaxNLocator\n",
    "import ast\n",
    "import itertools\n",
    "import time\n",
    "import sys\n",
    "import itertools\n",
    "import gzip\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### We are going to split this up into several jobs, each iterating through N isolates. Each job will produce a DF with all SNVs detected in any of the N isolates & corresponding alleles.  Once all jobs have finished, we will merge all DFs together to create one master DF of all SNVs detected in any of the 33,875 isolates in rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "#INPUTS\n",
    "first_isolate_index = int(sys.argv[1]) #INPUT - first isolate index\n",
    "last_isolate_index = int(sys.argv[2]) #INPUT - last isolate index\n",
    "job_number = str(sys.argv[3]) #INPUT - job number"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "#list of pickles DataFrame files for SNVs for each isolate in rollingDB\n",
    "SNV_DF_list = os.listdir('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB')\n",
    "\n",
    "#subset to only the pickled DFs for the isolates that will be searched for SNVs in this job\n",
    "SNV_DF_list = SNV_DF_list[first_isolate_index:last_isolate_index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "def store_SNV_info_from_SNV(SNV_DF_row):\n",
    "\n",
    "    ref_allele , alt_allele , ref_pos = list(SNV_DF_row)\n",
    "\n",
    "    #SNV is not documented at this Reference Position\n",
    "    if ref_pos not in ref_dict.keys():\n",
    "\n",
    "        #store a new row for SNV at this reference position; ref allele, alt allele 1, nan, nan, 1 isolate has polymorphism here (so far)\n",
    "        ref_dict[ref_pos] = ref_allele\n",
    "        alt1_dict[ref_pos] = alt_allele\n",
    "        alt2_dict[ref_pos] = np.nan\n",
    "        alt3_dict[ref_pos] = np.nan\n",
    "        num_isolates_dict[ref_pos] = 1\n",
    "\n",
    "    #a SNV is already documented at this Reference Position\n",
    "    else:\n",
    "\n",
    "        #check to see if specific allele is documented, if not add it\n",
    "        if alt_allele not in [alt1_dict[ref_pos] , alt2_dict[ref_pos] , alt3_dict[ref_pos]]:\n",
    "\n",
    "            #need to document presence of new allele in isolate pool (either going to be alt2 or alt3)\n",
    "            if ( type(alt2_dict[ref_pos])==float ) and ( np.isnan(alt2_dict[ref_pos]) ): #alt2 has not been found yet\n",
    "\n",
    "                alt2_dict[ref_pos] = alt_allele\n",
    "\n",
    "            elif ( type(alt3_dict[ref_pos])==float ) and ( np.isnan(alt3_dict[ref_pos]) ): #alt3 has not been found yet\n",
    "\n",
    "                alt3_dict[ref_pos] = alt_allele\n",
    "\n",
    "        #increment # of isolates that have a polymorphism here\n",
    "        num_isolates_dict[ref_pos] = num_isolates_dict[ref_pos] + 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": [
    "#create dictionaries to store SNV info for each isolate\n",
    "ref_dict = {}\n",
    "alt1_dict = {}\n",
    "alt2_dict = {}\n",
    "alt3_dict = {}\n",
    "num_isolates_dict = {}\n",
    "\n",
    "#iterate through all DataFrames for each isolate in rollingDB\n",
    "for pickled_SNV_DF in SNV_DF_list:\n",
    "    \n",
    "    #load SNV DF for isolates from pickled object\n",
    "    SNV_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB/' + pickled_SNV_DF)\n",
    "\n",
    "    #extract important info from SNV df\n",
    "    _ = SNV_DF.apply(lambda row: store_SNV_info_from_SNV(row) , axis = 1 )  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#construct DataFrame from dictionaries\n",
    "SNV_DF = pd.DataFrame( index = pd.Series( ref_dict.keys() ) )\n",
    "SNV_DF['ref'] = pd.Series(ref_dict)\n",
    "SNV_DF['alt1'] = pd.Series(alt1_dict)\n",
    "SNV_DF['alt2'] = pd.Series(alt2_dict)\n",
    "SNV_DF['alt3'] = pd.Series(alt3_dict)\n",
    "SNV_DF['num_isolates'] = pd.Series(num_isolates_dict)\n",
    "\n",
    "#make reference-position a column and re-index\n",
    "SNV_DF['pos'] = SNV_DF.index\n",
    "\n",
    "#sort according to reference position\n",
    "SNV_DF.sort_values(by = 'pos' , ascending = True , inplace = True)\n",
    "\n",
    "SNV_DF.reset_index(inplace = True , drop = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Output Pickle files for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "SNV_DF.to_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/seperate_HQ_SNV_DFs/HighQuality_SNVs_found_in_rollingDB_catalogue_' + job_number + '.pkl' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [3] This section of the notebook launches the script that was created in the previous section of the notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from slurmpy import Slurm\n",
    "import numpy as np\n",
    "import itertools\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#list of all SNV DFs for each isolate that was run through MegaPipe in rollingDB\n",
    "SNV_DF_list = os.listdir('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB')\n",
    "\n",
    "isolates_per_job = 500 #number of isolates each job will catalogue\n",
    "njobs = int( np.ceil( float( len( SNV_DF_list ) ) / float(isolates_per_job) ) ) #number of jobs required if we split for every 500 isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "68\n"
     ]
    }
   ],
   "source": [
    "print njobs #number of jobs we will submit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825966\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_1 : 6825966\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825967\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_2 : 6825967\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825968\n",
      "submitted: Submitted batch job 6825969\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_3 : 6825968\n",
      "SNV_4 : 6825969\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825970\n",
      "submitted: Submitted batch job 6825971\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_5 : 6825970\n",
      "SNV_6 : 6825971\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825972\n",
      "submitted: Submitted batch job 6825973\n",
      "submitted: Submitted batch job 6825974\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_7 : 6825972\n",
      "SNV_8 : 6825973\n",
      "SNV_9 : 6825974\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825975\n",
      "submitted: Submitted batch job 6825976\n",
      "submitted: Submitted batch job 6825977\n",
      "submitted: Submitted batch job 6825978\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_10 : 6825975\n",
      "SNV_11 : 6825976\n",
      "SNV_12 : 6825977\n",
      "SNV_13 : 6825978\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825979\n",
      "submitted: Submitted batch job 6825980\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_14 : 6825979\n",
      "SNV_15 : 6825980\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825981\n",
      "submitted: Submitted batch job 6825982\n",
      "submitted: Submitted batch job 6825983\n",
      "submitted: Submitted batch job 6825984\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_16 : 6825981\n",
      "SNV_17 : 6825982\n",
      "SNV_18 : 6825983\n",
      "SNV_19 : 6825984\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825985\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_20 : 6825985\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825986\n",
      "submitted: Submitted batch job 6825987\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_21 : 6825986\n",
      "SNV_22 : 6825987\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825988\n",
      "submitted: Submitted batch job 6825989\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_23 : 6825988\n",
      "SNV_24 : 6825989\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825990\n",
      "submitted: Submitted batch job 6825991\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_25 : 6825990\n",
      "SNV_26 : 6825991\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825992\n",
      "submitted: Submitted batch job 6825993\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_27 : 6825992\n",
      "SNV_28 : 6825993\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825994\n",
      "submitted: Submitted batch job 6825995\n",
      "submitted: Submitted batch job 6825996\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_29 : 6825994\n",
      "SNV_30 : 6825995\n",
      "SNV_31 : 6825996\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6825997\n",
      "submitted: Submitted batch job 6825998\n",
      "submitted: Submitted batch job 6825999\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_32 : 6825997\n",
      "SNV_33 : 6825998\n",
      "SNV_34 : 6825999\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826000\n",
      "submitted: Submitted batch job 6826001\n",
      "submitted: Submitted batch job 6826002\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_35 : 6826000\n",
      "SNV_36 : 6826001\n",
      "SNV_37 : 6826002\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826003\n",
      "submitted: Submitted batch job 6826004\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_38 : 6826003\n",
      "SNV_39 : 6826004\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826005\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_40 : 6826005\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826006\n",
      "submitted: Submitted batch job 6826007\n",
      "submitted: Submitted batch job 6826008\n",
      "submitted: Submitted batch job 6826009\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_41 : 6826006\n",
      "SNV_42 : 6826007\n",
      "SNV_43 : 6826008\n",
      "SNV_44 : 6826009\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826010\n",
      "submitted: Submitted batch job 6826011\n",
      "submitted: Submitted batch job 6826012\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_45 : 6826010\n",
      "SNV_46 : 6826011\n",
      "SNV_47 : 6826012\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826013\n",
      "submitted: Submitted batch job 6826014\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_48 : 6826013\n",
      "SNV_49 : 6826014\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826015\n",
      "submitted: Submitted batch job 6826016\n",
      "submitted: Submitted batch job 6826017\n",
      "submitted: Submitted batch job 6826018\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_50 : 6826015\n",
      "SNV_51 : 6826016\n",
      "SNV_52 : 6826017\n",
      "SNV_53 : 6826018\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826019\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_54 : 6826019\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826020\n",
      "submitted: Submitted batch job 6826021\n",
      "submitted: Submitted batch job 6826022\n",
      "submitted: Submitted batch job 6826023\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_55 : 6826020\n",
      "SNV_56 : 6826021\n",
      "SNV_57 : 6826022\n",
      "SNV_58 : 6826023\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826024\n",
      "submitted: Submitted batch job 6826025\n",
      "submitted: Submitted batch job 6826026\n",
      "submitted: Submitted batch job 6826027\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_59 : 6826024\n",
      "SNV_60 : 6826025\n",
      "SNV_61 : 6826026\n",
      "SNV_62 : 6826027\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826028\n",
      "submitted: Submitted batch job 6826029\n",
      "submitted: Submitted batch job 6826030\n",
      "submitted: Submitted batch job 6826031\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNV_63 : 6826028\n",
      "SNV_64 : 6826029\n",
      "SNV_65 : 6826030\n",
      "SNV_66 : 6826031\n",
      "SNV_67 : 6826032\n",
      "SNV_68 : 6826033\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6826032\n",
      "submitted: Submitted batch job 6826033\n"
     ]
    }
   ],
   "source": [
    "first_isolate_index = 0\n",
    "max_isolate_index = int( len( SNV_DF_list ) )\n",
    "\n",
    "#run each job using a difference set of indices for isolates that we will catalogue SNVs from\n",
    "for job_number in np.arange(1 , njobs+1):\n",
    "\n",
    "    last_isolate_index = first_isolate_index + isolates_per_job\n",
    "    if last_isolate_index > max_isolate_index: #watch out for the last job\n",
    "        last_isolate_index = max_isolate_index\n",
    "    \n",
    "    rollingDB_SNV_cataloging_job = 'python /home/rv76/Farhat_Lab/Python_Scripts/homoplasy_project/rollingDB_HQ_SNV_catalogue_scrape.py ' + str(first_isolate_index) + ' ' + str(last_isolate_index) + ' ' + str(job_number)\n",
    "    \n",
    "    #directory where you want output + error files\n",
    "    os.chdir('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/SNV_catalogue_scrape_jobs/')\n",
    "\n",
    "    job_name = 'SNV_' + str(job_number)\n",
    "\n",
    "    s = Slurm(job_name , {'partition':'short' , 'N':'1' , 't':'0-01:00:00' , 'mem':'4G' , 'mail-type':'FAIL' , 'mail-user':'roger_vargas@g.harvard.edu'})\n",
    "\n",
    "    #submits the job\n",
    "    job_id = s.run(rollingDB_SNV_cataloging_job)\n",
    "\n",
    "    print job_name  + ' : ' +  str(job_id)\n",
    "    \n",
    "    #setup isolate indices for next job\n",
    "    first_isolate_index = last_isolate_index"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [4] The purpose of this section of the notebook is generate the master catalogue of (high-quality) SNV calls made in rollingDB by merging the catalogues of SNVs found in seperate batches of isolates."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import ast\n",
    "import itertools\n",
    "import time\n",
    "import sys\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#INPUTS\n",
    "njobs = 68 #number of seperate SNV catalogues for subsets of isolates in rollingDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load in SNV catalogue(s) (High Quality SNV calls across all isolates in RollingDB)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We're going to iterate through each DataFrame (catalogue of SNVs detected within each batch of isolates) and create a master DF of all SNVs detected within all the isolates in rollingDB."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "def convert_SNV_DF_rows_to_DICT_elments(row):\n",
    "    \n",
    "    '''\n",
    "    This function converts the row of a SNV catalogue DF to a\n",
    "    list with two elements: (1) The Reference Position of an SNV detected within a subset of isolates\n",
    "    and (2) another list containing the Reference Allele, all Alternate Alleles, and Number of Isolates\n",
    "    with an SNV at this Reference Position\n",
    "    '''\n",
    "    \n",
    "    #create set of alternate alleles present for this SNV\n",
    "    alt_alleles_SNV_i = set( [row['alt1'] , row['alt2'] , row['alt3']] )\n",
    "    #purge nan values\n",
    "    alt_alleles_SNV_i = {x for x in alt_alleles_SNV_i if x==x}\n",
    "        \n",
    "    return [ row['pos'] , [row['ref'] , alt_alleles_SNV_i , row['num_isolates']] ]\n",
    "\n",
    "def convert_SNV_DF_to_DICT(SNV_DF):\n",
    "    \n",
    "    '''\n",
    "    This function takes an SNV DF catalogue for a subset of isolates and converts it into \n",
    "    an SNV DICT catalogue using the function above by applying it to each row.\n",
    "    '''\n",
    "    \n",
    "    SNV_dict = SNV_DF.apply( convert_SNV_DF_rows_to_DICT_elments , axis = 1 )\n",
    "    return dict(list( SNV_dict.values ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert SNV catalogue(s) DFs to SNV catalogue DICTs and store SNP catalogue info across jobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n",
      "4\n",
      "5\n",
      "6\n",
      "7\n",
      "8\n",
      "9\n",
      "10\n",
      "11\n",
      "12\n",
      "13\n",
      "14\n",
      "15\n",
      "16\n",
      "17\n",
      "18\n",
      "19\n",
      "20\n",
      "21\n",
      "22\n",
      "23\n",
      "24\n",
      "25\n",
      "26\n",
      "27\n",
      "28\n",
      "29\n",
      "30\n",
      "31\n",
      "32\n",
      "33\n",
      "34\n",
      "35\n",
      "36\n",
      "37\n",
      "38\n",
      "39\n",
      "40\n",
      "41\n",
      "42\n",
      "43\n",
      "44\n",
      "45\n",
      "46\n",
      "47\n",
      "48\n",
      "49\n",
      "50\n",
      "51\n",
      "52\n",
      "53\n",
      "54\n",
      "55\n",
      "56\n",
      "57\n",
      "58\n",
      "59\n",
      "60\n",
      "61\n",
      "62\n",
      "63\n",
      "64\n",
      "65\n",
      "66\n",
      "67\n",
      "68\n"
     ]
    }
   ],
   "source": [
    "#master set of all reference positions detected within rollingDB (faster to query a set)\n",
    "SNV_set_Ref_Position = set([])\n",
    "\n",
    "#master dictionaries of all SNVs detected within rollingDB\n",
    "SNV_dict_Ref_Allele = {} #keys - reference position; values - ref_allele (string)\n",
    "SNV_dict_Alt_Allele = {} #keys - reference position; values - alt_alleles (set)\n",
    "SNV_dict_Num_Isolates = {} #keys - reference position; values - num_isolates (int)\n",
    "\n",
    "for job_num in np.arange(1 , njobs+1):\n",
    "    \n",
    "    #load the SNV DataFrame holding SNVs from the batch of isolates\n",
    "    SNV_DF_subset = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/seperate_HQ_SNV_DFs/HighQuality_SNVs_found_in_rollingDB_catalogue_' + str(job_num) + '.pkl' )\n",
    "    \n",
    "    #convert SNV DataFrame to an SNV Dict\n",
    "    SNV_dict_subset = convert_SNV_DF_to_DICT(SNV_DF_subset)\n",
    "    \n",
    "    #iterate through each SNV in Dict\n",
    "    for ref_pos in SNV_dict_subset.keys():\n",
    "        \n",
    "        #update master SNV catalogue dict with SNV info\n",
    "        ref , alt_alleles_SNV_i , num_isolates = SNV_dict_subset[ref_pos]\n",
    "    \n",
    "        #[A] reference position not stored in master SNV dictionaries\n",
    "        ##if ref_pos not in SNV_dict_Ref_Allele.keys():\n",
    "        if ref_pos not in SNV_set_Ref_Position:\n",
    "            \n",
    "            SNV_set_Ref_Position.add(ref_pos)\n",
    "            \n",
    "            SNV_dict_Ref_Allele[ref_pos] = ref\n",
    "            SNV_dict_Alt_Allele[ref_pos] = alt_alleles_SNV_i\n",
    "            SNV_dict_Num_Isolates[ref_pos] = num_isolates\n",
    "        \n",
    "        #[B] reference position already stored in master SNV dictionaries\n",
    "        else:\n",
    "            \n",
    "            #update SNV info with new SNV info (ref_allele stays the same)\n",
    "            SNV_dict_Alt_Allele[ref_pos] = SNV_dict_Alt_Allele[ref_pos].union(alt_alleles_SNV_i) #store any new alleles detected\n",
    "            SNV_dict_Num_Isolates[ref_pos] = SNV_dict_Num_Isolates[ref_pos] + num_isolates #add number of isolates\n",
    "            \n",
    "    #keep track of progress\n",
    "    print job_num"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert master SNV dict to SNV df "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#convert dictionaries into pandas series objects\n",
    "SNV_series_Ref_Allele = pd.Series(SNV_dict_Ref_Allele , name = 'ref')\n",
    "SNV_series_Alt_Allele = pd.Series(SNV_dict_Alt_Allele , name = 'alt')\n",
    "SNV_series_Num_Isolates = pd.Series(SNV_dict_Num_Isolates , name = 'num_isolates')\n",
    "\n",
    "#construct DataFrame from Series objects\n",
    "SNV_DF = pd.DataFrame([SNV_series_Ref_Allele , SNV_series_Alt_Allele , SNV_series_Num_Isolates]).T\n",
    "\n",
    "#convert Alternate Allele(s) column to list\n",
    "SNV_DF['alt'] = SNV_DF.alt.apply(list)\n",
    "\n",
    "#create a column for reference position\n",
    "SNV_DF['pos'] = list(SNV_DF.index)\n",
    "\n",
    "#re-index DataFrame\n",
    "SNV_DF.reset_index(inplace = True , drop = True)\n",
    "\n",
    "#store master SNV DataFrame as a pickled object\n",
    "SNV_DF.to_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/all_HighQuality_SNVs_found_in_rollingDB_catalogue.pkl' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Use to load pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/all_HighQuality_SNVs_found_in_rollingDB_catalogue.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref  alt num_isolates  pos\n",
       "0   A  [C]            3    8\n",
       "1   G  [C]            1   10\n",
       "2   A  [C]           11   11\n",
       "3   C  [T]            1   15\n",
       "4   G  [C]            1   16"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>899030</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>4411513</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899031</th>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>1</td>\n",
       "      <td>4411514</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899032</th>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>1</td>\n",
       "      <td>4411516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899033</th>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>4</td>\n",
       "      <td>4411518</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>899034</th>\n",
       "      <td>G</td>\n",
       "      <td>[A]</td>\n",
       "      <td>3</td>\n",
       "      <td>4411522</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       ref  alt num_isolates      pos\n",
       "899030   A  [C]            1  4411513\n",
       "899031   G  [A]            1  4411514\n",
       "899032   A  [G]            1  4411516\n",
       "899033   C  [T]            4  4411518\n",
       "899034   G  [A]            3  4411522"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035, 4)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [5] The purpose of this section of the notebook is to analyze the catalogue of (high-quality) SNV calls made in rollingDB. For population genetic analyses, we will retain only SNVs that are present in at least 1 isolate. We then construct a skeleton genotype DF (rows = Reference Position , columns = isolates) for all of the remaining SNVs."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Construct a series for the isolate annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "HQ_SNV_per_isolate_files = os.listdir('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB/')\n",
    "isolate_ids_list = [isolate_id.split('_SNV')[0] for isolate_id in HQ_SNV_per_isolate_files]\n",
    "isolate_id_series = pd.Series(isolate_ids_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_id_series.head() #index = column indices of genotype matrix, values isolate IDs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33875,)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_id_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Remove 2 bad isolates from isolate annotation"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove:** Delete ID for isolate **IDR1400002533** since Full GZipped VCF looks corrupt and remove from list accordingly!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = isolate_id_series[isolate_id_series != 'IDR1400002533']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Remove:** Delete ID for **Peru4994** since F2 calculations keeps failing for this isolate (lineage-defining SNP depths df is empty) **Think this is because sample has very low coverage**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = isolate_id_series[isolate_id_series != 'Peru4994']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series.reset_index(inplace = True, drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_id_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_id_series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series.to_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load in catologue of SNVs detected within rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#SNV data for all isoaltes in rollingDB \n",
    "SNV_DF = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/all_HighQuality_SNVs_found_in_rollingDB_catalogues/all_HighQuality_SNVs_found_in_rollingDB_catalogue.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref  alt num_isolates  pos\n",
       "0   A  [C]            3    8\n",
       "1   G  [C]            1   10\n",
       "2   A  [C]           11   11\n",
       "3   C  [T]            1   15\n",
       "4   G  [C]            1   16"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035, 4)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNPs are not biallelic?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "64950"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#SNPs that are not biallelic will have more than 1 alternate allele\n",
    "np.sum( SNV_DF.alt.apply(len) > 1 )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "      <td>12</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>3</td>\n",
       "      <td>67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>C</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>17</td>\n",
       "      <td>108</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>C</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>2</td>\n",
       "      <td>116</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td>T</td>\n",
       "      <td>[C, G]</td>\n",
       "      <td>4</td>\n",
       "      <td>147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>2</td>\n",
       "      <td>186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>48</th>\n",
       "      <td>G</td>\n",
       "      <td>[C, T]</td>\n",
       "      <td>3</td>\n",
       "      <td>195</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>51</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>4</td>\n",
       "      <td>203</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, C]</td>\n",
       "      <td>2</td>\n",
       "      <td>234</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>2</td>\n",
       "      <td>236</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   ref     alt num_isolates  pos\n",
       "8    A  [C, G]           12   31\n",
       "13   G  [A, T]            3   67\n",
       "28   C  [A, T]           17  108\n",
       "30   C  [A, T]            2  116\n",
       "36   T  [C, G]            4  147\n",
       "45   G  [A, C]            2  186\n",
       "48   G  [C, T]            3  195\n",
       "51   G  [A, C]            4  203\n",
       "55   G  [A, C]            2  234\n",
       "56   G  [A, T]            2  236"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF[SNV_DF.alt.apply(len) > 1].head(n = 10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=3\n",
    "plt.rcParams['axes.facecolor']='1.0'\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['axes.grid']=True\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '1.0'\n",
    "plt.rcParams.update({'font.size': 12})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhMAAAFxCAYAAADJdyroAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XtcFPX+P/DXgqCwgYIUKZLCcscLomJmpGai4C0CxJNpGmllduSidcjuJqah5sk6pSb+7KSRAnJMRMwLpqJCEgchCEUEFTURj7CwKsv8/vDBfl25rQ57AV7Px4PHw/0MM5/37CC89jOfmZEIgiCAiIiI6CEZ6bsAIiIiat8YJoiIiEgUhgkiIiIShWGCiIiIRGGYICIiIlEYJoiIiEgUhglq94YPHw47OzscO3ZM36VAqVQiPDwc7u7usLOzwyeffKL1PsPDw2FnZ4dVq1ZpvS9DYUjHXBPHjh2DnZ0dhg8fru9S2lTDcUhNTW1y+apVq2BnZ4fw8HAdV0a6xjBBbaLhl4q3tzdqa2sBAKdPn4adnR3s7Oz0XJ3u7N69G9u3b4exsTFeeeUVPPnkk42+p+EPi7u7ux4q1M0ftoY+7v8aN26c1vpszenTp/G3v/0Nnp6ekMlkGD58OGbPno3//e9/jWr+4IMPVOt98MEHan8Qy8rK1PbJ3t4eQ4YMQXh4OK5evdpk37169UJYWBimT5/+ULUHBwc3+X4290fcUHh7eyMsLAyjRo3SdymkZV30XQB1LFeuXMGWLVvw2muv6bsUUe7cuQMTE5MHXq+4uBgA8Oyzz2Lp0qVtXVa7Y2pqipkzZ6pe9+rVSy911NbWYubMmbh69Sr8/Pzw2GOPoaysDL/++ivkcjm6d++u9v0//PAD5s+fj8cff7zF7Tbs2549e7B9+3acO3cOycnJjb7PwcGhTUapnnzySXh6eqpe9+3bV/Q2tWnMmDEYM2aMvssgHeDIBLUpiUSCr7/+WjU6cb+GT1RlZWUAGg+DxsfHw87ODs899xw++ugjODs7Y/To0Th9+jRWrlwJNzc3jBgxAunp6Y22nZubi3HjxsHFxQVhYWG4fv26atnJkycRHBwMDw8PeHt7IzIyUrX83k+aW7Zsgbe3N/72t781WX9ZWRnmzZuHwYMHw8PDA8HBwTh16pRqXz7//HMAQEJCAuzs7BAfH9/qeyYIApYvX46hQ4fCwcEBXl5eePHFF1X1VVRUYNGiRRg2bBhcXV0xadIkHDx4sNntHT58GH5+fnBzc0Pfvn3h4+OD2NhYAHc/fYeEhAAALly4oDZyVFtbi5iYGIwcORJOTk4YP3682iffw4cPY8KECXBycoKbmxvGjx+PlJSUFvetW7du+OSTT1Rfb7zxBoC7YW369Onw8vJCv3794O7ujtmzZ+PixYuqdS9duoSFCxfCx8cHjo6OGDVqFH7//Xe17Z8+fRr+/v5wcnLCzJkzcePGjSbrOHPmDK5evQp3d3fExcVhxYoV2Lp1K/773//i0UcfVfteiUQChUKBdevWtbhvABAdHY3PPvsMMTExAICsrCzVSMe97h8Nun37NhYvXgwvLy84ODhg6NChmD17dqv9TZgwQe39bBjdunPnDr788ks888wzcHJywqhRo7B+/XrU19cDaPp0w/3/FxtGF9etWwc/P78m39PNmzdj6NChGDBgAP71r3+1Wm9z/7+ff/55fPjhh3B3d8eQIUOQmJjY6rbIsDFMUJuaNGkSrl27hs2bN4vaTkFBAbKzs+Hs7IyioiKEhIRg9+7d8Pb2RmlpKSIjIxuts3r1agwYMADW1tZITU3FO++8o9pWaGgo/vvf/2L06NFwdHREfHw8XnvtNdx/N/kVK1ZgzJgxGDp0aKPt19TUYNq0adi9ezccHR0xcuRIZGRkYNq0aSgpKYG3tzcGDx4MAHB2dkZYWBhcXFxa3ddff/0V69atg7GxMaZPn47hw4ejoKAAcrkc9fX1mD17NrZt2wZra2v4+fkhNzcXs2bNQmZmZpPbKy8vh7W1NaZMmYKgoCDI5XKsWbMGycnJ6NWrFwICAgAAjzzyCMLCwhAWFgYAiIqKwldffQULCwsEBATg0qVLePXVV1XzEiIiIpCfn4+AgAAEBATAyMgIhYWFLe6bQqHABx98oPpKSkoCANTX1+Pq1asYNWoUXnzxRTzxxBPYt28f3n77bQB3g820adOwY8cOdO3aFUFBQejevTuuXLmitv2VK1fCxcUFXbt2xYEDB7B+/fom67CxsYGRkRH++OMPTJ06FTExMUhPT4eJiUmjESgLCwuMGjUKW7duxaVLl1rcPwC4deuW6lh069YNUqm01XV27NiBrVu3wtraGtOnT8fAgQObPZ73Sk1NVXs/KysrAdz9uf3ss88gl8sxdepUXL9+HR9//DG++uqrVrd5vzVr1sDd3b3Re3rs2DEsWbIEV65cwahRo5CQkKDR+9OUzMxMZGVlwcvLC5cvX8Y777yDqqqqh9oWGQae5qA2NWXKFBQWFuJf//oXhgwZ8tDbMTc3x48//ojs7GyEhITg5s2bSE5ORu/eveHq6orLly+joqICPXv2VK2zePFivPrqq8jLy4Ofnx9SUlIgl8uxZcsW3L59G56enrCxsYGNjQ1OnTqFY8eO4ezZs+jatatqG99++y2efvrpJmv65ZdfUFpair59+2LHjh2QSCQICwtDamoqtm3bhujoaJw6dQrZ2dnw8vLSeFi7rq4OANCvXz9MnjwZLi4u6NmzJwRBQE5ODk6dOgWpVIqkpCSYm5vD2toaGzduxObNmzFs2LBG2wsJCYGNjQ1yc3NRWVmJvn374saNGzh69CimTp2KOXPmICUlBT169FDVWFFRgeTkZBgZGWHo0KEwMjKCq6srMjIy8P333+Opp55CXV0dunbtqhr1cHR0bBTG7nf79m189913arUFBgaia9eu2LhxI/bt24e//voLbm5uOH36NDIyMlBfX4/9+/fj3LlzsLW1RVpaGszMzADc/QR+r6ioKLzxxhuIjY3FmjVrcPr06Sbr6NWrFz788EPExMQgKysLWVlZ+Oqrr9CvXz9s27YNTzzxRKPtTpkyBf/85z9hamra7P55eHiovV60aBG6dGn912rDfri5ueGFF16As7MzLCwsWl3v+PHjOH78uOr13Llz0aNHD/y///f/AADr1q3DiBEjsHfvXrzyyiuIi4vDW2+91ep279+Hpt7ThtGDadOmYdWqVaisrMTgwYNVox8PokePHkhISICxsTFkMhlqampQXFyMQYMGPfC2yDAwTFCbMjIyQkREBN544w3ExcW1+v1KpbLJdnt7e5iZmcHS0lLVJpPJYGxsrHpdU1OjFiacnZ0BAE5OTqq28vJy1TBudnY2srOz1fo5d+4c3NzcVK+b+uPc4MKFC6o6JBKJWl/3Ds8/qFGjRuHll19GQkKC6hSEl5cXNm3apKq9d+/eMDc3V+uzoZ77/eMf/8APP/zQqL2ioqLZGhr6qa+vb3TcSkpKANz99Lt06VLVfBgrKyssW7YMU6dObXa7lpaW+OOPPxq1nzhxAiEhIY2O/61bt1BdXa2qx83NTRUkADQaRWiYP9Dwc1JTU9NsLa+++iqmT5+OY8eO4dixY9i6dStKSkqwYcOGRvNbhgwZgmeffRbx8fEYO3Zss9ucOXMmzMzMYGtrC19fX7X5DC0JCQlBRkYG9u7di+TkZEgkEvj6+uK7775THeemfPTRR5g7d65a27Vr11T73fB/QCaTAbg7h+n27duNttPc/zug+ff08uXLatu2srKClZVVs5NOW+Ls7Ixu3boBuPvBoaqqCnK5/IG3Q4aDpzmozU2ePBnu7u7YtWtXo2UNfxgahjSbGyY3Mmr8o3lvkGhKUVERgLvnxxv06tUL9vb2AIB58+bh4sWLqq+MjIxGVxfcO0pxvz59+gC4O8my4RP52bNnAUDUFStKpRLLli3DH3/8gaNHjyI4OBi///47tm7dqqr90qVLqnkoDX021HO/hvd9zZo1KCsrw6xZswBAVXPD+3jvJ8qGfkxNTZGbm6t6j0pKSrBx40YAdyfTHT16FLm5uVi/fj0qKyuxYsWKh9rn3bt3Q6lUYsyYMThz5gx+/vln1TJBEFT1FBQUqM2/aRjFadAwCtAQ7ppTWVmJvLw8PPLII/Dz88NHH32kurKiurq6yXUWLVqE27dvt3jFRHR0ND788EO8/vrrGgcJ4O4x+Oabb1BYWIj09HT4+vri8OHDrc5BaUrPnj1VAaThZ79hIrCtrS1MTU1Vyxv2taCgoNntNfeeNkxGbfj5q6ysVJ1meVD3jt60duyofeDIBLU5iUSCiIgIzJs3r9Gy/v37IzMzE++99x5kMhn27t3bZv1+/vnnyM/PV53j9/f3h1QqxYwZM7B161Z89913KC0thbW1NYqKipCVldXsp/umPPfcc7C3t0dJSQlCQkJgZWWFPXv2oFu3bg99yR9wd9JeREQEhgwZgh49eiArKwsA0L17dwwaNAiDBw9GdnY2AgMD4eLiovok+/LLLze5PRsbG9y8eRObNm3C4cOHsWfPHrXlvXv3BnB31GbRokVwcHDAm2++icmTJ2PXrl2YNGkSfH19UVlZiZMnT2LmzJmIiorC+PHjYW9vj969e6vOld87ctSUhjkTDczMzBAdHa2a9JidnY333ntPbegeAMaOHQsHBwecO3cO48ePx5NPPomzZ89i3rx5GD9+/AO8u3dduXIFfn5+GDBggGokavfu3QCAZ555psl1Bg0ahOeeew6//PLLA/fXmp07d+Lrr7/GwIEDIZVKVX/cW3s/myKRSDBr1ix88803ePPNNzFmzBikpaUBgGpSZ0PQOXDgAD7++GMcOHDggft5/vnnsW3bNvz000+4desWCgoKGoU76rw4MkFaERAQ0OQntU8//RTu7u7Iy8tDeXk5QkND26zPyMhI5ObmoqKiAn5+fli5ciWAu79If/zxRwwfPhzHjx9HcnIyqqur8eabbz7Q9s3NzREfH4+AgACcOXMGR44cwYgRIxAfHw8HB4eHrvvxxx+Hg4MDjhw5gq1bt6ouY5wxYwaMjIywefNmhIaG4tq1a0hNTUX//v0RFxcHHx+fJrcXGxsLJycnFBUVobq6Gi+99JLacnt7e7z++uuwtLTEtm3bkJCQAODuzPsFCxZAIpFg+/btyMrKwpAhQzB69GgAgK+vL86ePYsdO3YgMzMTI0aMUF0l0pyGORMNX1u2bAEAzJkzBxMmTMCtW7dw/Phx/P3vf1dbz8zMDPHx8QgKCkJtbS127NiBa9euwdbW9mHeYtja2mLWrFm4desWUlNTsWvXLtjZ2eHjjz9GYGBgs+stWrToofprjUwmg7W1NQ4cOIAff/wRJiYmWLhw4UPfh+Odd97B4sWLYWZmhqSkJPTo0QPvv/++6mf8mWeeQVhYGLp164bU1FTMmTPngft4+umnsXTpUjz22GM4ePAgJk6c2KnuIUMtkwitzaAiIiIiagFHJoiIiEgUhgkiIiIShWGCiIiIRGGYICIiIlEYJoiIiEiUDh8meEMUIiIi7erwYYKIiIi0i2GCiIiIRGGYICIiIlEYJoiIiEiUdhEmjh07hmnTpiE4OLjRQ4uIiIhIvwz+qaEKhQLffvst/v3vf8PU1FTf5RAREdF9DH5kIisrC926dcPs2bMRFhaGq1ev6rskIiIiuofOwkRcXBz8/f3h4OCA8PBwtWWVlZUICwuDk5MTfHx8kJSUpFp27do1lJSUYPPmzXjxxRexatUqXZVMREREGtBZmLC1tcXChQsRGhraaNmSJUtgYmKCnJwcrFu3DtHR0SgsLAQAWFpaYtiwYTA1NcXTTz+NoqIiXZVMREREGtBZmAgICMCECRNgZWWl1l5TU4OUlBQsXrwYUqkUPj4+GDduHBISEgAAXl5eKCoqgiAIyMvLQ9++fXVVMhEREWlA7xMwi4uLYWxsDJlMpmrz9PRERkYGAMDa2hr+/v4ICgqCkZERT3MQEREZGL2HCblcDgsLC7U2CwsLyOVy1evZs2dj9uzZOq6MiIiINKH3qzmkUimqqqrU2qqqqiCVSvVUUcuUSuVDLSMiIuqo9D4y4ejoCKVSieLiYjg6OgIA8vPz4erqqufKmmZsbIy5c+eiuLhYrd3R0REbNmzQU1VERET6o7MwUVdXh7q6OtTX10OpVEKhUKBLly4wNzeHv78/YmNjERsbi7y8PKSlpSE5OVlXpT2w4uJiFBQU6LsMIiIig6Cz0xxr166FTCbDunXrkJiYCJlMhrVr1wIAYmJioFAoMHDgQMyfPx/Lly832JEJIiIiUicRBEHQdxHaJJFI0Na7OHbs2EYjE25ubti/f3+b9kNERNQe6H0CJhEREbVvDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoDBNEREQkCsMEERERicIwQURERKIwTBAREZEoBh8mysrKMGDAAAQHByM4OBgVFRX6LomIiIju0UXfBWjiySefxIYNG/RdBhERETXB4EcmACArKwuBgYFYvnw5BEHQdzlERER0D52Fibi4OPj7+8PBwQHh4eFqyyorKxEWFgYnJyf4+PggKSlJteyxxx7D0aNHkZiYiIqKCqSkpOiqZCIiItKAzsKEra0tFi5ciNDQ0EbLlixZAhMTE+Tk5GDdunWIjo5GYWEhAKBr164wNzeHRCKBv78/8vPzdVUyERERaUBnYSIgIAATJkyAlZWVWntNTQ1SUlKwePFiSKVS+Pj4YNy4cUhISAAAVFVVqb735MmT6Nevn65KJiIiIg3ofc5EcXExjI2NIZPJVG2enp6qkYmMjAxMmDABgYGBKC8vR2BgoL5KJSIioibo/WoOuVwOCwsLtTYLCwvI5XIAgJ+fH/z8/PRRGhEREWlA7yMTUqlU7VQGcPfUhlQq1VNFRERE9CD0HiYcHR2hVCpRXFysasvPz4erq6seqyIiIiJN6SxM1NXVQaFQoL6+HkqlEgqFAnV1dTA3N4e/vz9iY2NRU1ODzMxMpKWlISgoSFelERERkQg6CxNr166FTCbDunXrkJiYCJlMhrVr1wIAYmJioFAoMHDgQMyfPx/Lly/nyAQREVE7IRE6+C0lJRJJm981c+zYsSgoKFBrc3Nzw/79+9u0HyIiovZA73MmiIiIqH1jmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEkWjMJGamoovvvgCgiAgOTkZI0eOxMSJE5GXl6ft+oiIiMjAaRQmvvjiC+zbtw8A8N577+H8+fPIycnBJ598otXi7rVz504MGDBAZ/0RERGRZjQKE+fPn4e7uzuKi4tRWVmJ9evX49lnn0Vubq626wMA1NfXY/fu3ejdu7dO+iMiIiLNaRQm6urqUFdXh8LCQkgkEjz11FPo3bs3bt++re36AABJSUmYOHEijIw4xYOIiMjQaPTX2cnJCTt37kRERAT69euHHj164OLFi7C1tdW4o7i4OPj7+8PBwQHh4eFqyyorKxEWFgYnJyf4+PggKSlJtUypVGLXrl2YMmWKxn0RERGR7mgUJqKjo9GzZ0+YmZnh/fffx+3bt3HhwgWMHTtW445sbW2xcOFChIaGNlq2ZMkSmJiYICcnB+vWrUN0dDQKCwsBAAkJCZg8eTJHJYiIiAxUF02+6ZlnnsFvv/2m1nbw4MEH6iggIAAAkJOTg/LyclV7TU0NUlJSsH//fkilUvj4+GDcuHFISEjAu+++i6KiIpw+fRqJiYk4d+4c3n//fSxduvSB+iYiIiLtaTFM1NXVYfXq1Thw4AB69uyJ8PBwDBs2rE0LKC4uhrGxMWQymarN09MTGRkZAO6OWjTw9/dnkCAiIjIwLYaJf/7zn/jyyy8hCAIA4MSJEzh+/DhsbGzarAC5XA4LCwu1NgsLC8jl8kbfu2fPnjbrl4iIiNpGixMREhMTYWpqijlz5sDHxwe3bt3C7t2727QAqVSKqqoqtbaqqipIpdI27YeIiIi0o8WRiUuXLmHKlClYunQpKisrMWDAAFy6dKlNC3B0dIRSqURxcTEcHR0BAPn5+XB1dW3TfoiIiEg7WhyZuH37Nnr06AEAsLKyAgDcunXroTqqq6uDQqFAfX09lEolFAoF6urqYG5uDn9/f8TGxqKmpgaZmZlIS0tDUFDQQ/VDREREuiURGiZENKFPnz7o2bMnHn/8cQBAXl6e2muJRILU1FSNOlq1ahVWr16t1hYZGYmoqChUVlYiKioKhw8fhpWVFd59910EBgY+7D6pkUgkaGEXH8rYsWNRUFCg1ubm5ob9+/e3aT9ERETtQathosWVJRKUlZW1eVFtiWGCiIhIu1qcM3H/SAIRERHR/VoME9OmTdNVHURERNROtRgmtm/f3uoGQkJC2qwYIiIian9aDBMRERGQSCTNLpdIJAwTREREnVyLYcLd3b1RmKivr0dhYWGbT2okIiKi9qnFMLFv3z611//5z3+wZs0aCIKARx99FPPnz9dqcURERGT4NHpq6J49e7Bq1SoUFhbC2toa7733Hl5++WWYmZlpuz4iIiIycC2GibS0NKxatQr5+fmwsrJCdHQ05syZwxBBREREKi2GiVdeeQUSiQQSiQQeHh44deoUTp06pVoukUiwYcMGrRdJREREhqvV0xyCIEAQBBw5cqTRspau9CAiIqLOocUwERkZqas6iIiIqJ1imCAiIiJRNLqao4FSqcS+fftQXV2N4cOHw97eXlt1ERERUTvRYpj44IMPcOjQIezatQvdu3fHyy+/jPT0dABA165dERcXB19fX50USkRERIbJqKWFhw8fhpWVFbp3747c3FwcOnQIgiDA3d0dCoUCa9as0VWdREREZKBaDBOXL1+Gi4sLAKhGJAIDA5GWlgY/Pz8UFhZqv0IiIiIyaC2GCaVSibq6OgDAb7/9BolEgjFjxgAAbGxscPv2be1XSERERAatxTkTDg4OSElJgaWlJdLT0yGRSDBy5EgAQElJCR599FGdFElERESGq8WRiQULFqC2thbfffcdbt++jeeffx62tra4evUqMjMzMXjwYF3VSURERAaqxZGJKVOmwMHBARkZGbC1tcXkyZMBADdv3sSHH36IoUOH6qRIIiIiMlyt3mdiwIABGDBggFqbk5MTnJyctFYUERERtR8tnuYgIiIiag3DBBEREYnCMEFERESiNBsmPv30U+zbtw8AsH37duTk5OisKCIiImo/mg0T33zzDY4cOQIAiIiIwM6dO3VWFBEREbUfzV7NYWlpiaSkJPzvf/8DcPd22vc/klwikWDVqlXarZCIiIgMWrNhws/PDzt27MCOHTsgkUjw559/4s8//1T7HoaJ/2NjYwOlUgljY+Mml7e0jIiIqD1rNkysXr0aY8eOxdmzZxEbG4vBgwernstBjVlYWMDY2Bhz585FcXGx2jJHR0ds2LBBT5URERFpV7NhwsjISHXHy/r6enh7e2P06NG6qqvdKi4uRkFBgb7LICIi0plW74AJAJGRkSgtLcWKFStw4cIF2NvbIzQ0FH379tV2fURERGTgNAoT2dnZCA0NRW1tLQRBAABs3LgR8fHxfNgXERFRJ6fRTatiYmJQU1OD8ePH480338SECRNQU1OD5cuXa7s+/PXXX5gyZQqCgoIQEhKCK1euaL1PIiIi0pxGIxP5+fkICAjA+vXrVW3z5s3D0aNHtVZYA2tra+zcuRNGRkaIj4/Htm3bEB4ervV+iYiISDMahYmuXbvi+vXram3Xr19H165dtVLUve69nFIul8PV1VXrfRIREZHmNDrNMXToUJw4cQIjR47ErFmz8PTTT+PEiRMYNmyYxh3FxcXB398fDg4OjUYWKisrERYWBicnJ/j4+CApKUlt+enTpzFp0iTExcU1ehw6ERER6ZdGIxMffPABcnNzcf78eZw/fx4AYG9vj/fee0/jjmxtbbFw4UIcOnQICoVCbdmSJUtgYmKCnJwc5OXlYdasWfDw8FCNQvTv3x8///wz/vOf/+DLL7/EihUrNO6XiIiItEujMNGnTx8cOHAAv/zyCy5cuIA+ffrgueeeg5mZmcYdBQQEAABycnJQXl6uaq+pqUFKSgr2798PqVQKHx8fjBs3DgkJCXj33Xdx69Yt1ekUS0vLB+qTiIiItE+jMAEAZmZmqptYtaXi4mIYGxtDJpOp2jw9PZGRkQEAyM3NRUxMDIyMjNCtWzfevvsevH03EREZAo3DhLbI5XJYWFiotVlYWEAulwO4O18jMTFRH6UZPN6+m4iIDIHew4RUKkVVVZVaW1VVFaRSqZ4qal94+24iItI3ja7m0CZHR0colUq1T9f5+fm8BJSIiKidaDVM1NXVYdq0aVi5cqWojurq6qBQKFBfXw+lUgmFQoG6ujqYm5vD398fsbGxqKmpQWZmJtLS0hAUFCSqPyIiItKNVsNEly5dcPbsWfz111+iOlq7di1kMhnWrVuHxMREyGQyrF27FsDd23UrFAoMHDgQ8+fPx/LlyzkyQURE1E5oNGciPDwcK1euxAsvvIChQ4fCxMTkgTuKiopCVFRUk8usrKywadOmB94mERER6Z9GYSI6OhoSiQTTpk1Ta5dIJCgtLdVKYURERNQ+aHw1R8Ojx1sqJOaBAAAegUlEQVRrIyIios5FozBx/PhxbddBRERE7ZTGt9MGgLNnz6KkpARjx47ValFERETUfmh0n4nKykqEhoZi9OjRmDNnDsrKyvDEE0+IvlyUiIiI2j+NwsSnn36Ko0ePwtTUFIIgwN7eHt7e3vjll1+0XR8REREZOI3CxKFDhzBixAi89NJLqjYXFxdeyUFERESahQmFQgFbW1u1tuvXrz/U/SaIiIioY9EoTLi7u+OXX35BdnY2AGDp0qXYt28fPDw8tFocERERGT6NwsTbb7+N27dv49SpUxAEAd9++y2MjIyavaMlERERdR4aXRrq4+ODlJQUbNmyBRcuXIC9vT1mzJjBkQkN2djYQKlUwtjYuMnlLS0jIiIydBrfAdPNzQ3vv/8+ysvL0atXL5iZmWmzrg7FwsICxsbGmDt3rtqj1oG7j2DfsGGDniojIiIST6MwcePGDbz99tvYs2ePqs3f3x8rVqyAlZWV1orraIqLi1FQUKDvMoiIiNqURmFi0aJFSE1NVWvbs2cP6uvrsXHjRq0URkRERO2DRhMwf/31VzzxxBNIS0tDcXEx9u7diz59+uDXX3/Vdn1ERERk4DQamejbty8GDhyomnDp6emJp556CqdPn9ZqcURERGT4mg0T9z4pNDg4GGvWrIGnpydkMhnOnDmDlJQUREZG6qTIjoxXehARUXvXbJgIDg6GRCJRa/vggw9U/xYEAUuXLsXcuXO1V10noI0rPRhQiIhIl5oNE3Z2do3CBGlPW17pwUtRiYhIl5oNEydOnNBlHaQFvBSViIh0QeObVgF3H+5VW1ur1mZnZ9emBREREVH7olGYOHz4MKKionD58mW1dolEwseQa1Frcx86q+beE75XRET6oVGY+Mc//oHy8vJG7YIgtHlB9H9amvvg6+uLjz76SD+F6VlT7wnnghAR6Y9GYaKyshKjR4/Gt99+C6lUqu2a6D5NzX1wcHDQUzWGgfNBiIgMh0Z3wJwzZw4uXryI8vJyjkYQERGRGo1GJgICArB582aMGTNGrZ1zJoiIiEijMPHWW2/h5s2bjdo5SkFEREQahYmLFy/Cy8sLS5YsgaWlpbZrIiIionZEozAxY8YM/P777xgyZAhMTU21XRMRERG1IxqFiSNHjqCwsBADBw5Enz59VNfySyQSpKamarVAIiIiMmwahYk//vgDAFBdXa12OR6f3UFEREQahYnVq1druw4iIiJqpzQKE9OmTdN2HURERNROaRQmIiMjm2yXSCRYtWpVmxbUlKysLHz88ccwNTWFra0t1q5dCxMTE633S5pp6ZkYfF4GEVHHp1GY+OmnnyCRSFT3lWj4t67CRJ8+ffDTTz/BzMwMn332Gfbu3YtJkyZpvV/STHPPD+HzMoiIOocHHplQKpUoKChAWloapk+frrXC7vX444+r/t2lSxcYGWl0F3DSobZ+VgZHNIiI2o+HPs3x9ttvN/kk0ZbExcXhp59+QkFBAaZOnYovvvhCtayyshKLFi1Ceno6rK2tER0djcDAQLX1S0tLcfDgQfz9739/oH6p/eHTUomI2g+N74B5r6qqKpw9exb5+fkP1JmtrS0WLlyIQ4cOQaFQqC1bsmQJTExMkJOTg7y8PMyaNQseHh5wdXVV9RkeHo61a9fyxllaYmhzH/i0VCLqKAzt92tb0yhMPPnkk022u7m5PVBnAQEBAICcnBy1UY2amhqkpKRg//79kEql8PHxwbhx45CQkIB3330XdXV1mD9/PiIjI+Hk5PRAfZLmdD33oSP8B2rPOvovNyJD0tHnlmkUJu5/oJeZmRn69++PpUuXtkkRxcXFMDY2hkwmU7V5enoiIyMDALBz505kZ2fjiy++wBdffIGZM2di6tSpbdI3qWvruQ8t4akM/erov9yIDI0uf7/qmkZh4sKFC1otQi6Xw8LCQq3NwsICcrkcABAcHIzg4GCt1kD6wVMZ+tWRf7kRtQc2NjYdYpRQozChbVKpFFVVVWptVVVVkEqleqqIiIhI+ywsLDrEKGGLYcLe3r7FlSUSCUpLS0UX4ejoCKVSieLiYjg6OgIA8vPzVZMviYiIOrL2PkrY4g0bBEFo9etB1NXVQaFQoL6+HkqlEgqFAnV1dTA3N4e/vz9iY2NRU1ODzMxMpKWlISgoSNTOERERkfa1ODKxd+9etdc1NTWIi4vDrl27IAgCPD09H6iztWvXqj00LDExEZGRkYiKikJMTAyioqIwcOBAWFlZYfny5RyZICIiagdaDBMNYUGhUGDz5s345ptvcO3aNbi5uSEyMlJ1qaemoqKiEBUV1eQyKysrbNq06YG2R0RERPrXYpi4desWtmzZgq+//hrXrl2Ds7Mzli5dismTJ+uqPiIiIjJwLYaJp556ClevXoVEIsHkyZMxZcoUGBkZIS0tTfU9fn5+Wi+SiIjoQXWESy7bixbDxJUrV1RPCN21axd27dqltrytruYgIiJqax3hksv2osUwYWdnB4lEoqtaiIiI2lR7v+SyvWgxTJw4cUJXdRCRgegod+QjIt0xiDtgEpHh6Ch35CMi3WGYIKImcXiYiDTV4h0wiYjoLqVS+UDtRJ0JRyaIiDTQ1KkfnvYhuothopNpbXIdETWPp36ImsYw0ck0N7nO19cXH330kf4KIyKidothopO6/xOWg4ODHqvpeHhpJRF1JgwTRFrASyuJqDNhmCDSEp5fJ6LOgpeGEhERkSgME0RERG2gM99zhKc5iIiI2kBzc6U6w9VyDBNERERtpKm5Up3hajme5iAiIiJRGCaIiIhIFIYJ0pqGW3cTEVHHxjkTpDXN3bob6BwTkohIN3hXWf1jmCCt66wTkprS2oPW+EuR6MHxeUP6xzBBpEMtjdbwVttED4/PG9IvhgkiPeCtton0hyOEbY9hgoiIDF5b/oHnCGHbY5ggIiKDp43J3BwhbDsME0QdGIdrqSPhZG7DxTBB1IHx0lwi0gWGCaIOjp/miEjbeAdMIiIiEoVhgoiIiERhmCAiIiJRDD5M3Lx5ExMnToSzszMv4SEiIjJABh8mzMzMsGXLFkycOFHfpRARkRbxKcPtl8FfzWFiYoKePXvqu4xOrbVbzxIRtQVeytx+6SxMxMXF4aeffkJBQQGmTp2KL774QrWssrISixYtQnp6OqytrREdHY3AwEBdlUat4KPEiUhXeClz+6SzMGFra4uFCxfi0KFDUCgUasuWLFkCExMT5OTkIC8vD7NmzYKHhwdcXV11VR5pwJD/k/PBPURE+qOzMBEQEAAAyMnJQXl5uaq9pqYGKSkp2L9/P6RSKXx8fDBu3DgkJCTg3Xff1VV51M7xwT1ERPqj9zkTxcXFMDY2hkwmU7V5enoiIyND9XrmzJnIy8vD2bNn8dJLLyE0NFQfpVI7wAf3EBHpnt7DhFwuh4WFhVqbhYUF5HK56vX333+v67KIiIhIQ3q/NFQqlaKqqkqtraqqClKpVE8VERERNY2XrzZN7yMTjo6OUCqVKC4uhqOjIwAgPz+fky+JiMjg8Mq2puksTNTV1aGurg719fVQKpVQKBTo0qULzM3N4e/vj9jYWMTGxiIvLw9paWlITk7WVWlE7RqvVCHSLUO+sk1fdBYm1q5di9WrV6teJyYmIjIyElFRUYiJiUFUVBQGDhwIKysrLF++nCMTRBriJyUi0jedhYmoqChERUU1uczKygqbNm3SVSlEHQ4/KRGRPul9AiYREXUenMDYMel9AiYREXUePC3XMTFMEBGRTvG0XMfD0xzU4TU8t6M57WHYtT3U2Bl1hJ8torbAkQnq8DrCczs4NGyYOsLPFlFbYJigTqO9P7eDQ8OGq73/bBGJxdMcREREJArDBFELdHk+vLXz76TuYY+Nro4p51NQZ8LTHEQt0OX58ObOv3NeRNMe9tjo6phyPgV1JgwTRK3Q9fnw+/vjvIjmPeyx0eUx5XwK6gx4moOIiIhEYZigTu1h5ylwfkPH0xGOqS7naHTU+SAd4edAH3iagzq1ls5rtzRX4WHXI8PVEY6pLudodNT5IB3h50AfGCaI8PD3cOC9Hzqe9n5MOR+kbbT3nwNd42kOIiIiEoVhgoiIiERhmCAiIiJRGCaISGO8q2Pb4PtIHQ0nYBKRxnhXx7bB95E6GoYJInpgHXkWvy7xfaSOgqc5iIiISBSGCSIiIhKFYYKIiIhEYZggIiIiURgmiIiISBSGCSKiTq65+1rU1dU98Lb41M3OiZeGEhF1ck3d86LhCZl8oi5pgmGCiIga3fOi4QmZfKIuaYKnOYiIiEgUhgkiIiIShWGCiIiIRGGYICIiIlEYJoiIiEiUdhEmPvvsMwQGBmLu3Lmora3VdzlERER0D4MPEwUFBTh//jySkpLg6+uLH3/8Ud8lERER0T0MPkycOHECY8aMAQA8++yzOHnypJ4rIiIionvpLEzExcXB398fDg4OCA8PV1tWWVmJsLAwODk5wcfHB0lJSapl//vf/2BpaQng7p3Vbty4oauSiYiISAM6uwOmra0tFi5ciEOHDkGhUKgtW7JkCUxMTJCTk4O8vDzMmjULHh4ecHV1Rffu3XHz5k0AQFVVFXr06KGrkomIiEgDOhuZCAgIwIQJE2BlZaXWXlNTg5SUFCxevBhSqRQ+Pj4YN24cEhISAAA+Pj5IT08HABw6dAjDhg3TVclERESkAb3PmSguLoaxsTFkMpmqzdPTE4WFhQAAd3d39OnTB4GBgUhPT8f06dP1VSoRtaC1p0U29wTKlp5M+TBPn2zvT6182PeRSJ/0/qAvuVwOCwsLtTYLCwvI5XLV6+joaF2XRUQPSJOnRT7IkykdHR2xYcMGrdRhyB7mfbx3GZE+6D1MSKVSVFVVqbVVVVVBKpXqqSIiEqOlp0U+yJMptVlHe/Ag7+O9y4j0Qe+nORwdHaFUKtVSdn5+PlxdXfVYFREREWlKZ2Girq4OCoUC9fX1UCqVUCgUqKurg7m5Ofz9/REbG4uamhpkZmYiLS0NQUFBuiqNiIiIRNBZmFi7di1kMhnWrVuHxMREyGQyrF27FgAQExMDhUKBgQMHYv78+Vi+fDlHJoiIiNoJnc2ZiIqKQlRUVJPLrKyssGnTJl2VQkRERG1I73MmiIiIqH1jmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRJIIgCPouQpskEom+SyAiIjIY2vizr/dnc2hbB89KREREesfTHERERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKwwQRERGJwjBBREREojBMEBERkSgME0RERCQKw4SGKisrERYWBicnJ/j4+CApKUnfJT204OBgODo6wtnZGc7OzvD19VUtS0pKgo+PD5ycnPDKK6+gsrJStay196CldXUpLi4O/v7+cHBwQHh4uNqyX3/9Fc888wxkMhmCg4Nx4cIF1bJbt24hMjISrq6u8PLywrfffttm6+pCc/tdVlYGOzs71fF2dnbGmjVrNK7dkPf71q1biIqKgo+PD1xcXODn54cDBw60Se2Gut8t7XNHPtZvvfUWBg8eDFdXVzz99NPYunVrm9TdHvfZII+zQBp54403hNdee02orq4WTpw4Ibi6ugoFBQX6LuuhBAUFCT/88EOj9oKCAsHZ2VnIyMgQqqurhfnz5wuvv/66anlL70Fr6+rS7t27hT179gjvvPOOsHDhQlV7RUWF4OrqKvznP/8RamtrhU8++USYOHGianlMTIzw/PPPC5WVlcKff/4pDBo0SDhw4IDodXWluf0uLS0VevfuLdy5c6fJ9drzfsvlciE2NlYoLS0VlEqlkJaWJjg7OwulpaUd9ni3tM8d+VgXFBQICoVCEARBKCoqEgYNGiTk5OR02OPc0j4b4nFmmNCAXC4X+vbtK5w5c0bVtmDBAmHZsmV6rOrhNRcmYmJihPnz56tenzt3Tujbt69QVVXV6nvQ0rr68tlnn6n9Uf3++++FyZMnq17L5XLB0dFRKCoqEgRBELy9vYVDhw6plq9YsUIViMSsq2v373drv3g6yn43GDt2rPDzzz93muMtCP+3z53lWBcVFQleXl5CcnJypznO9+6zIR5nnubQQHFxMYyNjSGTyVRtnp6eKCws1GNV4ixfvhz9+/fH1KlTcezYMQDAn3/+CQ8PD9X39OvXDyYmJiguLm71PWhpXUNxf43m5ubo27cvCgsLcePGDVy+fFltuYeHB/7880/R6xqK4cOHY8iQIYiIiMD169cBoMPt919//YXi4mK4urp2muN97z436KjHOjo6GjKZDKNGjcJjjz2GsWPHdvjj3NQ+NzCk48wwoQG5XA4LCwu1NgsLC8jlcj1VJM67776LjIwM/Pbbb5gxYwZmz56NkpKSJvfT0tIS1dXVrb4HLa1rKORyOSwtLdXaLCwsVPvX8LrBvfWLWVffrK2tkZKSghMnTiA1NRXV1dVYsGABAHSo/b5z5w4WLFiA4OBgODk5dYrjff8+d/RjvXz5cvz5559ISkqCv78/TE1NO/xxbmqfDfE4M0xoQCqVoqqqSq2tqqoKUqlUTxWJ4+3tjUceeQRdu3bFtGnTMGzYMBw4cABSqbTRD01VVRUeeeSRVt+DltY1FE3tQ3V1tWr/Gl43uLd+Mevqm1QqxaBBg9ClSxc8+uijWLZsGdLT09WOX3vf7/r6evz973+Hqakpli1bBqDjH+/m9rmjH2tjY2P4+PigvLwcW7Zs6fDHGWh6nw3tODNMaMDR0RFKpVJtyD4/P19tWLE9k0gkEAQBLi4uyM/PV7WfP38et2/fhqOjY6vvQUvrGor7a6ypqUFJSQlcXV3Ro0cP2Nraqi3Pz8+Hi4uL6HUNjUQiAQAIgtAh9lsQBERFReGvv/7C+vXrYWJiIrp2Q9/v5vb5fh3tWN9LqVTi/PnzHfo4369hn+9nEMf5AeeAdFqvv/668MYbbwhyuVw4efJku72a48aNG8LBgweF2tpa4c6dO0JCQoIgk8mEoqIioaCgQHBxcRGOHz8uyOVyYcGCBWoTb1p6D1pbV5fu3Lkj1NbWCjExMcKCBQtU+3rt2jXB1dVV+Pnnn4Xa2lrh008/VZvFvGzZMuGFF14QKisrVZOdGmYxi1lX3/v922+/CUVFRYJSqRQqKiqE1157TQgKCuow+/32228LEydOFKqrq9XaO/Lxbm6fO+qx/uuvv4SdO3cK1dXVQl1dnXDw4EFBJpMJqampHfY4t7TPhnicGSY0dP36dWHOnDmCTCYThg4dKiQmJuq7pIdy7do1wd/fX3B2dhbc3NyEiRMnCunp6arliYmJwtChQwWZTCbMnj1buH79umpZa+9BS+vqUmxsrNC7d2+1r9jYWEEQBCE9PV3w9fUVHB0dhaCgIKG0tFS1nkKhECIiIgQXFxdh4MCBwjfffKO2XTHr6kJz+52UlCQMHz5ckMlkgpeXl/DWW28JV65c0bh2Q97vsrIyoXfv3oKDg4Pg5OSk+kpISBBdu6Hud0v73FGP9bVr14QXXnhBcHNzE1xcXIRnn31W+Pe//90mdbfHfTbE4ywRBEF42CEXIiIiIs6ZICIiIlEYJoiIiEgUhgkiIiIShWGCiIiIRGGYICIiIlEYJoiIiEgUhgmiTmD48OGws7NTPdRNn5RKJcLDw+Hu7g47Ozt88sknWu8zPDwcdnZ2WLVqldb7IuqMGCaIdKThD7q3tzdqa2sBAKdPn4adnR3s7Oz0XJ3u7N69G9u3b4exsTFeeeUVPPnkk42+59ixY7Czs4O7u7seKvy//ocPH66X/onaG4YJIh27cuUKtmzZou8yRLtz585DrdfwfJdnn30WS5cuhZ+fX1uWRUR6wDBBpGMSiQRff/21anTifg0jFWVlZQCAVatWwc7ODuHh4QCA+Ph42NnZ4bnnnsNHH30EZ2dnjB49GqdPn8bKlSvh5uaGESNGID09vdG2c3NzMW7cOLi4uCAsLAzXr19XLTt58iSCg4Ph4eEBb29vREZGqpaXlZWp6tqyZQu8vb3xt7/9rcn6y8rKMG/ePAwePBgeHh4IDg7GqVOnVPvy+eefAwASEhJgZ2eH+Pj4Vt8zQRCwfPlyDB06FA4ODvDy8sKLL76oqq+iogKLFi3CsGHD4OrqikmTJuHgwYPNbu/w4cPw8/ODm5sb+vbtCx8fH8TGxgK4OyoREhICALhw4YLayFFtbS1iYmIwcuRIODk5Yfz48UhNTVXb7oQJE+Dk5AQ3NzeMHz8eKSkpre4fUXvHMEGkY5MmTcK1a9ewefNmUdspKChAdnY2nJ2dUVRUhJCQEOzevRve3t4oLS1FZGRko3VWr16NAQMGwNraGqmpqXjnnXdU2woNDcV///tfjB49Go6OjoiPj8drr72G+++4v2LFCowZMwZDhw5ttP2amhpMmzYNu3fvhqOjI0aOHImMjAxMmzYNJSUl8Pb2xuDBgwEAzs7OCAsL0+iJhL/++ivWrVsHY2NjTJ8+HcOHD0dBQQHkcjnq6+sxe/ZsbNu2DdbW1vDz80Nubi5mzZqFzMzMJrdXXl4Oa2trTJkyBUFBQZDL5VizZg2Sk5PRq1cvBAQEAAAeeeQRhIWFISwsDAAQFRWFr776ChYWFggICMClS5fw6quvquaiREREID8/HwEBAQgICICRkREKCwtb3T+i9q6Lvgsg6mymTJmCwsJC/Otf/8KQIUMeejvm5ub48ccfkZ2djZCQENy8eRPJycno3bs3XF1dcfnyZVRUVKBnz56qdRYvXoxXX30VeXl58PPzQ0pKCuRyObZs2YLbt2/D09MTNjY2sLGxwalTp3Ds2DGcPXsWXbt2VW3j22+/xdNPP91kTb/88gtKS0vRt29f7NixAxKJBGFhYUhNTcW2bdsQHR2NU6dOITs7G15eXhpPvqyrqwMA9OvXD5MnT4aLiwt69uwJQRCQk5ODU6dOQSqVIikpCebm5rC2tsbGjRuxefNmDBs2rNH2QkJCYGNjg9zcXFRWVqJv3764ceMGjh49iqlTp2LOnDlISUlBjx49VDVWVFQgOTkZRkZGGDp0KIyMjODq6oqMjAx8//33eOqpp1BXV4euXbuqRj0cHR0bhTGijohhgkjHjIyMEBERgTfeeANxcXGtfr9SqWyy3d7eHmZmZrC0tFS1yWQyGBsbq17X1NSohQlnZ2cAgJOTk6qtvLxcdUolOzsb2dnZav2cO3cObm5uqtdN/XFucOHCBVUdEolEra+LFy+2sJctGzVqFF5++WUkJCSoTkF4eXlh06ZNqtp79+4Nc3NztT4b6rnfP/7xD/zwww+N2isqKpqtoaGf+vr6RsetpKQEwN1Rm6VLl+K1114DAFhZWWHZsmWYOnWqprtK1C7xNAeRHkyePBnu7u7YtWtXo2VmZmYAgKqqKgBodpjcyKjxf997g0RTioqKAABnzpxRtfXq1Qv29vYAgHnz5uHixYuqr4yMDIwbN05tG/eOUtyvT58+AO5Osmz4RH727FkAEHXFilKpxLJly/DHH3/g6NGjCA4Oxu+//46tW7eqar906ZJqHkpDnw313K/hfV+zZg3Kysowa9YsAFDV3PA+1tfXq9Zp6MfU1BS5ubmq96ikpAQbN24EAIwZMwZHjx5Fbm4u1q9fj8rKSqxYseKh95uoveDIBJEeSCQSREREYN68eY2W9e/fH5mZmXjvvfcgk8mwd+/eNuv3888/R35+vuocv7+/P6RSKWbMmIGtW7fiu+++Q2lpKaytrVFUVISsrKxmP9035bnnnoO9vT1KSkoQEhICKysr7NmzB926dcP06dMfuu6srCxERERgyJAh6NGjB7KysgAA3bt3x6BBgzB48GBkZ2cjMDAQLi4uSE5OhkQiwcsvv9zk9mxsbHDz5k1s2rQJhw8fxp49e9SW9+7dG8DdUZtFixbBwcEBb775JiZPnoxdu3Zh0qRJ8PX1RWVlJU6ePImZM2ciKioK48ePh729PXr37o1Lly4BgNrIEVFHxZEJIj0JCAiAp6dno/ZPP/0U7u7uyMvLQ3l5OUJDQ9usz8jISOTm5qKiogJ+fn5YuXIlAMDT0xM//vgjhg8fjuPHjyM5ORnV1dV48803H2j75ubmiI+PR0BAAM6cOYMjR45gxIgRiI+Ph4ODw0PX/fjjj8PBwQFHjhzB1q1bUVtbi5kzZ2LGjBkwMjLC5s2bERoaimvXriE1NRX9+/dHXFwcfHx8mtxebGwsnJycUFRUhOrqarz00ktqy+3t7fH666/D0tIS27ZtQ0JCAoC7V6MsWLAAEokE27dvR1ZWFoYMGYLRo0cDAHx9fXH27Fns2LEDmZmZGDFihOoqEaKOTCJwdhARERGJwJEJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEoVhgoiIiERhmCAiIiJRGCaIiIhIFIYJIiIiEuX/A7QmxZVSS54gAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 540x378 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig , ax = plt.subplots()\n",
    "\n",
    "n , bins , patches = plt.hist(SNV_DF.num_isolates.astype(float) , bins = 60 , color = 'k' , rwidth = 1.0 , edgecolor='white', linewidth=1.35)\n",
    "\n",
    "plt.title('Number of Isolates Each SNP is Found in', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.xlabel('Number of Isolates', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.ylabel('Number of SNPs', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "ax.set_yscale(\"log\", nonposy='clip')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(7.5, 5.25)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.tick_params(axis='y', which='major', labelsize=12 , labelcolor = 'k')\n",
    "plt.tick_params(axis='x', which='major', labelsize=12 , labelcolor = 'k')\n",
    "\n",
    "#file_name = '/home/rv76/Farhat_Lab/Figures/Final_5/Scraping_rollingDB/Mean_Coverage_Distribution_across_all_isolates.pdf'\n",
    "#plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *1* isolate with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035, 4)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 1])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *2* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(429880, 4)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 2])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *3* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(262010, 4)"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 3])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *4* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(193209, 4)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 4])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### How many SNP sites have at least *5* isolates with an alternate allele?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(153603, 4)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF[SNV_DF.num_isolates >= 5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Keep only SNVs in which the corresponding *Reference Position* has a SNV in $\\ge$ 1 isolate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SNVs found in only 0 isolates are dropped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_DF = SNV_DF[SNV_DF.num_isolates >= 1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035, 4)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "      <td>11</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>1</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>1</td>\n",
       "      <td>16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  ref  alt num_isolates  pos\n",
       "0   A  [C]            3    8\n",
       "1   G  [C]            1   10\n",
       "2   A  [C]           11   11\n",
       "3   C  [T]            1   15\n",
       "4   G  [C]            1   16"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_DF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Re-Structure DataFrame to include only the Reference Position, Reference Allele and Alternate Allele(s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_annotation_DF = pd.DataFrame(index = range( 0 , np.shape(SNV_DF)[0] ) ) #index for each SNV\n",
    "SNV_annotation_DF['pos'] = list(SNV_DF.pos) #H37Rv Reference Position\n",
    "SNV_annotation_DF['ref'] = list(SNV_DF.ref) #reference allele\n",
    "SNV_annotation_DF['alt'] = list(SNV_DF.alt) #alterante allele(s) list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035, 3)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNV_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref  alt\n",
       "0    8   A  [C]\n",
       "1   10   G  [C]\n",
       "2   11   A  [C]\n",
       "3   15   C  [T]\n",
       "4   16   G  [C]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_annotation_DF.head() #index = row indices of genotype matrix , 0 supports ref, 1 supports alt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### This is the DataFrame which we'll use to extract SNVs of interest when iterating over all of the VCF files in rollingDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Output Annotation for SNVs of interest in Genotypes folder"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create a text file with a single column of all Reference Positions that we're interested in based off of DF above. We will use this text file to subset the *full* VCF files to only the positions that Reference Positions that have a SNV in at least 1 isolate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_ref_positions = ['NC_000962.3\\t' + ref_pos for ref_pos in list( SNV_annotation_DF.pos.astype(str) ) ]\n",
    "\n",
    "#save each element as a new row in a text file\n",
    "with open('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/ref_positions_to_select.txt', 'w') as f:\n",
    "    for item in SNV_ref_positions:\n",
    "        f.write(\"%s\\n\" % item)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6] This section of the notebook prepares the script that iterates over all of the VCF files in rollingDB and outputs a genotype vector."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "import vcf\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_ID = 'Peru4988' #INPUT - isolate ID"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load SNV-Index annotation (for indexing genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNV_annotation_DF = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_SNP_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref  alt\n",
       "0    8   A  [C]\n",
       "1   10   G  [C]\n",
       "2   11   A  [C]\n",
       "3   15   C  [T]\n",
       "4   16   G  [C]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNV_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Convert SNV annotation into dictionaries with keys = ROW indices of genotype array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#keys: H37Rv Reference Position , values: Genotype Row Index\n",
    "ref_pos_dict = dict( zip( list(SNV_annotation_DF.pos) , list(SNV_annotation_DF.index) ) )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create set of H37Rv Reference Positions that are SNP sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "H37Rv_ref_positions_of_interest = set(SNV_annotation_DF.pos)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Construct 'skeleton' Numpy Array that will store the supported base for all Reference Positions of interest for a single isolate in rollingDB\n",
    "\n",
    "- **0** for _A_\n",
    "- **1** for _C_\n",
    "- **2** for _G_\n",
    "- **3** for _T_\n",
    "- **9** for Bad Quality SNV calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#number of Reference Positions we're interested in (where High Quality SNV calls were made)\n",
    "number_of_SNVs = len(ref_pos_dict.keys())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "899035"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "number_of_SNVs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### In constructing the Genotype Vector, we're going to iterate through all Reference Positions of interest and find the supported base from the VCF file for each position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create an empty numpy array to store genotypes\n",
    "genotypes_array = np.full(number_of_SNVs , 9 , dtype = int)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([9, 9, 9, ..., 9, 9, 9])"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035,)"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_array)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Iterate through the subset VCF file corresponding to the isolate in rollingDB and fill in genotypes vector"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "#genomic data directory\n",
    "VCF_genotypes_dir = '/n/data1/hms/dbmi/farhat/rollingDB/genomic_data/'\n",
    "\n",
    "#directory that stores files for each sequenced isolate\n",
    "directory_for_sequenced_isolate = VCF_genotypes_dir + isolate_ID + '/pilon/'\n",
    "\n",
    "#position-base dictionaries (order of output in Pilon)\n",
    "base_order = {0:'A' , 1:'C' , 2:'G', 3:'T'}\n",
    "base_order_r = {'A':0 , 'C':1 , 'G':2 , 'T':3}\n",
    "\n",
    "#load in VCF file if one was found\n",
    "vcf_reader = vcf.Reader( open(directory_for_sequenced_isolate  + isolate_ID + '_full.vcf.gz' , 'r') )\n",
    "\n",
    "#iterate through each Variant Call \n",
    "for record in vcf_reader:\n",
    "    \n",
    "    #check to see if H37Rv Reference Position is one of the SNP sites of interest\n",
    "    if record.POS in H37Rv_ref_positions_of_interest:\n",
    "    \n",
    "        #check to see if variant is SNP (length of Reference Allele is 1 base and there's only 1 alternate allele and length of alternate allele is 1 base is alternate allele exists)\n",
    "        if (len(record.REF) == 1) and (len(record.ALT) == 1) and ( (record.ALT[0] == None) or (len(str(record.ALT[0])) == 1) ):\n",
    "\n",
    "            #check to see if the call is a PASS by Pilon\n",
    "            if record.FILTER == []:\n",
    "\n",
    "                ##### Retrieve Relevant information for filtering quality of Base Call #####\n",
    "                # Mean Base Quality @ locus\n",
    "                BQ = record.INFO['BQ']\n",
    "                # Mean Mapping Quality @ locus\n",
    "                MQ = record.INFO['MQ']\n",
    "                # Number of Reads w/ Deletion \n",
    "                DC = record.INFO['DC']\n",
    "                # Number of Reads w/ Insertion\n",
    "                IC = record.INFO['IC']\n",
    "                # Depth of Valid Reads in Pileup\n",
    "                VD = record.INFO['DP']\n",
    "\n",
    "                ### Filtering Criteria\n",
    "                #---> Mean Base Quality > 20\n",
    "                #---> Mean Mapping Quality > 30\n",
    "                #---> No Reads Supporting Insertions\n",
    "                #---> No Reads Supporting Deletions\n",
    "                #---> Number of High Quality Reads >= 20\n",
    "                if (BQ > 20) and (MQ > 30) and (DC == 0) and (IC == 0) and (VD >= 20): \n",
    "\n",
    "                    # Quality-Percentage - Percentage of As, Cs, Gs, Ts weighted by Q & MQ at locus\n",
    "                    QP = record.INFO['QP']\n",
    "\n",
    "                    if (sum(np.array(QP) >= 75) == 1): #1 Allele is supported by at least 75% of the reads, store information for this Base Call\n",
    "\n",
    "                        #create a Boolean list for the Bases that have at least 75% of support from reads\n",
    "                        supported_bases_boolean = list( np.array( QP ) >= 75 )\n",
    "\n",
    "                        #position of Base [A,C,G,T] with at least 75% of support from reads\n",
    "                        supported_bases_i = [i for i,x in enumerate(supported_bases_boolean) if x == True] #returns list of length \n",
    "                        supported_base = base_order[supported_bases_i[0]]\n",
    "\n",
    "                        #write an exception in case a SNV (with a reference position that we're not interested in) 'sneaks' past the subsetting and filtering\n",
    "                        try:\n",
    "                            #store a '0', '1', '2', or '3' in Genotype Vector (Reference Position Index) referencing the Base with at least 90% support from the reads\n",
    "                            genotypes_array[ref_pos_dict[record.POS]] = base_order_r[supported_base]\n",
    "                        except KeyError:\n",
    "                            pass"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Pickle Genotypes Array for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0, 2, 0, ..., 0, 1, 2])"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.save('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + isolate_ID + '/' + isolate_ID + '_genotypes' , genotypes_array , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [7] This section of the notebook launches the script that was created in the previous section of the notebook and submits jobs for batches of isolates to make genotypes arrays for each isolate."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from slurmpy import Slurm\n",
    "import numpy as np\n",
    "import itertools\n",
    "import shutil"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Every isolate in the pool of **33,873** isolates in rollingDB had a reduced VCF file and a corresponding Full VCF file. We will iterate through each of these Full VCF files and extract the base calls for only the **899,035** reference positions that we're interested in."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_id_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_id_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Create directory for each isolate to store output for genotyping (only run this once)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "for tag in isolate_id_series.values:\n",
    "\n",
    "    #create directory to store subset VCF and genotypes for each isolate\n",
    "    isolate_genotype_dir = '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "    if os.path.exists(isolate_genotype_dir):\n",
    "        shutil.rmtree(isolate_genotype_dir)\n",
    "        os.makedirs(isolate_genotype_dir)\n",
    "    elif not os.path.exists(isolate_genotype_dir):\n",
    "        os.makedirs(isolate_genotype_dir)\n",
    "'''"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Check Progress of Jobs:__ Isolates that need to have Full VCF >> Genotypes Array Reference Positions of Interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_ids_to_run_VCF_extraction = []\n",
    "isolate_ids_completed = []\n",
    "\n",
    "for tag in isolate_id_series.values:\n",
    "\n",
    "    #create directory to store subset VCF and genotypes for each isolate\n",
    "    isolate_genotype_dir = '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "    \n",
    "    #genotypes array file\n",
    "    genotypes_array_path = isolate_genotype_dir + '/' + tag + '_genotypes.npy'\n",
    "    \n",
    "    #check for the existence of a genotypes file, if TRUE job has either started or completed\n",
    "    if os.path.exists(genotypes_array_path):\n",
    "            \n",
    "        isolate_ids_completed.append(tag)\n",
    "            \n",
    "    #genotypes file doesn't exist, task never started\n",
    "    else:\n",
    "        \n",
    "        isolate_ids_to_run_VCF_extraction.append(tag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32567"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_completed)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1306"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_to_run_VCF_extraction)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Construct genotypes array for reference positions of interest for isolates that had Full VCF files "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.Series(isolate_ids_to_run_VCF_extraction)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Extract genotypes for each isolate from VCF file containing a subset of Reference Positions from the Full VCF files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Create BASH scripts to extract genotypes of N isolates per job"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "N_isolates_per_job = 115\n",
    "\n",
    "njobs = int( np.ceil( float( len( isolate_id_series ) ) / float(N_isolates_per_job) ) ) #number of jobs required if we split for every N isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "44\n"
     ]
    }
   ],
   "source": [
    "print njobs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "bash_scripts = []\n",
    "\n",
    "#split jobs up into tasks of N\n",
    "for first_isolate_i_for_job in np.arange(0 , (njobs)*N_isolates_per_job , N_isolates_per_job):\n",
    "\n",
    "    last_isolate_i_for_job = first_isolate_i_for_job + N_isolates_per_job\n",
    "    \n",
    "    commands_list = []\n",
    "\n",
    "    for isolate_i in range(first_isolate_i_for_job , last_isolate_i_for_job): \n",
    "\n",
    "        try:\n",
    "            tag = isolate_id_series[isolate_i]\n",
    "\n",
    "            commands_list.append( 'python /home/rv76/Farhat_Lab/Python_Scripts/homoplasy_project/rollingDB_extract_HQ_SNV_genotypes.py {}'.format(tag) )\n",
    "            \n",
    "        except KeyError: #applies only to the last job if it has < N isolates\n",
    "            continue\n",
    "        \n",
    "    bash_scripts.append(commands_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Submit each job to O2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893609\n",
      "submitted: Submitted batch job 6893611\n",
      "submitted: Submitted batch job 6893613\n",
      "submitted: Submitted batch job 6893615\n",
      "submitted: Submitted batch job 6893617\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno1 : 6893609\n",
      "Geno2 : 6893611\n",
      "Geno3 : 6893613\n",
      "Geno4 : 6893615\n",
      "Geno5 : 6893617\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893619\n",
      "submitted: Submitted batch job 6893621\n",
      "submitted: Submitted batch job 6893623\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno6 : 6893619\n",
      "Geno7 : 6893621\n",
      "Geno8 : 6893623\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893624\n",
      "submitted: Submitted batch job 6893625\n",
      "submitted: Submitted batch job 6893626\n",
      "submitted: Submitted batch job 6893627\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno9 : 6893624\n",
      "Geno10 : 6893625\n",
      "Geno11 : 6893626\n",
      "Geno12 : 6893627\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893628\n",
      "submitted: Submitted batch job 6893629\n",
      "submitted: Submitted batch job 6893630\n",
      "submitted: Submitted batch job 6893631\n",
      "submitted: Submitted batch job 6893632\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno13 : 6893628\n",
      "Geno14 : 6893629\n",
      "Geno15 : 6893630\n",
      "Geno16 : 6893631\n",
      "Geno17 : 6893632\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893633\n",
      "submitted: Submitted batch job 6893634\n",
      "submitted: Submitted batch job 6893635\n",
      "submitted: Submitted batch job 6893636\n",
      "submitted: Submitted batch job 6893637\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno18 : 6893633\n",
      "Geno19 : 6893634\n",
      "Geno20 : 6893635\n",
      "Geno21 : 6893636\n",
      "Geno22 : 6893637\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893638\n",
      "submitted: Submitted batch job 6893639\n",
      "submitted: Submitted batch job 6893640\n",
      "submitted: Submitted batch job 6893641\n",
      "submitted: Submitted batch job 6893642\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno23 : 6893638\n",
      "Geno24 : 6893639\n",
      "Geno25 : 6893640\n",
      "Geno26 : 6893641\n",
      "Geno27 : 6893642\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893643\n",
      "submitted: Submitted batch job 6893644\n",
      "submitted: Submitted batch job 6893645\n",
      "submitted: Submitted batch job 6893646\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno28 : 6893643\n",
      "Geno29 : 6893644\n",
      "Geno30 : 6893645\n",
      "Geno31 : 6893646\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893647\n",
      "submitted: Submitted batch job 6893648\n",
      "submitted: Submitted batch job 6893649\n",
      "submitted: Submitted batch job 6893651\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno32 : 6893647\n",
      "Geno33 : 6893648\n",
      "Geno34 : 6893649\n",
      "Geno35 : 6893651\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893653\n",
      "submitted: Submitted batch job 6893654\n",
      "submitted: Submitted batch job 6893656\n",
      "submitted: Submitted batch job 6893657\n",
      "submitted: Submitted batch job 6893658\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Geno36 : 6893653\n",
      "Geno37 : 6893654\n",
      "Geno38 : 6893656\n",
      "Geno39 : 6893657\n",
      "Geno40 : 6893658\n",
      "Geno41 : 6893660\n",
      "Geno42 : 6893661\n",
      "Geno43 : 6893663\n",
      "Geno44 : 6893665\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "submitted: Submitted batch job 6893660\n",
      "submitted: Submitted batch job 6893661\n",
      "submitted: Submitted batch job 6893663\n",
      "submitted: Submitted batch job 6893665\n"
     ]
    }
   ],
   "source": [
    "job_num = 1\n",
    "for job_i_commands_list in bash_scripts:\n",
    "    \n",
    "    #append all commands in a single string to be submitted as a job\n",
    "    genotypes_extract_job = ''\n",
    "    for command_i in job_i_commands_list:\n",
    "        genotypes_extract_job  = genotypes_extract_job  + '\\n' + command_i\n",
    "    \n",
    "    #directory where you want output + error files\n",
    "    os.chdir('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/O2_SLURM_logs/make_genotypes_arrays_from_VCF/')\n",
    "\n",
    "    job_name = 'Geno' + str(job_num)\n",
    "\n",
    "    s = Slurm(job_name , {'partition':'short' , 'N':'1' , 't':'0-12:00:00' , 'mem':'2G' , 'mail-type':'FAIL' , 'mail-user':'roger_vargas@g.harvard.edu'})\n",
    "\n",
    "    #submits the job\n",
    "    job_id = s.run(genotypes_extract_job)\n",
    "\n",
    "    print job_name  + ' : ' +  str(job_id)\n",
    "    job_num += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [8] Inspect the distribution of missing values after Genotype Matrix is constructed (from genotype vectors) and filter out SNPs & Isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib as mpl\n",
    "import matplotlib.ticker as ticker\n",
    "from itertools import compress\n",
    "from matplotlib.colors import LogNorm\n",
    "from matplotlib import gridspec\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Make sure all isolates have a corresponding _genotypes_ array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_id_series = pd.read_pickle( '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33873,)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_id_series)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "__Check for Genotypes Arrays:__ Get isolate IDs that have a corresponding genotypes array"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_ids_with_genotypes = []\n",
    "isolate_ids_without_genotypes = []\n",
    "\n",
    "for tag in isolate_id_series.values:\n",
    "\n",
    "    #create directory to store subset VCF and genotypes for each isolate\n",
    "    isolate_genotype_dir = '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/' + tag\n",
    "    \n",
    "    #genotypes_array\n",
    "    genotypes_array_pickle_file = isolate_genotype_dir + '/' + tag + '_genotypes.npy'\n",
    "    \n",
    "    #check for the existence of a genotypes file for this isolate\n",
    "    if os.path.exists(genotypes_array_pickle_file):\n",
    "            \n",
    "        isolate_ids_with_genotypes.append(tag)\n",
    "            \n",
    "    else:\n",
    "        \n",
    "        isolate_ids_without_genotypes.append(tag)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_with_genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolate_ids_without_genotypes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "del isolate_ids_with_genotypes #clear variable from memory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Concatenate genotypes arrays into a genotypes matrix for all isolates in rollingDB"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Store the __genotypes array__ for each isolate into a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.0295220382015\n",
      "0.059044076403\n",
      "0.0885661146046\n",
      "0.118088152806\n",
      "0.147610191008\n",
      "0.177132229209\n",
      "0.206654267411\n",
      "0.236176305612\n",
      "0.265698343814\n",
      "0.295220382015\n",
      "0.324742420217\n",
      "0.354264458418\n",
      "0.38378649662\n",
      "0.413308534821\n",
      "0.442830573023\n",
      "0.472352611224\n",
      "0.501874649426\n",
      "0.531396687627\n",
      "0.560918725829\n",
      "0.59044076403\n",
      "0.619962802232\n",
      "0.649484840433\n",
      "0.679006878635\n",
      "0.708528916836\n",
      "0.738050955038\n",
      "0.767572993239\n",
      "0.797095031441\n",
      "0.826617069642\n",
      "0.856139107844\n",
      "0.885661146046\n",
      "0.915183184247\n",
      "0.944705222449\n",
      "0.97422726065\n"
     ]
    }
   ],
   "source": [
    "genotypes_directory = '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/Genotypes_per_isolate/'\n",
    "genotype_arrays_per_isolate_list = []\n",
    "\n",
    "#iterate through the genotypes arrays for each isolate\n",
    "isolate_i = 0\n",
    "for isolate_ID in isolate_id_series:\n",
    "    \n",
    "    #load genotypes array\n",
    "    genotypes_array = np.load(genotypes_directory + isolate_ID + '/' + isolate_ID + '_genotypes.npy')\n",
    "    \n",
    "    #convert genotypes array from dtype 'int64' to 'int8' to save lots of memory\n",
    "    genotypes_array = genotypes_array.astype('int8')\n",
    "    \n",
    "    #store genotypes array for this isolate in a list with all other genotypes arrays\n",
    "    genotype_arrays_per_isolate_list.append(genotypes_array)\n",
    "    \n",
    "    if isolate_i % 1000 == 0:\n",
    "        print float(isolate_i) / float(len(isolate_id_series))\n",
    "    isolate_i += 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Stack genotype arrays for all isolates to create one __genotypes matrix__ for all isolates "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_matrix = np.stack(genotype_arrays_per_isolate_list , axis = -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035, 33873)"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([[0, 9, 0, ..., 0, 0, 0],\n",
       "       [2, 9, 2, ..., 2, 2, 2],\n",
       "       [0, 9, 0, ..., 0, 0, 0],\n",
       "       ...,\n",
       "       [0, 9, 9, ..., 0, 0, 0],\n",
       "       [1, 9, 9, ..., 1, 1, 1],\n",
       "       [2, 9, 9, ..., 2, 2, 2]], dtype=int8)"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genotypes_matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "30453012555 bytes\n"
     ]
    }
   ],
   "source": [
    "print(\"%d bytes\" % (genotypes_matrix.size * genotypes_matrix.itemsize))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Genotypes Matrix is ~30.45 GB large"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __Genotypes Matrix__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save Genotypes Matrix\n",
    "np.save('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix' , genotypes_matrix , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Load __SNP annotation file__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>10</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>11</td>\n",
       "      <td>A</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>16</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref  alt\n",
       "0    8   A  [C]\n",
       "1   10   G  [C]\n",
       "2   11   A  [C]\n",
       "3   15   C  [T]\n",
       "4   16   G  [C]"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(899035, 3)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "del genotype_arrays_per_isolate_list #clear variable from memory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [8.1] Drop SNP sites that *[1]* have low Empirical Base Pair Recall (EBPR) Scores, *[2]* are located in Mobile Genetic Element (MGE) regions, *[3]* have missing base calls $\\ge 10\\%$ of isolates or *[4]* are located in a region with overlapping genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 1)  *FILTER* out SNPs with low EBR scores"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load Genotypes Matrix\n",
    "## genotypes_matrix =  np.load('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix.npy')\n",
    "\n",
    "#load SNP annotation file\n",
    "## SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load list of low EBR score Reference positions to Drop (IL mapping/variant calling: sites with empirical score < 0.90)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/H37Rv_sites_with_low_EBR_to_drop.pkl', 'rb') as f:\n",
    "    H37Rv_positions_to_drop = pickle.load(f)\n",
    "    \n",
    "#convert to set, faster to check\n",
    "H37Rv_positions_to_drop = set(H37Rv_positions_to_drop)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Filter out sites with low EBR scores"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs from the __GENOTYPE MATRIX__ and __SNP ANNOTATION FILE__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#construct boolean filter\n",
    "SNPs_to_keep_filter = []\n",
    "for H37Rv_pos_i in SNP_annotation_DF.pos:\n",
    "    \n",
    "    if H37Rv_pos_i not in H37Rv_positions_to_drop:\n",
    "        \n",
    "        SNPs_to_keep_filter.append(True)\n",
    "        \n",
    "    elif H37Rv_pos_i in H37Rv_positions_to_drop:\n",
    "        \n",
    "        SNPs_to_keep_filter.append(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20360"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SNPs_to_keep_filter) - sum(SNPs_to_keep_filter) #number of sites dropped by this filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(20360, 3)"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#store dataframe of SNPs that were dropped by this filtering step\n",
    "SNPs_to_drop_filter = [not good_SNP for good_SNP in SNPs_to_keep_filter]\n",
    "dropped_SNP_annotation_DF = SNP_annotation_DF[SNPs_to_drop_filter]\n",
    "dropped_SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/SNP_sites_dropped/1_dropped_SNPs_annotation_low_EBR_score.pkl')\n",
    "np.shape(dropped_SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter Genotypes Matrix\n",
    "genotypes_matrix = genotypes_matrix[SNPs_to_keep_filter , :]\n",
    "\n",
    "#filter SNP annotation file\n",
    "SNP_annotation_DF = SNP_annotation_DF[SNPs_to_keep_filter]\n",
    "SNP_annotation_DF.reset_index(drop = True , inplace = True) #re-index new filtered SNP annotation DF (so new index matches indexing of genotypes matrix rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(878675, 33873)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>44</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   29   C     [G]\n",
       "1   31   A  [C, G]\n",
       "2   33   A     [G]\n",
       "3   44   C     [T]\n",
       "4   48   C     [T]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(878675, 3)"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF) #Annotation for Rows of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __Genotypes Matrix__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save Genotypes Matrix\n",
    "np.save('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix' , genotypes_matrix , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __SNP annotation file__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 2) *FILTER* out SNPs in MGE regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load Genotypes Matrix\n",
    "## genotypes_matrix =  np.load('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix.npy')\n",
    "\n",
    "#load SNP annotation file\n",
    "## SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(878675, 33873)"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(878675, 3)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Collect list of MGE Reference Positions to Drop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#csv files with category for each gene\n",
    "gene_categories_df = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/gene_categories/gene_categories.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>Gene_Category</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Rv0001</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Rv0002</td>\n",
       "      <td>Rv0002</td>\n",
       "      <td>Non-Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Rv0003</td>\n",
       "      <td>Rv0003</td>\n",
       "      <td>Non-Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Rv0004</td>\n",
       "      <td>Rv0004</td>\n",
       "      <td>Non-Essential</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Rv0005</td>\n",
       "      <td>Rv0005</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     name gene_id          Gene_Category\n",
       "0  Rv0001  Rv0001              Essential\n",
       "1  Rv0002  Rv0002          Non-Essential\n",
       "2  Rv0003  Rv0003          Non-Essential\n",
       "3  Rv0004  Rv0004          Non-Essential\n",
       "4  Rv0005  Rv0005  Antibiotic Resistance"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_categories_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#reference genome annotation\n",
    "reference_genome_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Collect positions for **MGEs**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "H37Rv_positions_MGE_genes = []\n",
    "MGE_sites_concat_length = 0\n",
    "\n",
    "for gene_id in gene_categories_df[gene_categories_df.Gene_Category == 'Mobile Genetic Element'].gene_id:\n",
    "    \n",
    "    #get reference positions tht corresond to each gene\n",
    "    gene_id_info = reference_genome_annotation.loc[gene_id, :]\n",
    "    gene_id = gene_id_info.name\n",
    "\n",
    "    chrom_start = gene_id_info.chromStart\n",
    "    chrom_end = gene_id_info.chromEnd\n",
    "\n",
    "    #find the position of the first base relative to H37Rv in 5' -> 3'\n",
    "    H37Rv_start = min(chrom_start , chrom_end)\n",
    "    H37Rv_end = max(chrom_start , chrom_end)\n",
    "    \n",
    "    # add up positions to get total lenth of MGE regions in genome\n",
    "    MGE_sites_concat_length += H37Rv_end - H37Rv_start\n",
    "    \n",
    "    #store all corresponding H37Rv Reference Positions to gene_id in list of reference positions to scan\n",
    "    H37Rv_positions_MGE_genes.append([H37Rv_start+1 , H37Rv_end+1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "108"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(H37Rv_positions_MGE_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "79659"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MGE_sites_concat_length"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER out sites in MGE regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "#construct boolean filter\n",
    "SNPs_to_keep_filter = []\n",
    "for H37Rv_pos_i in SNP_annotation_DF.pos:\n",
    "\n",
    "    bad_pos_i = False\n",
    "\n",
    "    for problem_region in H37Rv_positions_MGE_genes:\n",
    "\n",
    "        if (H37Rv_pos_i > problem_region[0]) and (H37Rv_pos_i < problem_region[1]):\n",
    "            bad_pos_i = True\n",
    "            break\n",
    "\n",
    "    if bad_pos_i == False:\n",
    "        SNPs_to_keep_filter.append(True)\n",
    "    else:\n",
    "        SNPs_to_keep_filter.append(False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "9137"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SNPs_to_keep_filter) - sum(SNPs_to_keep_filter) #number of sites dropped by this filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(9137, 3)"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#store dataframe of SNPs that were dropped by this filtering step\n",
    "SNPs_to_drop_filter = [not good_SNP for good_SNP in SNPs_to_keep_filter]\n",
    "dropped_SNP_annotation_DF = SNP_annotation_DF[SNPs_to_drop_filter]\n",
    "dropped_SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/SNP_sites_dropped/2_dropped_SNPs_annotation_MGE_region.pkl')\n",
    "np.shape(dropped_SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter Genotypes Matrix\n",
    "genotypes_matrix = genotypes_matrix[SNPs_to_keep_filter , :]\n",
    "\n",
    "#filter SNP annotation file\n",
    "SNP_annotation_DF = SNP_annotation_DF[SNPs_to_keep_filter]\n",
    "SNP_annotation_DF.reset_index(drop = True , inplace = True) #re-index new filtered SNP annotation DF (so new index matches indexing of genotypes matrix rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(869538, 33873)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>29</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>31</td>\n",
       "      <td>A</td>\n",
       "      <td>[C, G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>33</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>44</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   29   C     [G]\n",
       "1   31   A  [C, G]\n",
       "2   33   A     [G]\n",
       "3   44   C     [T]\n",
       "4   48   C     [T]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(869538, 3)"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF) #Annotation for Rows of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __Genotypes Matrix__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save Genotypes Matrix\n",
    "np.save('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix' , genotypes_matrix , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __SNP annotation file__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 3) *FILTER* out SNPs that are missing in $> 10\\%$ of isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load Genotypes Matrix\n",
    "## genotypes_matrix =  np.load('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix.npy')\n",
    "\n",
    "#load SNP annotation file\n",
    "## SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(869538, 33873)"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(869538, 3)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Investigate the Number of Missing Calls for each Unique SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=3\n",
    "plt.rcParams['axes.facecolor']='1.0'\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['axes.grid']=True\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '1.0'\n",
    "plt.rcParams.update({'font.size': 12})"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look for SNPs with an abundance of missing values, drop SNPs that have missing values $> 10\\%$ of the isolates"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_values = genotypes_matrix == 9\n",
    "missing_values_per_SNP = missing_values.sum(axis = 1) #will range from 0 - 33,873"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#impose a threshold for maximum number of missing values per SNP (across all isolates)\n",
    "max_proportion_missing = 0.10\n",
    "number_of_isolates = float(np.shape(genotypes_matrix)[1])\n",
    "\n",
    "max_number_values_missing_per_SNP = np.floor( max_proportion_missing * number_of_isolates )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3387.0\n"
     ]
    }
   ],
   "source": [
    "print max_number_values_missing_per_SNP #that we're going to allow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcoAAAFxCAYAAADzkehqAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XtYVNX+BvB3RBQYMcEMdQSFQUBARFS8ZOb9OCgZMqJpmh6yi5Uilh3Tzqkw7YIXTnRR69DBMk0B+6moWHkHFdRIQcREUQg1BXMERmXYvz94mOMEbGaQuQDv53l4ctba7P2dHfDO2pe1JYIgCCAiIqJatTJ3AURERJaMQUlERCSCQUlERCSCQUlERCSCQUlERCSCQUlERCSCQUlERCSCQUlERCSiSQRlamoqwsLCoFQqsWvXLnOXQ0RELUhrcxdQH7VajbVr1+Kbb75BmzZtzF0OERG1MBY/oszIyICNjQ1mzZqF8PBwXL9+3dwlERFRC2KyoIyLi4NCoYCrqysiIiJ0+kpKShAeHg53d3cEBgYiKSlJ23fjxg1cunQJX3/9NaZNm4aVK1eaqmQiIiLTBaWTkxPmz5+PKVOm1OhbsmQJrK2tkZmZidjYWCxevBjnzp0DALRv3x4DBgxAmzZtMHToUJw/f96g7Uokkkapn4iIWiaTBWVQUBDGjRsHBwcHnfaysjIkJyfjjTfegFQqRWBgIMaMGYOEhAQAgL+/P86fPw9BEJCVlYXu3bubqmQiIiLzX8yTl5cHKysryOVybZuPjw/S0tIAAI6OjlAoFAgNDUWrVq146JWIiEzK7EFZWloKe3t7nTZ7e3uUlpZqX8+aNQuzZs0ycWVEREQWcNWrVCqFSqXSaVOpVJBKpWaqiIiI6H/MHpRubm7QaDTIy8vTtmVnZ8PT09OMVREREVUxWVBWVFRArVajsrISGo0GarUaFRUVsLOzg0KhQHR0NMrKypCeno6UlBSEhoaaqjQiIqI6SQRBEEyxoZUrV2LVqlU6bZGRkVi4cCFKSkqwcOFCHDx4EA4ODnjrrbcQEhLSKNuVSCQw0VskIqJmyGRBaS4MSiIiehhmP0dJRERkyRiUREREIhiUREREIhiUREREIhiUREREIhiUelKr1Q3qIyKipo23hxhAJpPV2l5YWNgo6yciIsvDESUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBiUREZEIBmUjUKvVDeojIiLL19rcBTQHNjY2kMlktfYVFhaauBoiImpMHFESERGJYFASERGJYFASERGJYFASERGJYFASERGJYFASERGJYFASERGJsPigvHLlCnr37g2lUgmlUombN2+auyQiImpBmsSEA4MGDcL69evNXQYREbVAFj+iBICMjAyEhIRgxYoVEATB3OUQEVELYrKgjIuLg0KhgKurKyIiInT6SkpKEB4eDnd3dwQGBiIpKUnb99hjj+HIkSNITEzEzZs3kZycbKqSiYiITBeUTk5OmD9/PqZMmVKjb8mSJbC2tkZmZiZiY2OxePFinDt3DgDQtm1b2NnZQSKRQKFQIDs721QlExERmS4og4KCMG7cODg4OOi0l5WVITk5GW+88QakUikCAwMxZswYJCQkAABUKpV22ePHj6NHjx6mKpmIiMj85yjz8vJgZWUFuVyubfPx8dGOKNPS0jBu3DiEhISgqKgIISEh5iqViIhaILNf9VpaWgp7e3udNnt7e5SWlgIAxo4di7Fjx5qjNCIiIvOPKKVSqc7hVaDqcKtUKjVTRURERP9j9qB0c3ODRqNBXl6eti07Oxuenp5mrIqIiKiKyYKyoqICarUalZWV0Gg0UKvVqKiogJ2dHRQKBaKjo1FWVob09HSkpKQgNDTUVKURERHVyWRBGRMTA7lcjtjYWCQmJkIulyMmJgYAsHz5cqjVavj5+WHu3LlYsWIFR5RERGQRJEIzn+pGIpE02mw+Mpms1vbCwkLRPiIiarrMfo6SiIjIkjEoiYiIRDAoiYiIRDAoiYiIRDAoiYiIRDAoiYiIRDAoiYiIRDAoiYiIRDAoiYiIRDAoiYiIRDAoiYiIRDAojUytVjeoj4iILENrcxfQ3NnY2HDCdCKiJowjSiIiIhEMSiIiIhEMSiIiIhEMSiIiIhF6BWV6ejq2bt0KQRBw9OhRTJ06FS+++CIvRiEiomZPr6teo6KiUFpaCqVSiddeew1FRUWQSCS4d+8e4uLijF0jERGR2eg1orxw4QJ8fX1x+fJlFBUV4b333sOAAQOQnp5u7PqIiIjMSq+gVKvVaNu2LXJzcyGRSBAcHAwvLy+Ul5cbuz4iIiKz0uvQq4uLC7Zt24Z9+/bByckJnTp1wrVr1/Doo48auz4iIiKz0mtEOW/ePNy7dw/Xrl1DZGQk7t27h7S0NPTt29fY9REREZmVRBAEQZ8F1Wo1KisrYWdnZ+yaGpVEIoGeb7FeYlPRNbSPiIgsW72HXr/77jv8/PPPcHR0xEsvvQRXV1dT1EVERGQRRIMyLi4O//znP7Wvd+3ahSNHjsDe3t7ohREREVkC0XOUGzZsAACMGjUKPXr0QElJCXbs2GGSwoiIiCyB6IgyPz8fQUFBWLt2LYqKijBgwABcvnzZVLURERGZneiI8u7du+jSpQsAaP/LeyeJiKglqfdinr1796KgoKDW1xKJBOvXrzdedURERGZWb1Dm5+cjPz+/1tcSicR4lbUAarUaNjY2BvcREZHpiAZlZGSkqepokWxsbHiPJRGRhWNQEhERiRANyqNHj9a7gkGDBjVaMURERJZGNCiVSqXoeUiJRMLbRYiIqFkTDUp7e/tag1KlUjXa/KlUO17oQ0RkGUSD8uzZszqvT506hZUrV2L//v1o27Ytnn32WaMW96Bt27bh7bffxunTp022TXPihT5ERJZBr+dRnj59GtHR0fj555/Rpk0bzJo1C6+++io6d+5s7PoAAJWVldi5cye6du1qku0RERFVEw3KrKwsrFy5Env37oW1tTWeffZZvPbaayYPrKSkJIwfPx5r16416XaJiIhEp7AbN24c9u7dC4lEglGjRqFt27ZYt24d3nnnHbzzzjt499139d5QXFwcFAoFXF1dERERodNXUlKC8PBwuLu7IzAwEElJSdo+jUaD7du346mnnjLwrRERET080RFl9QU7giBg9+7dNfokEgn+9a9/6bUhJycnzJ8/H/v374dardbpW7JkCaytrZGZmYmsrCzMnDkT3t7e8PT0REJCAoKDg9GqlWimExERGYVoUE6ePLnRNhQUFAQAyMzMRFFRkba9rKwMycnJ+OmnnyCVShEYGIgxY8YgISEBb731Fs6fP48zZ84gMTERFy9exNtvv42oqKhGq4uIiEiMaFCuXr3a6AXk5eXBysoKcrlc2+bj44O0tDQAVaPNagqFgiFJREQmpddVrw86c+YMSktL4ePjg3bt2j10AaWlpbC3t9dps7e3R2lpaY1ld+3a9dDbIyIiMoToib9PPvkE06ZNw59//gkAeP3116FQKKBUKjF06FD8+uuvD12AVCqFSqXSaVOpVJBKpQ+9biIiooclGpQ//PADbt68iUceeQTnz5/Hpk2bIAgC2rdvjxs3buCjjz566ALc3Nyg0WiQl5enbcvOzoanp+dDr5uIiOhhiQZlUVERevXqBQDYt28fAGDUqFHIysrCkCFDDBpRVlRUQK1Wo7KyEhqNBmq1GhUVFbCzs4NCoUB0dDTKysqQnp6OlJQUhIaGPsTbIiIiahyiQalWq9GmTRsAwIkTJyCRSKBQKABUjQRrO49Yl5iYGMjlcsTGxiIxMRFyuRwxMTEAgOXLl0OtVsPPzw9z587FihUrOKIkIiKLIHoxj7OzM/bs2QMPDw/s378fADBkyBAAQEFBATp27Kj3hhYuXIiFCxfW2ufg4ID//Oc/eq+rpRPKyyGxtTW4j4iIDCc6opw1axZu3ryJd999F6WlpXjyySfh4uKC27dv49ixY/Dz8zNVnfQAia0tIJHU+sWQJCJqXKIjylmzZqFz585ITU3FY489htmzZwMALl++jJkzZ2LEiBEmKZKIiMhc6r2Pcty4cRg3bpxOm6+vL3x9fY1WFBERkaXgBKpEREQiGJREREQiGJREREQiGJREREQi9A7Ko0eP4pVXXsGECRNQVFSE1atX49SpU8asjYiIyOz0enrInj17MGfOHFRWVkIikaBTp06Ij49HdnY21q9fb+waiYiIzEavEeXq1avRoUMHDBo0CADQunVrDBgwACdPnjRqcUREROamV1BeuHABI0eOhI+Pj7bN0dERt27dMlphRERElkCvoHzssceQm5uLyspKAEBJSQkOHTqErl27GrU4agC1us4uobzchIUQETUPegXl+PHj8euvvyI+Ph4A0L9/f1y+fBnjx483anHUADY2nAeWiKgR6RWUCxYswFNPPQWNRgNBEHD37l1MmDAB8+fPN3Z9REREZqXXVa+2trb47LPPsGzZMly5cgXOzs5wdHQ0dm1ERERmp1dQHj16VOd1bm6u9t/VV8ISERE1R3oFpVKphEQiqdEukUhw+fLlRi+KiIjIUugVlDKZTBuUGo0Gf/zxByoqKiCTyYxaHBERkbnpFZTHjh3TeV1cXIyZM2di+PDhxqiJiIjIYjRoUnRHR0f06dMHmzdvbux6iIiILIpeI8rJkyfrvL5z5w6ysrLg4OBglKKIiIgshV5BmZaWVqNNIpHg5ZdfbvSCiIiILIleQblq1Sqd13Z2dvD29oabm5tRiiIjUaurZu6phVBezpl7iIhqoVdQhoWFGbsOMoXq6e1qIREEExdDRNQ06BWUzz//fL3LSCQSPpuSiIiaHb2Ccvfu3ZBIJBAeGHXU9pqIiKi50SsoZ8+ejfj4eIwdOxZubm7Iy8tDSkoKZs6cyStfySjEzpnyfCoRmZJeQZmfn4+goCB8/vnn2raXXnoJly9fRlRUlNGKo5ZLYmvL86lEZBH0CsrU1FR4eXnh7t27aNu2LdRqNQoKCnDu3Dlj10dERGRWegWll5cXMjMz4e/vjy5duqCoqAh37tyBv7+/sesjU+GtI0REtdJrCrtVq1bB2dkZKpUKubm5UKlUcHFxwcqVK41dH5lK9a0jtXwxJImoJdNrROnh4YFDhw7hxIkTuHr1Krp06YKAgABYWVkZuz4iIiKzqjMoCwsL0a5dOzzyyCMoLCwEUPW4repHa129elXbRs0cD8sSUQtWZ1AOGjQIzz//PP71r39h4MCBfHBzS8YZfYioBavzHKUgCDoTClS/fvCrsrLSJEVS8yOUl5u7BCIivdQ5oiwoKKj130SNQew+SQAAR6pEZCH0upjHnP744w+Eh4fD2toarVq1QmxsLJycnMxdFhERtRB63R6SnZ2N0NBQeHp6wtnZWfvl4uJi7Prg6OiIbdu2ISEhAUqlEt99953Rt0kGUKvr7OLhVSJqDvQaUc6bNw85OTk12gUTHB578BaU0tJSeHp6Gn2bZABe6ENEzZxeI8pLly7B09MT33//PXbv3o09e/Zov/QVFxcHhUIBV1dXRERE6PSVlJQgPDwc7u7uCAwMRFJSkk7/mTNnMGHCBMTFxaF37956b5PMTGS0SUTUVOg1ohw6dCisra0xZMiQBm/IyckJ8+fPx/79+6H+yx/QJUuWwNraGpmZmcjKysLMmTPh7e2tHT36+vpix44d+L//+z988skn+PDDDxtcB5mQyGiTF+sQUVOhV1B27twZ3377LWbNmgUfHx+0bv2/b1uwYIFeGwoKCgIAZGZmoqioSNteVlaG5ORk/PTTT5BKpQgMDMSYMWOQkJCAt956SzsROwC0b98etry5nTgBAhGZkF5B+c033wAAfvzxR/z0008Aqs5PSiQSvYOyLnl5ebCysoJcLte2+fj4IC0tDQBw+vRpLF++HK1atYKNjQ3nlyWeFyUik9IrKJVKZa0z8zSG0tJS2Nvb67TZ29ujtLQUANC/f38kJiYaZdtERET10Sso16xZY7QCpFIpVCqVTptKpYJUKjXaNomIiPQlGpRHjx6t0WZra4vu3bujQ4cOjVKAm5sbNBoN8vLy4ObmBqDqvs0WfxuIWq2djJ6IiMxHNCjFDrlOmDABH3/8Mdq1a6fXhioqKlBRUYHKykpoNBqo1Wq0bt0adnZ2UCgUiI6ORnR0NLKyspCSkoIffvjB8HfTnPCKUSIiiyB6H6W9vX2Nr7Zt20IQBOzYscOgC2tiYmIgl8sRGxuLxMREyOVyxMTEAACWL18OtVoNPz8/zJ07FytWrOCIkoiILIJEaMD0OmfOnMHkyZPh6OiII0eOGKOuRiORSBptBqG6nr1ZWFholD7REWVz7nuY7y0v560jRNSo9JqZ5698fX3h7e2Na9euNXY9RA+n+pB1LV8MSSJqiAYF5dmzZ5Gdnc2neFDTwgnciagBRC/m8fHxqdF279497RR0YWFhxqmKyBg4UQERNYBoUP7555919o0bNw6vv/56oxdEZGnEzm3yvCdR8ycalFu2bKnRZmNjg+7du8PR0dFoRRFZEomtbYNGoqYOWAY6kXGIBuXgwYNNVQdRs9PQgG0q2yNqKRp0MQ9Rs8NnZxJRHfSa65Wo2eNMSERUB44oiR4GR6JEzV6dQbls2TLs3bsXQNVFPZmZmSYriqjJEJnggIiahzqD8osvvsDhw4cBAAsWLMC2bdtMVhQREZGlqPMcZfv27ZGUlKS9l/LAgQOIjIzUWUYikRg0MToREVFTU2dQjh07Flu3bsXWrVshkUiQm5uL3NxcnWUYlERE1NzVGZSrVq3CqFGjcOHCBURHR6Nv374YMWKEKWsjIiIyuzqDslWrVggODgYAVFZWIiAgAMOHDzdVXURERBZBr/soIyMjcfnyZXz44YcoKCiAs7MzpkyZgu7duxu7PiIiIrPSKyhPnTqFKVOmoLy8XPsQ5C+//BKbN29G3759jVogERGROek14cDy5ctRVlaGv/3tb3jllVcwbtw4lJWVYcWKFcauj4gaA5/FSdRgeo0os7OzERQUhHXr1mnbXnjhBRw5csRohRE1a2p11WQFtTDKkz74LE6iBtMrKNu2bYvi4mKdtuLiYrRt29YoRRE1ewwuoiZDr6Ds378/du3ahccffxxyuRx5eXnIz89HUFCQsesjIiIyK72C8p///CdOnz6N/Px85OfnAwCcnZ2xdOlSoxZHRERkbnoFZbdu3fDzzz/jxx9/REFBAbp164bRo0fDlk9MJyKiZk7v51Ha2tpqJyAgIiJqKfg8SiIiIhEMSiIiIhEMSiIiIhH1BmVFRQXCwsLw0UcfmaIeImomxGb84WxA1JTUG5StW7fGhQsX8Mcff5iiHiIS0ZQCRmJrWzWpQi1fjT7zEJER6XXoNSIiArt370ZaWhru379v7JqIWjaReVnFwoeIjEOv20MWL14MiUSCsLAwnXaJRILLly8bpTCiFktkejtwejsik9P7Pkqhll/Q2tqIiIiaE72C8ujRo8aug4iIyCLpPYUdAFy4cAGXLl3CqFGjjFoUERGRpdDrYp6SkhJMmTIFw4cPx+zZs3HlyhW4uLjwlhEiImr29ArKZcuW4ciRI2jTpg0EQYCzszMCAgLw448/Grs+AEBGRgaCg4MRGhqKuXPn8spbIiIyGb2Ccv/+/Rg8eDCeffZZbZuHh4fJrnjt1q0bvv/+eyQkJMDFxQV79uwxyXaJiIj0OkepVqvh5OSk01ZcXAxra2ujFPVXnTt31v67devWaNWKM+8REZFp6JU4vXr1wo8//ohTp04BAKKiorB37154e3sbtLG4uDgoFAq4uroiIiJCp6+kpATh4eFwd3dHYGAgkpKSanz/5cuXsW/fPowePdqg7RIRETWUXkG5aNEi3Lt3DydPnoQgCFi7di1atWqFhQsXGrQxJycnzJ8/H1OmTKnRt2TJElhbWyMzMxOxsbFYvHgxzp07p+1XqVSIiIhATEwM2rRpY9B2iYiIGkqvQ6+BgYFITk5GfHw8CgoK4OzsjOnTpxs8ogwKCgIAZGZmoqioSNteVlaG5ORk/PTTT5BKpQgMDMSYMWOQkJCAt956CxUVFZg7dy4iIyPh7u5u0DaJiIgeht4z83h5eeHtt99GUVERunTpAttGnNQ4Ly8PVlZWkMvl2jYfHx+kpaUBALZt24ZTp05hzZo1WLNmDWbMmIGJEyc22vaJiIjqoldQ3rp1C4sWLcKuXbu0bQqFAh9++CEcHBweuojS0lLY29vrtNnb26O0tBQAoFQqoVQqH3o7REREhtIrKF9//XXs3r1bp23Xrl2orKzEl19++dBFSKVSqFQqnTaVSgWpVPrQ6yaieqjVVROx10EoL+djsahF0+tinkOHDsHFxQUpKSnIy8vDnj170K1bNxw6dKhRinBzc4NGo0FeXp62LTs7G56eno2yfiISUf20Ej47kqhWegVl9+7dMWTIEHh7e6NNmzbw8fHBkCFD0KNHD4M2VlFRAbVajcrKSmg0GqjValRUVMDOzg4KhQLR0dEoKytDeno6UlJSEBoa2pD3RERE1GjqPPT64BNDlEolVq9eDR8fH8jlcvz2229ITk5GZGSkQRuLiYnBqlWrtK8TExMRGRmJhQsXYvny5Vi4cCH8/Pzg4OCAFStWcERJRERmJxHqeKhkt27dIBF5arogCGjVqpXFP7hZIpE02nMzZTJZre2FhYVG6RN9eG9z7rO0elpyX3V/Q/EB1NQM1DmilMlkokFJRNRgIhcQ8eIhsjR1BuWxY8dMWQcRtSTVFxDVQsLRJlkYvSccAKomQi8vL9dpq+vQIRERUXOgV1AePHgQCxcuxNWrV3XaJRKJxZ+jJCIiehh6BeU//vEPnblZqzXWRTJERESWSq+gLCkpwfDhw7F27VrOlkNERC2KXhMOzJ49G4WFhSgqKuIokoiIWhS9RpRBQUH4+uuvMWLECJ12nqMkIqLmTq+gfO2113D79u0a7RxdEhFRc6dXUBYWFsLf3x9LlixB+/btjV0TERGRxdArKKdPn45ffvkF/fr1Q5s2bYxdExFZElPPosNZe8jC6BWUhw8fxrlz5+Dn54du3brBysoKQNU5yr8+p5KImhlTz6LDWXvIwugVlGfPngUA3LlzBzk5Odp2zgVLRETNnV5B+eCjsYiIiFoSvYIyLCzM2HUQEdWP5y/JDPQKyroe0CyRSLBy5cpGLYiIqE48f0lmoFdQfv/99zoPQK7+N4OSiIiaO4NHlBqNBjk5OUhJScHUqVONVhgREZElaPCh10WLFtX6RBEiIqLmRO+ZeR6kUqlw4cIFZGdnG6UoIiIiS6FXUA4aNKjWdi8vr0YthoiaGJGrUImaC72C8q+Tn9va2sLX1xdRUVFGKYqImgiRq1DBq1CpmdArKAsKCoxdBxERkUXS68HNRERELZXoiNLZ2Vn0m/ngZiIiau5Eg5IPZiYiopZONCj37Nmj87qsrAxxcXHYvn07BEGAj4+PUYsjImoMYvPAco5Yqo9oUFYHoVqtxtdff40vvvgCN27cgJeXFyIjIxEUFGSSIomIHobE1pZzxFKDiQbl3bt3ER8fj88++ww3btxAz549ERUVheDgYFPVR0REZFaiQTlkyBBcv34dEokEwcHBeOqpp9CqVSukpKRolxk7dqzRiyQiIjIX0aC8du2a9kkh27dvx/bt23X6edUrERE1d6JBKZPJIKlr1g0iIqIWQDQojx07Zqo6iIiILBJn5iEiIhLBoCQiIhLBoCSi5kGtNncF1ExZfFDevn0b48ePR8+ePZGTk2PucojIUlU/8qu2L6KHYPFBaWtri/j4eIwfP97cpRARUQtk8UFpbW2Njh07mrsMIiIyMqG8vEF9xmayoIyLi4NCoYCrqysiIiJ0+kpKShAeHg53d3cEBgYiKSnJVGUREZGF0M7JW8uXOSeuF72PsjE5OTlh/vz52L9/P9R/Oem+ZMkSWFtbIzMzE1lZWZg5cya8vb3h6elpqvKIiIhqZbIRZVBQEMaNGwcHBwed9rKyMiQnJ+ONN96AVCpFYGAgxowZg4SEBFOVRkREVCezn6PMy8uDlZUV5HK5ts3Hxwfnzp3Tvp4xYwYOHjyIN954A5s3bzZHmURE1EKZ7NBrXUpLS2Fvb6/TZm9vj9LSUu3rDRs2mLosIiIiABYwopRKpVCpVDptKpUKUqnUTBURERH9j9mD0s3NDRqNBnl5edq27OxsXshDRGZnqbcrkGmZLCgrKiqgVqtRWVkJjUYDtVqNiooK2NnZQaFQIDo6GmVlZUhPT0dKSgpCQ0NNVRoRUa0s9XYFMi2TBWVMTAzkcjliY2ORmJgIuVyOmJgYAMDy5cuhVqvh5+eHuXPnYsWKFRxREpFpcI5YqodEEATB3EUYk0QiQWO9RZlMVmt7YWGhUfrqnKNSEJp3n6XV05L7LK0eS+ujxmeB+9vs5yiJiIgsGYOSiIhIBIOSiIhIBIOSiIhIBIOSiIhMpinef2r2KeyIiKjl0N6bWhsLvZKYI0oiIiIRDEoiIiIRDEoiIiIRDEoiIiIRDEoiIiIRDEoiIiIRDEoiIiIRDEoiooYQeTxXU7ypnurGCQeIiBrCxqbOG+clFnrjPDUMR5REREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQiGJREREQimkRQfvDBBwgJCcGcOXNQXl5u7nKIiKgFsfigzMnJQX5+PpKSkvDEE09g06ZN5i6JiIhaEIsPymPHjmHEiBEAgJEjR+L48eNmroiIiFoSkwVlXFwcFAoFXF1dERERodNXUlKC8PBwuLu7IzAwEElJSdq+P//8E+3btwcA2Nvb49atW6YqmYiICK1NtSEnJyfMnz8f+/fgJuM+AAAXkUlEQVTvh1qt1ulbsmQJrK2tkZmZiaysLMycORPe3t7w9PTEI488gtu3bwMAVCoVOnToYKqSiYiITDeiDAoKwrhx4+Dg4KDTXlZWhuTkZLzxxhuQSqUIDAzEmDFjkJCQAAAIDAzEgQMHAAD79+/HgAEDTFUyERGR+c9R5uXlwcrKCnK5XNvm4+ODc+fOAQB69eqFbt26ISQkBAcOHMDUqVPNVSoRUbMjiNxJYIy+pshkh17rUlpaCnt7e502e3t7lJaWal8vXrzY1GUREbUIEltbQCKpvU8QGr0PgtCgOs3J7CNKqVQKlUql06ZSqSCVSs1UERER0f+YPSjd3Nyg0WiQl5enbcvOzoanp6cZqyIiIqpisqCsqKiAWq1GZWUlNBoN1Go1KioqYGdnB4VCgejoaJSVlSE9PR0pKSkIDQ01VWlERER1MllQxsTEQC6XIzY2FomJiZDL5YiJiQEALF++HGq1Gn5+fpg7dy5WrFjBESUREVkEiSA0wTOrBpBIJGistyiTyWptLywsNEqf6Mnw5txnafW05D5Lq6cp9TUllrTfLHCfmv0cJRERkSVjUBIREYlgUBIREYlgUBIREYlgUBIREYlgUBIREYlgUBIREYloEfdREhERVTM09sz+9BBja+afA4iIyMh46JWIiEgEg5KIiEgEg5KIiEgEg5KIiEgEg5KIiEgEg5KIiEgEg5KIiEgEg5KIiEgEg5KIiEgEgxJASUkJwsPD4e7ujsDAQCQlJdW6nCAIeP/99+Hj4wMfHx8sW7as2c/8o++++fzzzzFy5Eh4eHhg0KBB+Pzzz01cqenpu2+q3bt3D8OGDUO/fv1MVKH5GLJvTp8+jUmTJqFnz57o06cPvvzySxNWanr67pu7d+/izTffRJ8+feDj44PnnnsORUVFJq7WdOLi4qBQKODq6oqIiAjRZdetWwd/f394eXkhMjISd+/eNW5xAgkvv/yy8OKLLwp37twRjh07Jnh6ego5OTk1louPjxeGDh0qFBYWCr///rvw5JNPCv/973/NULHp6LtvPv30U+HXX38V7t+/L5w/f14YMGCAsG3bNjNUbDr67ptqa9asEUJCQoSAgAATVmke+u6bmzdvCr179xYSEhIEtVotqFQqITc31wwVm44hv1OjRo0Srl+/LpSXlwuvvvqqEB4eboaKTWPnzp3Crl27hDfffFOYP39+ncvt27dP8PPzE3JycoSSkhIhNDRUeP/9941aW4sPytLSUqF79+7Cb7/9pm179dVXa93xwcHBwoYNG7SvN27cKIwfP94kdZqDIfvmr5YuXSosWbLEmOWZlaH7Jj8/Xxg2bJjw008/NfugNGTfLF++XHj11VdNWZ5ZGbJv3nzzTSEqKkr7eu/evcLQoUNNUqc5ffDBB6JBOXfuXGH58uXa1wcPHhT69Olj1Jpa/KHXvLw8WFlZQS6Xa9t8fHxw7ty5Gsvm5ubC29tb+9rb2xu5ubkmqdMcDNk3DxIEAceOHYOHh4exSzQbQ/fN0qVL8Y9//AM2NjamKtFsDNk3J0+ehIODA5566in4+fnhueeeQ2FhoSnLNSlD9s0zzzyD9PR0XL16FeXl5UhKSsKIESNMWa5FOnfunM7fYR8fH/zxxx8oLi422jZbfFCWlpbC3t5ep83e3h6lpaW1Ltu+ffsaywnN9DylIfvmQStXrkRlZSWmTJlizPLMypB9s2vXLmg0GigUClOVZ1aG7JuioiJs2bIF7733Ho4fPw4XFxfMnTvXVKWanCH7xs3NDTKZDP369YOnpyfOnz+PBQsWmKpUi1VWVlbj7zCAev8uPYwWH5RSqRQqlUqnTaVSQSqV1rvsnTt3IJVKm+0zLw3ZN9Xi4uKwdetWxMfHo23btsYu0Wz03TdlZWVYtmwZoqKiTFmeWRnyc2NjY4Nx48bB398fNjY2WLBgATIyMnD79m1TlWtShuybxYsX4+7duzhz5gzOnz8PhUKBGTNmmKpUi2VnZ6ezD6v/LfZ36WG1+KB0c3ODRqNBXl6eti07Oxuenp41lvXw8EB2drbOcs358KIh+wYANm3ahNjYWGzevBldu3Y1VZlmoe++ycvLQ0FBASZNmgR/f3/MmTMH169fh7+/P65cuWLqsk3CkJ+bXr166XzQrP53cz1KY8i+yc7ORlhYGBwcHNC2bVv8/e9/x6lTp4x6iLEp8PT0rPF3uFOnTnB0dDTaNlt8UNrZ2UGhUCA6OhplZWVIT09HSkoKQkNDayyrVCqxbt06FBUV4erVq1i7di3CwsLMULVpGLJvEhMT8cEHH2DTpk3o3r27Gao1LX33jZeXl7YvJSUFH3/8MTp16oSUlJRm+2HCkJ+bKVOmYPfu3Thz5gzu37+PNWvWIDAwEI888ogZKjc+Q/ZNnz59sHXrVty+fRv379/Hf//7X3Tu3NmogWBOFRUVUKvVqKyshEajgVqtRkVFRY3llEolNm3ahNzcXNy6dQsxMTHG/zts1EuFmoji4mJh9uzZglwuF/r37y8kJiYKgiAIR48eFdzd3bXLVVZWClFRUYK3t7fg7e0tREVFCZWVleYq2yT03TcDBw4UXFxcBHd3d+3XokWLzFW2Sei7bx505MiRZn/VqyAYtm++/vprISAgQOjVq5cwc+ZMoaCgwBwlm4y+++bmzZvCK6+8IvTu3Vvw8vISJk6cKJw8edJcZRtddHS00LVrV52v6OhooaCgQHB3d9f5ufjiiy8EPz8/wcPDQ4iIiBDUarVRa5MIQjM9xkFERNQIWvyhVyIiIjEMSiIiIhEMSiIiIhEMSiIiIhEMSiIiIhEMSiIiIhEMSjKbzZs3QyaTQalUmrWOwsJCKJVKyOVyyGQy7N271+jbHDhwIGQyGVJTU42+LWNYuXIlZDJZvc8NVCqVkMlk2Lx5s9FqkclkkMlkDz3TUUREBGQyGVauXNlIlRmPKfYr/Q+Dsomp/gNb/eXr64tp06YhMzPT3KWJqi0YPDw8EB4ejvHjx5uxMiA2NhZpaWno0aMHwsPDa51ZqDoY/v73v5uhQv2DyVQCAgIQHh6OJ598EgCQmpoKmUyGgQMHPvS6H/wZT09P17YfO3ZM2/7gdsLDwxEeHo527do91HaffPJJhIeHIyAg4KHWUxdBEBAYGFhrwG3fvh0ymQx9+vSpdTYaMq/W5i6AGmb06NFwcXFBWloaDhw4gMzMTBw4cACPPvpojWXv378Pa2trM1Qpvu2+ffuib9++Jq6opup5N+fMmYOpU6eauZqmYcSIESZ55FN8fDwGDBgAANiwYUOty7z33nuNsq2QkBCEhIQ0yrpqI5FIEBISgtjYWPzwww86T9fZtm0bAGDixIlo3Zp/li0NR5RN1DPPPIOoqCh8//33AIBbt27hxIkTuHLlivZTd3x8PAICAvDMM88AqJo8ePr06fD19UXv3r3x3HPP4bffftOus/qT/KeffopRo0ZpR3wPTsJ89OhRTJo0Cb169UJAQABeffVVXL16Vdtfve3169dj0KBBGDZsGAYOHIiCggIAwOTJk7WfqGs79Krv+uPi4jB06FB4eHjgtddew7179+rcV2LrVCqVOHz4MABg4cKFeh/Cu3XrFl544QX4+vrCzc0NgwYNwqJFi7T99e3rv/riiy/w+OOPw93dHa6urhg9ejR27NgBoGo0uWrVKgDAli1bdPZZYWEhXn75ZfTr1w+9evXCM888g5ycHO16169fjyFDhsDNzQ29e/eGUqmstY6bN29CJpPBy8sLlZWVKCkpQbdu3eDv76/djkwmg5+fHwRB0BnhpqamYvLkyQCAgoIC7f+jB125cgWTJ0+GXC7HxIkTtT8PYjp06ICdO3eiuLgYN2/eRHJyMjp06FBjub8eehV7z0lJSRg+fDjkcjl8fHwQHByM48ePA6h56LX6Pc6ZMwfz5s1Dz5498fjjj+PgwYPabefm5mLChAmQy+WYMWMGli5dKnrkoXpO18OHD+PGjRsAgNu3b2Pfvn06/e+++y4GDhwINzc3yOVyTJgwQfQw/V8PxdY2ws/JycGMGTPg5+eH3r17Y86cOdpnfwqCgBUrVqB///5wdXWFv78/pk2b1uInYK/GoGzCKisrkZaWpn3918mSP/zwQ4wYMQL9+/fHtWvXoFQqsX//fgQEBMDX1xc//vgjJk+ejFu3bul8X0xMDPr06QNHR0fs3r0bb775JoCqP/5Tp07F8ePHMXz4cMhkMiQlJWH69Om4f/++zjo++OADDBw4EMOGDcPUqVO1h8WCgoIQHh5e61NXDFl/dHQ0+vfvD41Gg8TERCQkJNS6j+pb5/jx49G5c2cAwLBhw/Q+hLd27Vrs3LkTrq6uCAsLQ8+ePXHixAkAMGhfV7t8+TJ69eqFsLAwjB07Frm5uZg3bx6uXLmCgIAA7ci7Z8+e2sPV5eXlCAsLw/bt29GrVy888cQTSEtLw+TJk1FcXIyLFy/inXfegUqlwuTJkzFs2DAUFhbi+vXrNbbfsWNHeHh4QKVSIScnBxkZGRAEAX/88QcuXryIjIwMAFUfpv76WLkuXbogKCgIANCuXTvtodAHffLJJ3BycoKjoyMyMjLw4Ycf1ruPlUol7t69i02bNmHTpk24e/euNpDrIvaey8vLsWDBAhQUFCAkJAQjR47EnTt3cOnSJdF1Jicn49q1a/D09MSlS5ewcOFCAFWTeM+ePRunTp2Ch4cHbGxsEB8fL7ouDw8P+Pj4QKPRaD8I7dy5E3fv3oVcLkefPn0AVH2w6Nu3L6ZOnYohQ4bg1KlTePHFF3Hnzp1691ttrl+/jtDQUBw6dAiBgYHo27cvkpOTMW3aNNy9exeHDh1CbGwsrKysMHXqVAwcOBA5OTlGfcZjU8IxfhP11z9EY8aMQb9+/XSeDr927VoMHToUAPDZZ5/hzz//xODBg7W/zGPHjkVWVhZ27NiBZ599Vvt9ixYtwvPPP4+srCyMHTsWycnJKC0txYYNG3D//n2EhYVh9erVuH//Pvr374+cnBykpqZqz1cBwPvvv69zGHPTpk24c+cOZs+ejSFDhgCo+jT+IEPW/8EHHyA4OBiCIGDr1q04c+ZMrfupvnXOnj0bO3fuxNWrV/H000/r/bDp6vNIffv2RUhIiPYPJQAkJCTova+rLV26FDt37sTFixdhbW2Njh074vr168jIyEBISAhOnjyJU6dOwd/fX3uocfv27bh06RI6d+4MNzc3AFWjq0uXLmHHjh0YPHgwAMDJyQlBQUHo2bMnunbtCo1GU+t7GjRoEHJzc5GRkYHCwkI4Ozvj2rVrOH78OE6fPg0A2nU+yNXVFbNnz9aO+Go7FDp9+nQsX74cmzdvRmRkJLKysurdx4MHD8bBgwfx7bffQhAEeHh4YNCgQVi/fn2d31P9/6W296xWq6HRaNCxY0f87W9/g4eHB7p3717n/qjm6emJTZs24cqVKxg8eDB+//13FBcX47fffsOlS5cglUqRkJAAOzs7zJ49GykpKaLrmzRpErKysrBt2zbMmjVLe9j1wSeIfPzxx9i5cyeuXLkCV1dX2Nraori4GGfPntUeijZEQkICbt26pd0fQNWHo99++w2pqanax5r16NEDwcHB8PDwQMeOHZvt484MxaBsokaPHo0ePXrAwcEBfn5+GDFiRI1P+g/+QlUflurZs6e2zd3dHVlZWTUOg1Uv4+7urm0rKirSLlfdb21tDRcXF9y4caPGOvr372/wezJk/b6+vgCgfRxTXZ98DVmnIZ5//nlkZ2cjPj4eX331FaysrPDUU0/h3//+t0H7GgDu3buH4OBgnUOm1W7evFlnDdXrunr1Kr766iudvkuXLmHmzJl4/fXX8dVXX2HatGkAALlcjnXr1sHLy6vG+qqDPSMjAwUFBRgyZAjOnz+PjIwM0aDUR/X/r+on0+s7UpkxYwbefvttAMCyZcvqXb5nz56i73nFihVYvXo1Zs2aBaBqNPzvf/9b++GtNj4+PpBIJNraq+svKioCAHTt2hV2dnYAqkaM9QVlSEgI3n//fWRkZODkyZNIS0uDRCLBpEmTAADFxcUYPXo0rl27VuN79T0U+tfwr/6ZPH/+PM6fP6/TV/2z8txzzyEhIUE7avf398d//vMfODk56bXN5oyHXpuoZ555Bu+++y4iIiIwcuTIGiEJAG3bttX+29nZGQB0zk9duHABANCtWzed76v+RXpw2S5dumiXq26/f/8+Ll++XOs6Htw2AFhZWQGoOlxcF0PWr+8FD4as0xAdOnTAt99+i3PnzmHv3r3w8PBAUlIS0tPTDdrXQNXIOicnB1ZWVjh8+DAKCgq0h6arP9G3alX1q/rg/qtel5+fHwoKClBYWIjCwkJkZ2dj3rx50Gg0mDdvHs6cOYPjx4/jlVdewYULF+ockVWH4LFjx5CZmYkBAwZgwIABOHToELKzs+Hg4FBrwAL1//+t7q/t51SMUqmEra0t7Ozsan1m41/V954nT56MEydO4OTJk3jvvfdQVFSENWvWiK6zrtq7dOkCoOpDZHl5OQCInoeu5uTkhMcffxyCIOC1116DRqNBYGCg9ufm+PHjuHbtGjp27IhffvkFFy9e1H4grGuEZ2trCwDaQ7Pnzp3T6a9ed1BQkPbnpLCwEKdOncLUqVOh0Wjw/vvv4+zZszhy5AiUSiV++eUXbNy4sd730xJwRNlCTJo0CZ988glSU1Mxa9Ys3L9/H2fOnEGnTp1q3J7x8ccfIzs7W3vxgEKhgFQqxfTp07Fx40Zs2bIFarUaBQUFuHHjBjw9PesdaXTt2hX5+fmIjo5GSkoKXnzxxRrLPMz662KMdQLAp59+ipSUFHh5eaFNmzba0Z29vb1B+xqoOrfcqlUraDQavPvuuygvL8fFixd1lqk+XLZv3z4sXboUgwcPxqhRo9C9e3f8+uuvmDhxInr16oXff/8dqamp2LBhA5ydnTFhwgQMHDgQjz76qPZWiwdHRg/q1KkT3N3dtX/sBwwYgA4dOmDt2rUAqg7N1hV01fUVFRXh9ddfh6urK1555RVDd2sN7du3R2JiomjdD/r9999F37O/vz8GDx4MJycn7Qi+oQ+JDggIQI8ePXDp0iVMnjwZXbt2rXc0WW3SpEk4dOiQ9vxo9WgSgPbK9Zs3b+Kdd95Bfn5+vSNwX19f/Pzzz1i3bh0KCwvx3Xff6fSHhITgk08+0Z6X7NatG/Lz83H06FEcPnwYV65cwYIFC9CvXz906NBBe066uT5A21AcUbYQnTt3xpYtW/Dkk08iPT0dmZmZGDVqFLZs2QIHBwedZSMjI3H69GncvHkTY8eOxUcffQSg6pdx48aN6NevH37++WdcuXIFEydOxDfffIM2bdqIbj8yMhI9evTAiRMn8NVXX2mv+HvQw6y/LsZYZ/V6W7dujT179mDr1q149NFHERUVBW9vb4P2NVAVMlFRUejUqROOHj0KPz+/Goeug4ODMXz4cJSVlSEuLg6pqamws7PD5s2b8fTTT6OwsBBbt27FhQsXEBoaCrlcjnbt2sHf3x8ZGRnYuHEjrl27hokTJ2L+/Pl1vq9BgwYBABwcHCCXy3UO31f31cbZ2RkvvfQS2rdvj++++67Oi6saws/PD35+fnotW997fuKJJ3D69Gls2rQJubm5GDVqFP75z382qK7WrVsjLi4Offv2xdmzZ1FeXq69wry+n63x48drR4Ft2rTBhAkTtH39+/fHvHnz0KFDBxw8eBBPP/209oKzurzwwgsYMWIEiouLkZqaijlz5uj0d+7cGVu3bsXo0aORlZWFxMREXL16Fc899xwcHR3RuXNnuLq64vDhw9i4cSPKy8sxY8YMTJ8+vSG7ptnhg5tJq/o2ji1btoiesyGiKrdv39YZ6U6bNg0HDhzA/PnzdW4XoqaNh16JiBpo5cqVyM/PR58+fZCVlYUDBw7Azs6OE1c0MwxKIqIG8vb2xt69e3HgwAF06NABI0eOxKJFi+Di4mLu0qgR8dArERGRCF7MQ0REJIJBSUREJIJBSUREJIJBSUREJIJBSUREJIJBSUREJOL/AeI5dgXMOzr3AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 468x378 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig , ax = plt.subplots()\n",
    "\n",
    "n , bins , patches = plt.hist( missing_values_per_SNP / number_of_isolates , bins = 50 , color = 'k' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "plt.hist( missing_values_per_SNP[missing_values_per_SNP > max_number_values_missing_per_SNP] / number_of_isolates , bins = bins , color = 'red' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "\n",
    "#plt.title('Mean Coverage across sequenced isolates in rollingDB', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.xlabel('Proportion of Isolates with Missing Values', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.ylabel('Number of Unique SNPs', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "ax.set_yscale(\"log\", nonposy='clip')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(6.5, 5.25)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.tick_params(axis='y', which='major', labelsize=12 , labelcolor = 'k')\n",
    "plt.tick_params(axis='x', which='major', labelsize=12 , labelcolor = 'k')\n",
    "\n",
    "file_name = '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/Figures/genotypes_matrix_filtering/SNP_sites_dropped_per_missing_values.pdf'\n",
    "plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of Isolates = 33873.0\n"
     ]
    }
   ],
   "source": [
    "print 'Number of Isolates = ' + str( number_of_isolates )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Range and Average Number of Missing Values per SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "62"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.min(missing_values_per_SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "948.6810697174822"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(missing_values_per_SNP)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33870"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(missing_values_per_SNP) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SNP with most missing values is missing from 99.9911433885% of isolates\n"
     ]
    }
   ],
   "source": [
    "print 'SNP with most missing values is missing from ' + str( float(np.max(missing_values_per_SNP) ) / number_of_isolates * 100) + '% of isolates' "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop SNPs that are flagged as 'missing values' in more than 10% of isolates in rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "31215\n"
     ]
    }
   ],
   "source": [
    "print sum( missing_values_per_SNP > max_number_values_missing_per_SNP )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Percentage of SNPs with many missing values that we're going to drop from analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3.589837361909428\n"
     ]
    }
   ],
   "source": [
    "number_of_unique_SNPs = float(np.shape(genotypes_matrix)[0])\n",
    "print 100.0 * ( sum( missing_values_per_SNP > max_number_values_missing_per_SNP ) / number_of_unique_SNPs )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER out sites with missing calls in $> 10\\%$ of isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs with a high abundance of missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(869538, 33873)"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape( genotypes_matrix )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs from the __GENOTYPE MATRIX__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "869538"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( missing_values_per_SNP <= max_number_values_missing_per_SNP )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNPs_to_keep_filter = missing_values_per_SNP <= max_number_values_missing_per_SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "838323"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(SNPs_to_keep_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "31215"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SNPs_to_keep_filter) - sum(SNPs_to_keep_filter) #number of sites dropped by this filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(31215, 3)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#store dataframe of SNPs that were dropped by this filtering step\n",
    "SNPs_to_drop_filter = [not good_SNP for good_SNP in SNPs_to_keep_filter]\n",
    "dropped_SNP_annotation_DF = SNP_annotation_DF[SNPs_to_drop_filter]\n",
    "dropped_SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/SNP_sites_dropped/3_dropped_SNPs_annotation_calls_missing_10perc_isolates.pkl')\n",
    "np.shape(dropped_SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_matrix = genotypes_matrix[SNPs_to_keep_filter , :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(838323, 33873)"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix) #Genotype Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique SNPs with a high abundance of missing values from the __SNP ANNOTATION FILE__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF = SNP_annotation_DF[SNPs_to_keep_filter]\n",
    "\n",
    "#re-index new filtered SNP annotation DF (so new index matches indexing of genotypes matrix rows)\n",
    "SNP_annotation_DF.reset_index(drop = True , inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>44</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>67</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>69</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   44   C     [T]\n",
       "1   48   C     [T]\n",
       "2   64   G     [C]\n",
       "3   67   G  [A, T]\n",
       "4   69   C     [T]"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(838323, 3)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF) #Annotation for Rows of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __Genotypes Matrix__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save Genotypes Matrix\n",
    "np.save('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix' , genotypes_matrix , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __SNP annotation file__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 4) *FILTER* out SNPs that located in H37Rv Reference Positions with overlapping genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load Genotypes Matrix\n",
    "## genotypes_matrix =  np.load('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix.npy')\n",
    "\n",
    "#load SNP annotation file\n",
    "## SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(838323, 33873)"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(838323, 3)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Load dictionary that contains the mapping between each H37Rv Reference Position and H37Rv locus tag, from the file: */n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/pickled_files/dicts_for_SNP_annotation/H37Rv_coord_gene_mapping.pickle', 'rb') as handle:\n",
    "    ReferencePosition_Gene_mapping = pickle.load(handle)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Rv0001']"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ReferencePosition_Gene_mapping[1] #one gene mapped to position"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Rv2035', 'Rv2036']"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ReferencePosition_Gene_mapping[2282101] #two genes mapped to position (overlapping)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ReferencePosition_Gene_mapping[2281250] #no genes mapped to position (intergenic)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER out sites that map to 2 (or more) H37Rv locus tags"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNPs_to_keep_filter = np.array( [(len(ReferencePosition_Gene_mapping[ref_pos_i]) < 2) for ref_pos_i in SNP_annotation_DF.pos] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "835979"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(SNPs_to_keep_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2344"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(SNPs_to_keep_filter) - sum(SNPs_to_keep_filter) #number of sites dropped by this filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2344, 3)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#store dataframe of SNPs that were dropped by this filtering step\n",
    "SNPs_to_drop_filter = [not good_SNP for good_SNP in SNPs_to_keep_filter]\n",
    "dropped_SNP_annotation_DF = SNP_annotation_DF[SNPs_to_drop_filter]\n",
    "dropped_SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/SNP_sites_dropped/4_dropped_SNPs_annotation_overlapping_genes.pkl')\n",
    "np.shape(dropped_SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter Genotypes Matrix\n",
    "genotypes_matrix = genotypes_matrix[SNPs_to_keep_filter , :]\n",
    "\n",
    "#filter SNP annotation file\n",
    "SNP_annotation_DF = SNP_annotation_DF[SNPs_to_keep_filter]\n",
    "SNP_annotation_DF.reset_index(drop = True , inplace = True) #re-index new filtered SNP annotation DF (so new index matches indexing of genotypes matrix rows)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 33873)"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>44</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>67</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>69</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   44   C     [T]\n",
       "1   48   C     [T]\n",
       "2   64   G     [C]\n",
       "3   67   G  [A, T]\n",
       "4   69   C     [T]"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 3)"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF) #Annotation for Rows of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __Genotypes Matrix__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save Genotypes Matrix\n",
    "np.save('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix' , genotypes_matrix , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Save __SNP annotation file__ in case kernel crashes when filtering"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [8.2] Drop low-quality isolates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### (Filter 5) *FILTER* out isolates that have missing base calls $> 10\\%$ of SNP sites"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load Genotypes Matrix\n",
    "## genotypes_matrix =  np.load('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix.npy')\n",
    "\n",
    "#load SNP annotation file\n",
    "## SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 33873)"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 3)"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Investigate the Number of Missing Calls for each Sequenced Isolate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Look for isolates with an abundance of missing values, drop isolates that have missing values > 10% of SNP calls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "missing_values = genotypes_matrix == 9\n",
    "missing_values_per_isolate = missing_values.sum(axis = 0) #will range from 0 - 835,979"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "#impose a threshold for maximum number of missing values per isolate (across all SNPs)\n",
    "max_proportion_missing = 0.10\n",
    "number_of_SNPs = float(np.shape(genotypes_matrix)[0])\n",
    "\n",
    "max_number_values_missing_per_isolate = np.floor( max_proportion_missing * number_of_SNPs )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "83597.0\n"
     ]
    }
   ],
   "source": [
    "print max_number_values_missing_per_isolate #that we're going to allow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAhIAAAFxCAYAAAAmtUHWAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzt3XlcVfW+//H3diiU8IjWwRgs3QgKhhPiUGlqegQ1j4pDetQ8aOc+bBA17ZR1bmVpg2bcbK6r2cm0FG0yI085lOaUkUKKSiqQWikmgtsE1u8Pf+wryrBYsjd76+v5ePB4sNdarPVZS2S/93d91/drMwzDEAAAgAW1aroAAADgvQgSAADAMoIEAACwjCABAAAsI0gAAADLCBIAAMAyggQAALCMIAEAACwjSAAAAMsIEgAAwDKCBAAAsIwgAQAALLvsg4TNZqvpEgAAuGxd9kECAAC4DkECAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWEaQAAAAlhEkAACAZQQJAABgGUHCAofDcUnrAQC4XNgMwzBqughXstlscsUpBgUFlbsuJyen2o8HAIAnokUCAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWEaQAAAAlhEkAACAZQQJAABgGUECAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWOY1QWLlypW66aabaroMAABwHq8IEsXFxfr0008VGBhY06UAAIDzeEWQWLFihfr166datbyiXAAArhhue2desGCBYmNj1axZMyUmJpZal5ubq4SEBIWGhiomJkYrVqxwrisqKtLHH3+sO+64w12lAgAAk+q460ABAQGaNGmS1q5dK4fDUWrdjBkzVLduXaWmpiotLU1jxoxRRESEwsPDtXz5cg0YMIDWCAAAPJDb3p3j4uLUt29f+fv7l1peUFCgVatWadq0afL19VVMTIx69+6t5cuXS5L27t2rZcuWadSoUfrpp5/06KOPuqtkAABQCbe1SJQnMzNTtWvXlt1udy6LjIzUpk2bJJ1rrSgRGxurmTNnur1GAABQthq/X5Cfny8/P79Sy/z8/JSfn3/Rtp999pm7ygIAACbUeJDw9fVVXl5eqWV5eXny9fWtoYoAAIBZNR4kmjdvrqKiImVmZjqXpaenKzw8vAarAgAAZrgtSBQWFsrhcKi4uFhFRUVyOBwqLCxU/fr1FRsbqzlz5qigoEBbt25VSkqKhgwZ4q7SAACARW4LEklJSbLb7Zo/f76Sk5Nlt9uVlJQkSZo1a5YcDoeioqI0ceJEzZ49mxYJAAC8gM0wDKOmi3Alm80mV5xiUFBQuetycnKq/XgAAHiiGu8jAQAAvBdBAgAAWEaQAAAAlhEkAACAZQQJAABgGUECAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWEaQAAAAlhEkAACAZQQJAABgGUECAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWEaQAAAAlhEkAACAZQQJAABgGUECAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWEaQAAAAlhEkAACAZQQJAABgGUHCBRwOR7VsAwCAp6tjZqPVq1dr9+7dmjRpkj766CM9++yzatiwoZ599llFRka6ukav4+Pjo6CgoAq3ycnJcVM1AAC4js0wDKOyjfr27avatWvrk08+UVRUlHJzcyVJN998s5YuXeryIi+FzWaTiVOssoqCQk5ODkECAHBFMHVr4+DBg2rVqpUyMzOVm5ur119/XT179tTOnTtdXR8AAPBgpoJEYWGhCgsLtWfPHtlsNnXt2lWBgYH6448/XF0fAADwYKb6SISGhmrlypX67LPPdOONN6phw4bKyclRQECAq+sDAAAezFSLxEMPPaTGjRurXr16evTRR/XHH38oOztbvXr1cnV9AADAg5nqbOnN6GwJAIDrmB5H4ttvv9U999yj/v376/Dhw5o3b5527NjhytoAAICHM9VH4vPPP9eECRNUXFwsm82m6667TosWLVJ6erreeOMNlxb466+/KiEhQXXr1lWtWrU0f/58+mYAAOAhTLVIzJs3Tw0bNlTnzp0lSXXq1FHHjh313XffubQ4SWrUqJFWrlyp5cuXKz4+Xu+9957LjwkAAMwxFST279+vnj17lhrFslGjRjpx4oTLCitRu3Zt1ap1rsz8/HyFh4e7/JgAAMAcU0Hiz3/+szIyMlRcXCxJys3N1YYNGxQYGGj6QAsWLFBsbKyaNWumxMTEUutyc3OVkJCg0NBQxcTEaMWKFaXW79q1S/3799eCBQt00003mT4mAABwLVNBol+/fvrhhx+0aNEiSVJ0dLQOHTqkfv36mT5QQECAJk2apOHDh1+0bsaMGapbt65SU1M1f/58PfTQQ9qzZ49zfevWrfXJJ59o2rRpevHFF00fEwAAuJapIDF58mTdcccdKioqkmEYOnPmjPr376/777/f9IHi4uLUt29f+fv7l1peUFCgVatWadq0afL19VVMTIx69+6t5cuXS5LOnDnj3LZBgwaqV6+e6WMCAADXMvXURr169fTyyy/rySefVFZWlkJCQtSoUaNqKSAzM1O1a9eW3W53LouMjNSmTZskSTt37tSsWbNUq1Yt+fj4aO7cudVyXAAAcOlMBYmQkBCNHz9e//3f/+0MEC+//LKSk5O1Zs2aSyogPz9ffn5+pZb5+fkpPz9f0rnbKMnJyZd0DAAA4Bqmbm0YhnHR6JA//fRTqX4MVvn6+iovL6/Usry8PPn6+l7yvgEAgGtV2CLRpUsX5/dLly7V559/LkkqLi7W4cOH1aBBg0suoHnz5ioqKlJmZqaaN28uSUpPT+cxTwAAvECFQSIrK0vSufkq8vLyLmo5GDlypOkDlUxFXlxcrKKiIjkcDtWpU0f169dXbGys5syZozlz5igtLU0pKSn68MMPLZwOAABwpwon7dq0aZMMw9CwYcPUr18/3XXXXZLODRIVGBio4OBg0weaO3eunn/++VLLpkyZoqlTpyo3N1dTp07V+vXr5e/vr4cffliDBg2ydkYXYNIuAABcx9Tsn5s2bdL111+vG2+80Q0lVS+CBAAArmPqqY0uXbroo48+0iuvvKLjx48735htNpvLJ+0CAACey1SQmDt3rl544QXn6/ODBAAAuHKZevzz/fffV4MGDRQVFSVJGjNmjHx8fDR+/HiXFgcAADybqRaJo0ePKj4+Xtdcc41++OEHzZo1S3l5eTp48KCr6wMAAB7M9BDZderUcc6TsWzZMu3bt0979+51aXGXM4fDIR8fH8vrAQDwBKaCRNOmTXXo0CH16dNHhmFo8uTJkqTQ0FCXFnc58/HxqfTJDwAAPJ2pIDFjxgwdOXJEPXr00MiRI/XJJ5/o2muv1TPPPOPq+gAAgAczNY6EN/PkcSRokQAAeLtyWyTmzZtnagcltzkAAMCVp9wWieDg4ArHiTAMQzabzTkfh6eiRQIAANcpt0UiPj6eAacAAECFyg0S549kCQAAUBZTT21I0rZt2/T2228rOztbwcHBGjt2rKKjo11ZGwAA8HCmgsTatWs1duxYFRUVSZK2bt2qDz/8UAsXLlTPnj1dWiAAAPBcpubamDt3rmw2m8aPH6+nn35a48ePV61atUw/2QEAAC5Pplok9u7dq4EDB+qxxx5zLsvNzdXnn3/uqroAAIAXMNUi0aBBA2VkZMjhcEg6Nw9ERkaGGjRo4NLiAACAZzPVItGtWzctWbJEbdu2VWBgoH7++Wfl5+dr+PDhrq4PAAB4MFMtEg8//LDatWunU6dOKSMjQ6dOnVLbtm318MMPu7o+AADgwUy1SDRq1EgfffSRUlNTlZWVpZCQELVp04YBqwAAuMKZapE4ceKEfv75Z7Vt21adO3fWunXr9Pbbb+vs2bOuru+KVdIfxep6AADcwVSLxD333KM9e/Zo27ZtGjdunFJTUyWde5rjqaeecmmBVyofHx/m4gAAeDxTLRLp6enq2LGjjh07pu+//15DhgxR06ZNtXr1alfXBwAAPJipIPH777+rUaNGysjIkM1m05QpU3TbbbcpNzfX1fUBAAAPZurWRuPGjfXll18qMzNT11xzjZo2baoTJ04wjgQAAFc4Uy0SgwcPVlZWljZs2KAhQ4ZIkrZv366IiAiXFgcAADybqRaJhx56SLfccosKCwvVvXt3nT17VvPmzVOTJk1cXR8AAPBgpqcRv/XWW53f16pVS126dHFJQTDH4XDIx8fnkrcBAOBSVBgkIiMjK91BWlpatRUD8yp7PFTiEVEAgOtVGCR+//33Cn+YkS0BALiyVRgkPvjgA3fVAQAAvFCFQYJ+EAAAoCKmHv8EAAAoC0ECAABYRpAAAACWlRsknnzySX3xxReSznW6LJnxEwAAoES5QeLVV1/V119/LUmaPHmyVq5c6baiAACAdyj3qY0GDRpoxYoVzrEk1q1bpylTppTaxmazae7cua6tEJYZp0/LVq+e5fUAAFSm3CDRp08fLVu2TMuWLZPNZlNGRoYyMjJKbUOQ8Gy2evWkCgYNsxmGG6sBAFyOyg0Szz//vHr16qX9+/drzpw5ateunXr06OHO2uBqDodUyVwctFoAACpSbpCoVauWBgwYIEkqLi5W+/btddttt7mrLriDj0+FLRYSrRYAgIqZmv1zypQpOnTokJ555hllZ2crJCREw4cP1w033ODq+gAAgAczFSR27Nih4cOH6/Tp0zL+/yfUN998U0uXLlW7du1cWiAAAPBcpgakmjVrlgoKCvSXv/xF99xzj/r27auCggLNnj3b1fUBAAAPZqpFIj09XXFxcXr99dedy+6++2598803LisMHqKSDpl0xgSAK5upIHH11Vfr+PHjpZYdP35cV199tUuKggeppEMmnTEB4MpmKkhER0frs88+08033yy73a7MzEwdPHhQcXFxrq4PAAB4MFNB4l//+pd27typgwcP6uDBg5KkkJAQPfLIIy4tDgAAeDZTQSI4OFhffvml1qxZo+zsbAUHB+v2229XPe6NAwBwRTMVJCSpXr16zgGqALPMdMakwyYAeC/TQaImbdu2TY8//riuuuoqBQQEKCkpSXXr1q3psiBV+lRHZfN9SHTYBABvZmociZoWHBys999/X8uXL1fTpk31+eef13RJKFHyVEd5XwCAy5pXtEg0adLE+X2dOnVUq5ZX5B8AAC57lb4jFxYWatiwYXr22Wcv+WALFixQbGysmjVrpsTExFLrcnNzlZCQoNDQUMXExGjFihUX/fyhQ4f01Vdf6fbbb7/kWgAAwKWrtEWiTp062r9/f7VM0BUQEKBJkyZp7dq1cjgcpdbNmDFDdevWVWpqqtLS0jRmzBhFREQoPDxckpSXl6fExEQlJSXpqquuuuRaAADApTN1jyAxMVGrV6/Wpk2bdPbsWcsHi4uLU9++feXv719qeUFBgVatWqVp06bJ19dXMTEx6t27t5YvXy7pXKvIxIkTNWXKFIWGhlo+PgAAqF6mgsRDDz2kEydOaNiwYWrevLlCQkIUEhKipk2bVksRmZmZql27tux2u3NZZGSk9uzZI0lauXKlduzYoRdeeEHx8fH68MMPq+W4AADg0pjubGmU8YheWcusyM/Pl5+fX6llfn5+ys/PlyTFx8crPj6+Wo4FAACqj6kg8e2337q0CF9fX+Xl5ZValpeXJ19fX5ceFwAAXBpTtzaCg4MVHBysM2fOaM+ePc7XwcHB1VJE8+bNVVRUpMzMTOey9PR0Z0dLAADgmUwFidzcXA0fPly33Xabxo0bp6ysLDVt2rTKj4QWFhbK4XCouLhYRUVFcjgcKiwsVP369RUbG6s5c+aooKBAW7duVUpKioYMGWLppAAAgHuYChJPPvmkvvnmG1111VUyDEMhISFq37691qxZU6WDJSUlyW63a/78+UpOTpbdbldSUpIkadasWXI4HIqKitLEiRM1e/ZsWiQAAPBwpvpIrF27Vl26dFFERIT+93//V5IUFhamjz76qEoHmzp1qqZOnVrmOn9/f+e+cYWpZL4OJvUCAM9lKkg4HA4FBASUWnb8+HEmzkL1KJmvoxxM6gUAnsvUrY1WrVppzZo12rFjhyRp5syZ+uKLLxQREeHS4gB3MU6frpZtAOBKYypITJ8+XX/88Ye+++47GYah1157TbVq1Sr3NgXgbZzTnVfwxe0VALiYqVsbMTExWrVqlRYtWqTs7GyFhIRo1KhRtEjAPaqhDwX9LADANUyPbNmyZUs9+uijOnz4sK6//nrV448y3MVMH4oK1ju3AQBUO1O3Nk6cOKG7775bYWFh6t69u8LCwnT33XcrNzfX1fUBAAAPZqpF4oEHHtDq1atLLfvss89UXFysN9980yWFAQAAz2eqRWLDhg1q2rSpUlJSlJmZqc8//1zBwcHasGGDq+sDPIfDUeFqnuoAcCUy1SJxww03KCoqytm5MjIyUl27dtWuXbtcWhzgURjvAgAuUm6QOH/Gz/j4eM2bN0+RkZGy2+3at2+fVq1apSlTprilSAAA4JlshlH2x6jg4GDZKvj0ZRiGatWqpUOHDrmsuOpgs9lUzilekqCgoHLX5eTkVLjezDaXur5kmwqfZjDxtEOl21zqenceozLVUScAXGHKbZEICgqqMEgAAACUGyQ2b97szjoA16pkUCt3YOAsAJcj0wNSSecm6jp9Qc/0yprXAY9QSUdJd9yWcA7DXdE23B4B4GVMBYn169dr6tSpOnLkSKnlNpvN4/tIAAAA1zE1jsQ///lPHT58WIZhlPoqLi52dX2A96hknAkAuByZapHIzc3Vbbfdptdee02+vr6urgnwTh5w+wQA3M1Ui8S4ceOUk5PjbJUAAACQTLZIxMXFaeHCherRo0ep5fSRAADgymaqReK+++7TyZMn6SMBuBrzeQDwMqZaJHJyctS2bVvNmDFDDRo0cHVNwJWL+TwAeBlTQWLUqFH6/vvv1aFDB1111VWurgnAJahsUCsGvQJQnUwFia+//lp79uxRVFSUgoODVbt2bUnn+kisXr3apQUCqJrKBr6iVQNAdTIVJH788UdJ0qlTp7R7927ncubiAADgymYqSDz//POurgMAAHghU0Fi2LBhrq4DAAB4IVNBYsqUKWUut9lsmjt3brUWBMDFKpkJtTo6Y9LhE7hymAoS77//vmw2m3NUy5LvCRKAF3LDI6Z0+ASuHFVukSgqKtLu3buVkpKiESNGuKwwAGWopDXBa5g4D1otAO9g+dbG9OnTdfjw4WovCEAFKpsYTPKOycFMnAetFoB3MD2y5fny8vK0f/9+paenu6QoADWI1gIAVWAqSHTu3LnM5S1btqzWYgB4AFoLAFSBqUm7Lpysy8fHR9HR0Zo3b56r6wMAAB7MVItEdna2q+sAAABeyFSLBAAAQFkqbJEICQmp8IdtNpsOHTpUrQUBAADvUWGLxIV9I8r6AgBPZZw+fUnrAVSuwhaJzz//vNTrgoICLViwQB9//LEMw1BkZKRLiwOA8ph5BJURNgHXqzBIlAQFh8OhhQsX6tVXX9Vvv/2mli1basqUKYqLi3NLkQBwocpCgiTvGJwL8HIVBokzZ85o0aJFevnll/Xbb7+pRYsWmjlzpgYMGOCu+gB4ostlqG4Al6zCING1a1f98ssvstlsGjBggO644w7VqlVLKSkpzm369Onj8iIBeJjKBq2iJQC4YlQYJI4ePeqc6fPjjz/Wxx9/XGo9T20AAHBlqzBIBAUFyVbZPUgAAHDFqjBIbN682V11AAAAL8TIlgAAwDKCBAAAsIwgAQAALCNIALhyORwVrmYIbaBypqYRB4DLUiXjYTCENlA5WiQAAIBlHh8kTp48qX79+qlFixbavXt3TZcDAADO4/FBol69elq0aJH69etX06UAcKdK+i8A8Awe30eibt26aty4cU2XAcDdmM8D8Apua5FYsGCBYmNj1axZMyUmJpZal5ubq4SEBIWGhiomJkYrVqxwV1kAAOASuK1FIiAgQJMmTdLatWvluKDJcsaMGapbt65SU1OVlpamMWPGKCIiQuHh4e4qDwAAWOC2Fom4uDj17dtX/v7+pZYXFBRo1apVmjZtmnx9fRUTE6PevXtr+fLl7ioNAABYVOOdLTMzM1W7dm3Z7XbnssjISO3Zs8f5evTo0Vq/fr2mTZumpUuX1kSZAFCmygatYlArXO5qvLNlfn6+/Pz8Si3z8/NTfn6+8/U777zj7rIAwBRbvXoMaoUrWo23SPj6+iovL6/Usry8PPn6+tZQRQAAwKwaDxLNmzdXUVGRMjMzncvS09PpaAmg5jGWBVAptwWJwsJCORwOFRcXq6ioSA6HQ4WFhapfv75iY2M1Z84cFRQUaOvWrUpJSdGQIUPcVRoAlK1kLIuKvoArnNuCRFJSkux2u+bPn6/k5GTZ7XYlJSVJkmbNmiWHw6GoqChNnDhRs2fPpkUCAAAvYDOMy7snkM1mkytOMSgoqNx1OTk5Fa43s82lri/ZptKRASv7RFXZNpe63luOQZ3edwxPqhO4jNV4HwkAAOC9CBIAAMAyggQAALCMIAEArlTJI6SeMvIlI3TCqhof2RIALmuVTIfuKSNfMkInrKJFAgAAWEaQAAAAlhEkAKAmmRiGm/4J8GT0kQCAmlRJHwqJ/gnwbLRIAAAAywgSAADAMoIEAACwjD4S3srhODcpF4DLn8Nxri9FOYzTp8+NA1EBM9sAVhAkvJWJDlrMOghcJswMakWHTdQQbm0AAADLCBIAAMAyggQAALCMIAEAACwjSAAAAMsIEgAAwDKCBAAAsIwgAQAALCNIAAAAywgSAADAMoIEAACwjCABAAAsI0gAAOCBjNOnL2m9uzD7JwAAHshWr17ls756AFokAACAZQQJAABgGUECAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQBA5RyOSjepjiGbvWVYaPwfhsgGAFTOx6fC4Zql6hmy2VuGhcb/oUUCAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWEaQAAAAlhEkAACAZV4RJJ5++mkNGjRIEyZM0GkmbAEAwGN4fJDYvXu3Dh48qBUrVujWW2/VkiVLarokAADw/3l8kNi8ebN69OghSerZs6e2bNlSwxUBAIASbgsSCxYsUGxsrJo1a6bExMRS63Jzc5WQkKDQ0FDFxMRoxYoVznW///67GjRoIEny8/PTiRMn3FUyAACoRB13HSggIECTJk3S2rVr5XA4Sq2bMWOG6tatq9TUVKWlpWnMmDGKiIhQeHi4/vSnP+nkyZOSpLy8PDVs2NBdJQMAgEq4rUUiLi5Offv2lb+/f6nlBQUFWrVqlaZNmyZfX1/FxMSod+/eWr58uSQpJiZG69atkyStXbtWHTt2dFfJAACgEjXeRyIzM1O1a9eW3W53LouMjNSePXskSa1atVJwcLAGDRqkdevWacSIETVVKgB4rwtagmviGIYbnrozc4zKtnFHnZcTt93aKE9+fr78/PxKLfPz81N+fr7z9UMPPeTusgDg8uLjI9ls5a83DJcfw1Ydx6iErV69is+zpI4arvNyUuMtEr6+vsrLyyu1LC8vT76+vjVUEQAAMKvGg0Tz5s1VVFSkzMxM57L09HSFh4fXYFUAAMAMtwWJwsJCORwOFRcXq6ioSA6HQ4WFhapfv75iY2M1Z84cFRQUaOvWrUpJSdGQIUPcVRoAALDIbUEiKSlJdrtd8+fPV3Jysux2u5KSkiRJs2bNksPhUFRUlCZOnKjZs2fTIgEAgBewGcbl3avEZrPJFacYFBRU7rqcnJwK15vZxsz6yjoUqZIORZWur459XC7HoE7vO4a31HmlXQszLnUf7qrTHbygzhrvIwEAALwXQQIAAFhGkAAAAJYRJAAAgGUECQAAYBlBAgAAWEaQAAAAll0R40gAAABzqhoLanz2T1e7zHMSAAA1ilsbAADAMoIEAACwjCABAAAsI0gAAADLCBIAAMAyggQAALCMIAEAACwjSAAAAMsIEgAAwDKCRBlyc3OVkJCg0NBQxcTEaMWKFWVuZxiGnnrqKUVGRioyMlJPPvkkI2mWwez1fOWVV9SzZ0+FhYWpc+fOeuWVV9xcqXcwez1L/PHHH+rWrZs6dOjgpgq9S1Wu586dOzV48GC1aNFCbdq00ZtvvunGSj2f2Wt55swZPfjgg2rTpo0iIyM1duxYHT582M3Ver4FCxYoNjZWzZo1U2JiYoXbvv7662rbtq1atmypKVOm6MyZM26q8goYItuKGTNmqG7dukpNTVVaWprGjBmjiIgIhYeHl9ru3//+t1avXq0vvvhCNptNd955p5o2baoxY8bUUOWeyez1NAxDSUlJatWqlQ4cOKCRI0cqMDBQAwcOrKHKPZPZ61nilVde0bXXXqv8/Hw3V+odzF7P48ePa9SoUXrsscfUr18/nT17lje/C5i9lm+99Za2b9+uNWvWyM/PT9OmTdOjjz5KMLtAQECAJk2apLVr18rhcJS73dq1a/XSSy/p/fffV0BAgMaPH6+5c+fq4Ycfdk+hBkrJz883brjhBmPfvn3OZffee6/x1FNPXbTtgAEDjHfeecf5evHixUa/fv3cUqe3qMr1vNAjjzxizJgxw5XleZ2qXs+DBw8a3bp1M/7zn/8Y7du3d1eZXqMq13PWrFnGvffe687yvEpVruWDDz5ozJw50/n6iy++MG655Ra31OmNnn76aWPSpEnlrp84caIxa9Ys5+v169cbbdq0cUdphmEYBrc2LpCZmanatWvLbrc7l0VGRmrPnj0XbZuRkaGIiAjn64iICGVkZLilTm9Rlet5PsMwtHnzZoWFhbm6RK9S1ev5yCOP6J///KeBKbZUAAATLElEQVR8fHzcVaJXqcr1/O677+Tv76877rhDUVFRGjt2rHJyctxZrkeryrW88847tXXrVh05ckSnT5/WihUr1KNHD3eWe1nZs2dPqfeiyMhI/frrrzp+/Lhbjk+QuEB+fr78/PxKLfPz8yuzWTg/P18NGjS4aDuDfhJOVbme55s7d66Ki4s1fPhwV5bndapyPT/77DMVFRUpNjbWXeV5napcz8OHD+uDDz7QE088oS1btqhp06aaOHGiu0r1eFW5ls2bN1dQUJA6dOig8PBw7d27V5MnT3ZXqZedgoKCi96LJLntdiZB4gK+vr7Ky8srtSwvL0++vr6Vbnvq1Cn5+vrKZrO5vE5vUZXrWWLBggVatmyZFi1apKuvvtrVJXoVs9ezoKBATz75pGbOnOnO8rxOVX4/fXx81LdvX7Vt21Y+Pj6aPHmytm3bppMnT7qrXI9WlWv50EMP6cyZM9q1a5f27t2r2NhYjR492l2lXnbq169f6tqXfF/R39nqRJC4QPPmzVVUVKTMzEznsvT09DI7soWFhSk9Pb3UdjTFl1aV6ylJS5Ys0fz587V06VIFBga6q0yvYfZ6ZmZmKjs7W4MHD1bbtm01YcIE/fLLL2rbtq2ysrLcXbbHqsrvZ6tWrUp9SCj5nhbIc6pyLdPT0zVs2DD5+/vr6quv1t///nft2LHDbU3xl5vw8PCL3ouuu+46NWrUyC3HJ0hcoH79+oqNjdWcOXNUUFCgrVu3KiUlRUOGDLlo2/j4eL3++us6fPiwjhw5otdee03Dhg2rgao9V1WuZ3Jysp5++mktWbJEN9xwQw1U6/nMXs+WLVs616WkpOi5557Tddddp5SUFALaeary+zl8+HCtXr1au3bt0tmzZ/XCCy8oJiZGf/rTn2qgcs9TlWvZpk0bLVu2TCdPntTZs2f19ttvq0mTJm574/MWhYWFcjgcKi4uVlFRkRwOhwoLCy/aLj4+XkuWLFFGRoZOnDihpKQk974Xua1bpxc5fvy4MW7cOMNutxvR0dFGcnKyYRiG8e233xqhoaHO7YqLi42ZM2caERERRkREhDFz5kyjuLi4psr2WGavZ6dOnYymTZsaoaGhzq/p06fXVNkey+z1PN8333zDUxvlqMr1XLhwodG+fXujVatWxpgxY4zs7OyaKNljmb2Wx44dM+655x7jpptuMlq2bGkMHDjQ+O6772qqbI81Z84cIzAwsNTXnDlzjOzsbCM0NLTU79+rr75qREVFGWFhYUZiYqLhcDjcVqfNMGiXAwAA1nBrAwAAWEaQAAAAlhEkAACAZQQJAABgGUECAABYRpAAAACWESSAKli6dKmCgoIUHx9fo3Xk5OQoPj5edrtdQUFB+uKLL2q0HleaO3eugoKClJiYWOF28fHxCgoK0tKlS11WS1BQkIKCgi55dNDExEQFBQVp7ty51VSZ67jjusK7ESTgcp06dXL+AQ4KClLr1q01cuRIpaam1nRpFSqpe+PGjc5lYWFhSkhIUL9+/WqwMmn+/PnatGmTbrzxRiUkJJQ5EmhxcbGef/55de3aVc2aNVNkZKTi4uL03nvvObcpeZOw2+06evSoJOn333+/6A2z5I2v5CssLEx9+/bVhx9+6PJzbd++vRISEtS9e3dJ0saNGxUUFKROnTpd8r7P/93cunWrc/nmzZudy88/TkJCghISEnTNNddc0nG7d++uhIQEtW/f/pL2Ux7DMBQTE1NmAPj4448VFBSkNm3alDlKIlBVdWq6AFw5br/9djVt2lSbNm3SunXrlJqaqnXr1unaa6+9aNuzZ8+qbt26NVBlxcdu166d2rVr5+aKLlYyn8GECRM0YsSIMrd58803nZ/mhw0bplOnTumHH37Q9u3bdeedd5ba1uFwaP78+ZVO8hUZGanOnTtr//79Wrt2rSZOnCh/f39169atek6sDD169HDLFNOLFi1Sx44dJUnvvPNOmds88cQT1XKsQYMGadCgQdWyr7LYbDYNGjRI8+fP14cfflhqFt2VK1dKkgYOHKg6dXgLwKWjRQJuc+edd2rmzJl6//33JUknTpzQ9u3blZWV5fz0t2jRIrVv3975Rpeenq5Ro0apdevWuummmzR27Fjt27fPuc+ST5QvvfSSevXq5WwxOH/yn2+//VaDBw9Wq1at1L59e9177706cuSIc33Jsd944w117txZ3bp1U6dOnZSdnS1JGjp0qPOTXVm3Nszuf8GCBbrlllsUFham++67T3/88Ue516qifcbHx+vrr7+WJE2dOrXcpvaSlpTHH39czzzzjF566SVt2LBBM2bMuGhbm82md999V4cPHy63Jknq3LmznnjiCb377rtq2bKlJOk///mPJOmNN95Q165d1bx5c910002Kj48v9W9V4tixYwoKClLLli1VXFys3NxcBQcHq23btpLO3bYJCgpSVFSUDMModWtj48aNGjp0qCQpOzvbeW3Pl5WVpaFDh8put2vgwIHOf8eKNGzYUJ9++qmOHz+uY8eOadWqVWrYsOFF213YUlPROa9YsUK33Xab7Ha7IiMjNWDAAG3ZskXSxbc2Ss5xwoQJuv/++9WiRQvdfPPNWr9+vfPYGRkZ6t+/v+x2u0aPHq1HHnlEQUFB+vvf/17mOZXMcfH111/rt99+kySdPHlSX331Van1jz/+uDp16qTmzZvLbrerf//+pVrhLnThrY6yWoh2796t0aNHKyoqSjfddJMmTJignJwcSedaS2bPnq3o6Gg1a9ZMbdu21ciRI5mwy4sRJOBWxcXF2rRpk/P1hZP0PPPMM+rRo4eio6N19OhRxcfHa+3atWrfvr1at26tNWvWaOjQoTpx4kSpn0tKSlKbNm3UqFEjrV69Wg8++KCkc0FkxIgR2rJli2677TYFBQVpxYoVGjVqlM6ePVtqH08//bQ6deqkbt26acSIEc7m67i4OCUkJJQ5s2tV9j9nzhxFR0erqKhIycnJWr58eZnXqLJ99uvXT02aNJEkdevWrdym9j//+c+SpGnTpum+++7TwoULlZWVJX9//4u27d+/v86cOaMXX3yxzJoutG/fPmewadSokX766Sc99thjysvL09ChQ9WtWzfl5OTol19+uehnGzdurLCwMOXl5Wn37t3atm2bDMPQr7/+qp9++knbtm2TdC4knj/bpiRdf/31iouLkyRdc801zlsN53vxxRcVEBCgRo0aadu2bXrmmWcqPZ/4+HidOXNGS5Ys0ZIlS3TmzBlnYClPRed8+vRpTZ48WdnZ2Ro0aJB69uypU6dO6cCBAxXuc9WqVTp69KjCw8N14MABTZ06VdK5yZvGjRunHTt2KCwsTD4+Plq0aFGF+woLC1NkZKSKior0ySefSJI+/fRTnTlzRna7XW3atJF0Lni1a9dOI0aMUNeuXbVjxw794x//0KlTpyq9bmX55ZdfNGTIEG3YsEExMTFq166dVq1apZEjR+rMmTPasGGD5s+fr9q1a2vEiBHq1KmTdu/erfz8fEvHQ82jXQtuc+Ef/N69e6tDhw7OTyqS9Nprr+mWW26RJL388sv6/fff1aVLF+cfzT59+igtLU2ffPKJ/va3vzl/bvr06Ro/frzS0tLUp08frVq1Svn5+XrnnXd09uxZDRs2TPPmzdPZs2cVHR2t3bt3a+PGjc777pL01FNPlbpNsGTJEp06dUrjxo1T165dJZ37VHi+quz/6aef1oABA2QYhpYtW6Zdu3aVeZ0q2+e4ceP06aef6siRI/rrX/9aqtn6fImJidq1a5dSU1OVnJys5ORkPfroo0pMTHS+QZXo0KGD8vLy9N5772n06NFl7k+S3nrrLb311lvO1yEhIfrb3/7m/DQZEBCguLg4tWjRQoGBgSoqKipzP507d1ZGRoa2bdumnJwchYSE6OjRo9qyZYt27twpSerSpctFP9esWTONGzfO2WJQ1q2GUaNGadasWVq6dKmmTJmitLS0cs+nRJcuXbR+/Xq9++67MgxDYWFh6ty5s954441yf6akf0FZ5+xwOFRUVKTGjRvrL3/5i8LCwnTDDTeUez1KhIeHa8mSJcrKylKXLl30888/6/jx49q3b58OHDggX19fLV++XPXr19e4ceOUkpJS4f4GDx6stLQ0rVy5UnfddZfztsb5M3I+99xz+vTTT5WVlaVmzZqpXr16On78uH788UfnrZ6qWL58uU6cOOG8HtK58Lhv3z5t3LjROe36jTfeqAEDBigsLEyNGzdmOnYvRosE3Ob222/X+PHjNW3aNL3zzjtasGCBatUq/St4/h+ukubjFi1aOJeFhoZK0kXN1SXblKyXpMOHDzu3K1lft25dNW3atMx9REdHV/mcqrL/1q1bS5Jz2unyPoFVZZ8VCQwM1KpVq7RmzRo9+eST6tixo4qLizVv3rwym5EfeOAB/fHHH/qf//mfcvcZGRmphIQETZo0SUlJSfrqq6/UuHFjtWjRQg888ICOHDmikSNHqmPHjurWrZv27t1b5n5KQsK2bdu0detWde3aVa1bt9a2bducLRJlBQkzSq5zgwYNJJV/nS80evRoHThwQAcPHtSYMWMq3b6ic/b19dXs2bMlSXfddZe6du2q6Ohobd68ucJ9RkZGymazOWsvqb/kllNgYKDq168vSWW2kF1o0KBBqlWrlrZt26bvvvtOmzZtks1m0+DBgyVJx48fV69evfTggw9q/vz5euutt3T69GnnOjMuDEcl/2/37t3rDJ7Hjh2TJB04cEDdu3fX2LFj9f3332vo0KFq06aN+vfvr19//dXU8eB5CBJwmzvvvFOPP/64EhMT1bNnz4uarSXp6quvdn4fEhIiSaXus+/fv1+SFBwcXOrnSt6wzt/2+uuvd25Xsvzs2bM6dOhQmfs4/9iSVLt2bUnnbseUpyr7N9uxrSr7rMj27dt1+vRptWrVSuPGjdPbb78t6dw96oKCgou2b9eunXr16qWPP/643H2W9JGYPn264uPjVa9ePUnn3kzuv/9+7dq1S1u2bNE999yj/fv3l/uJviQkbN68WampqerYsaM6duyoDRs2KD09Xf7+/s4+GBeq7N+lZH1Zv18VKTmf+vXrl/rEXp7Kznno0KHavn27vvvuOz3xxBM6fPiwXnjhhQr3WV7t119/vaRz4bjkjb6s/icXCggI0M033yzDMHTfffepqKhIMTExzv9bW7Zs0dGjR9W4cWN9//33+umnn5xBt7wWgpJ/85JbH3v27Cm1vmTfcXFxysnJcX7t2LFDI0aMUFFRkZ566in9+OOP+uabbxQfH6/vv/9eixcvrvR84Jm4tQGPNXjwYL344ovauHGj7rrrLp09e1a7du3Sddddd9Hjl88995zS09OdncRiY2Pl6+urUaNGafHixfrggw/kcDiUnZ2t3377TeHh4ZV+4g0MDNTBgwc1Z84cpaSk6B//+MdF21zK/stTXft89dVXtXHjRkVHR+v666933jJo3rz5RR0USzzwwAPOzpNV8fPPP6t///7q1KmTrr32WuejlOd/sj7fddddp9DQUOebYceOHdWwYUO99tprks4FlvKCQElz+eHDh/XAAw+oWbNmuueee6pc84UaNGig5OTkCus+X2Xn3LZtW3Xp0kUBAQHavXu3pP9rjaqq9u3b68Ybb9SBAwc0dOhQBQYGVnpbo8TgwYO1YcMGZ/+MktYISc4npo4dO6bHHntMBw8erLQFp3Xr1vryyy/1+uuvKycnp9TjxNK5VpAXX3zR2S8iODhYBw8e1Lfffquvv/5aWVlZmjx5sjp06KCGDRs6W6CsXhvUPFok4LGaNGmiDz74QN27d9fWrVuVmpqqXr166YMPPriow+CUKVO0c+dOHTt2TH369NGzzz4r6dwfvcWLF6tDhw768ssvlZWVpYEDB+rf//63rrrqqgqPP2XKFN14443avn273nrrLWfP9/Ndyv7LU137vOOOO9SmTRvt3LlTS5cuVU5OjmJjY7Vw4cJy36SjoqLUp0+fKtd8zTXXqG3bttq2bZsWL16so0ePauDAgZo0aVK5P9O5c2dJkr+/v+x2e6nbWiXryhISEqL/+q//UoMGDfTee++V22nViqioKEVFRZnatrJzvvXWW7Vz504tWbJEGRkZ6tWrl/71r39ZqqtOnTpasGCB2rVrpx9//FGnT592PtlU2e9Ev379nK0IV111lfr37+9cFx0drfvvv18NGzbU+vXr9de//tXZkbc8d999t3r06KHjx49r48aNmjBhQqn1TZo00bJly3T77bcrLS1NycnJOnLkiMaOHatGjRqpSZMmatasmb7++mstXrxYp0+f1ujRozVq1CgrlwYewGbQwwVerOQxzQ8++MDZIRK4HJ08ebJUS8nIkSO1bt06TZo0SdOnT6/BynCl49YGAHiBuXPn6uDBg2rTpo3S0tK0bt061a9fv9wByQB3IUgAgBeIiIjQF198oXXr1qlhw4bq2bOnpk+f7nyiB6gp3NoAAACW0dkSAABYRpAAAACWESQAAIBlBAkAAGAZQQIAAFhGkAAAAJb9P7Cf92NWCAbgAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 540x378 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig , ax = plt.subplots()\n",
    "\n",
    "n , bins , patches = plt.hist( missing_values_per_isolate / number_of_SNPs , bins = 50 , color = 'k' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "plt.hist( missing_values_per_isolate[missing_values_per_isolate > max_number_values_missing_per_isolate] / number_of_SNPs , bins = bins , color = 'red' , rwidth = 1.0 , edgecolor='white', linewidth=1.0)\n",
    "\n",
    "\n",
    "#plt.title('Mean Coverage across sequenced isolates in rollingDB', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.xlabel('Proportion of SNPs with Missing Values', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "plt.ylabel('Number of Isolates', fontweight = 'bold' , fontsize = 12, color = 'k')\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "ax.set_yscale(\"log\", nonposy='clip')\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(7.5, 5.25)\n",
    "fig.tight_layout()\n",
    "\n",
    "plt.tick_params(axis='y', which='major', labelsize=12 , labelcolor = 'k')\n",
    "plt.tick_params(axis='x', which='major', labelsize=12 , labelcolor = 'k')\n",
    "\n",
    "file_name = '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/Figures/genotypes_matrix_filtering/Samples_dropped_per_missing_values.pdf'\n",
    "plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True)\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique SNPs = 835979.0\n"
     ]
    }
   ],
   "source": [
    "print 'Number of unique SNPs = ' + str( number_of_SNPs )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Range and Average Number of Missing Values per unique SNP"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "170"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.min(missing_values_per_isolate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "17680.30714728545"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.mean(missing_values_per_isolate)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "835979"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.max(missing_values_per_isolate) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Isolate with most missing values has 100.0% missing values\n"
     ]
    }
   ],
   "source": [
    "print 'Isolate with most missing values has ' + str( float(np.max(missing_values_per_isolate) ) / number_of_SNPs * 100) + '% missing values' "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop isolates that are flagged as 'missing values' in more than 10% of isolates in rollingDB"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1663\n"
     ]
    }
   ],
   "source": [
    "print sum( missing_values_per_isolate > max_number_values_missing_per_isolate )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Percentage of isolates with many missing values that we're going to drop from analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "4.90951495291235\n"
     ]
    }
   ],
   "source": [
    "number_of_unique_isolates = float(np.shape(genotypes_matrix)[1])\n",
    "print 100.0 * ( sum( missing_values_per_isolate > max_number_values_missing_per_isolate ) / number_of_unique_isolates )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### FILTER out low-quality isolates with $> 10\\%$ missing SNP calls"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop unique isolates with a high abundance of missing values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 33873)"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape( genotypes_matrix )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop isolates from the __GENOTYPE MATRIX__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33873"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len( missing_values_per_isolate <= max_number_values_missing_per_isolate )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolates_to_keep_filter = missing_values_per_isolate <= max_number_values_missing_per_isolate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "32210"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sum(isolates_to_keep_filter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1663"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(isolates_to_keep_filter) - sum(isolates_to_keep_filter) #number of isolates dropped by this filter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genotypes_matrix = genotypes_matrix[: , isolates_to_keep_filter]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 32210)"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(genotypes_matrix) #Genotype Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Drop isolates with a high abundance of missing values from __ISOLATE ANNOTATION FILE__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_annotation_series = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes/genotypes_isolate_annotation.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN05575260\n",
       "2    SAMN03648641\n",
       "3    SAMN03647419\n",
       "4    SAMEA3671418\n",
       "dtype: object"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_annotation_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(33873,)"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_annotation_series) #Annotation for Columns of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(1663,)"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#store dataframe of isolates that were dropped by this filtering step\n",
    "isolates_to_drop_filter = [not good_isolate for good_isolate in isolates_to_keep_filter]\n",
    "dropped_isolate_annotation_series = isolate_annotation_series[isolates_to_drop_filter]\n",
    "dropped_isolate_annotation_series.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/isolate_IDs_dropped/1_dropped_isolates_annotation_calls_missing_10perc_SNPs.pkl')\n",
    "dropped_isolate_annotation_series.to_csv('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/isolate_IDs_dropped/1_dropped_isolates_annotation_calls_missing_10perc_SNPs.csv')\n",
    "np.shape(dropped_isolate_annotation_series)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "isolate_annotation_series = isolate_annotation_series[isolates_to_keep_filter]\n",
    "\n",
    "#re-index new filtered isolate annotation series (so new index matches indexing of genotypes matrix columns)\n",
    "isolate_annotation_series.reset_index(drop = True , inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    SAMEA3558733\n",
       "1    SAMN03648641\n",
       "2    SAMN03647419\n",
       "3    SAMEA3671418\n",
       "4    SAMN07659096\n",
       "dtype: object"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "isolate_annotation_series.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(32210,)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(isolate_annotation_series) #Annotation for Columns of Genotypes Matrix"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Save annotation files and genotypes matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "#save isolate annotation file (columns of Genotype Matrix)\n",
    "isolate_annotation_series.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_isolate_annotation.pkl')\n",
    "\n",
    "#save SNP annotation file (rows of Genotype Matrix)\n",
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')\n",
    "\n",
    "#save Genotypes Matrix\n",
    "np.save('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_matrix' , genotypes_matrix , allow_pickle = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [9] Annotate SNPs w/ H37Rv Locus Info"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retrieve information for H37Rv Reference Genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 150,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/rv76/anaconda2/lib/python2.7/site-packages/ipykernel_launcher.py:18: DeprecationWarning: \n",
      ".ix is deprecated. Please use\n",
      ".loc for label based indexing or\n",
      ".iloc for positional indexing\n",
      "\n",
      "See the documentation here:\n",
      "http://pandas.pydata.org/pandas-docs/stable/indexing.html#ix-indexer-is-deprecated\n"
     ]
    }
   ],
   "source": [
    "####################################################################################################################################################################################\n",
    "#load reference genome and reference annotation\n",
    "reference_genome = '/home/rv76/Farhat_Lab/Reference_Seqs/H37Rv/h37rv.fasta'\n",
    "for reference_genome in SeqIO.parse(reference_genome, \"fasta\"):\n",
    "    reference_genome.seq.alphabet = IUPAC.unambiguous_dna\n",
    "\n",
    "reference_genome_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')\n",
    "\n",
    "####### Make a dictionary of all H37Rv reference positions and corresponding genes #######\n",
    "ReferencePosition_Gene_mapping = {} #keys: H37Rv Reference Positions , values: gene_ids (may be multiple genes for some Reference Positions)\n",
    "\n",
    "#store a list corresponding to every Reference Position (to store all genes that map to the Reference Position)\n",
    "for H37Rv_RefPos in range(0 , len(reference_genome.seq) + 1):\n",
    "    ReferencePosition_Gene_mapping[H37Rv_RefPos] = []\n",
    "\n",
    "for gene_id_index in range(0 , len(reference_genome_annotation.index)):\n",
    "\n",
    "    gene_id_info = reference_genome_annotation.ix[gene_id_index , :]\n",
    "    gene_id = gene_id_info.name\n",
    "\n",
    "    chrom_start = gene_id_info.chromStart\n",
    "    chrom_end = gene_id_info.chromEnd\n",
    "\n",
    "    #find the position of the first base relative to H37Rv in 5' -> 3'\n",
    "    H37Rv_start = min(chrom_start , chrom_end)\n",
    "    H37Rv_end = max(chrom_start , chrom_end)\n",
    "\n",
    "    #store all corresponding H37Rv Reference Positions to gene_id in dictionary\n",
    "    for H37Rv_RefPos in range(H37Rv_start+1 , H37Rv_end+1):\n",
    "\n",
    "        ReferencePosition_Gene_mapping[H37Rv_RefPos].append(gene_id) #append gene_id to list already in dict\n",
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Annotate SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To annotate SNPs, we get gene_ids that the SNPs 'land in' and the gene-relative coordinate."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>44</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>67</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>69</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt\n",
       "0   44   C     [T]\n",
       "1   48   C     [T]\n",
       "2   64   G     [C]\n",
       "3   67   G  [A, T]\n",
       "4   69   C     [T]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#load SNP annotation file (rows of Genotype Matrix)\n",
    "SNP_annotation_DF = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation.pkl')\n",
    "\n",
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 3)"
      ]
     },
     "execution_count": 152,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SNP_annotation_DF)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Categories"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get Gene Categories\n",
    "gene_categories = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/CSV_files/gene_categories/gene_categories.csv').set_index('name')\n",
    "\n",
    "#create dictionary, keys: H37Rv locus, values: gene category\n",
    "gene_categories_map = dict( zip( list(gene_categories.gene_id) , list(gene_categories.Gene_Category) ) ) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Gene Names"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Parse through Mycobrowser (tuberculist) to collect gene names"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [],
   "source": [
    "mycobrowser_table = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/Mycobrowser_Release3/Mycobacterium_tuberculosis_H37Rv_txt_v3.txt' , sep = '\\t')\n",
    "\n",
    "#retreive columns for locus tag & gene name\n",
    "mycobrowser_table = mycobrowser_table.loc[: , ['Locus' , 'Name']]\n",
    "\n",
    "#create dictionary, keys: H37Rv loucs, values: gene category\n",
    "gene_names_map = dict( zip( list(mycobrowser_table.Locus) , list(mycobrowser_table.Name) ) ) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Iterate through each SNP to annotate"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Some SNPs are present in more than 1 gene, have to account for this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 162,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0.0\n",
      "0.20000023924\n",
      "0.400000478481\n",
      "0.600000717721\n",
      "0.800000956962\n"
     ]
    }
   ],
   "source": [
    "#create dictionaries for each columsn of information\n",
    "locus_tag_dict = {} #H37Rv id\n",
    "genomic_coord_dict = {} #coordinate on gene, 5' -> 3'\n",
    "gene_categ_dict = {} #gene Category\n",
    "gene_name_dict = {} #gene Name\n",
    "\n",
    "for SNP_index in SNP_annotation_DF.index:\n",
    "\n",
    "    #reference position, ref allele and alt allele for SNP\n",
    "    SNP_info = SNP_annotation_DF.loc[SNP_index , :]\n",
    "\n",
    "    #find the gene that SNP occurs on\n",
    "    ref_seq_position = SNP_info.pos\n",
    "\n",
    "    #check list corresponding to H37Rv coordinate to see if there are any genes associated with RefPosition\n",
    "    if len(ReferencePosition_Gene_mapping[ref_seq_position]) == 1:\n",
    "        \n",
    "        #get gene that ReferencePosition is mapped to (i.e. SNP filtering ensures that SNP corresponds to only 1 gene)\n",
    "        gene_intergenic_id = ReferencePosition_Gene_mapping[ref_seq_position][0]\n",
    "\n",
    "        #find genomic coordinate of SNP relative to gene (subtract 1 since reference seq starts counting at 1)\n",
    "        gene_relative_coord = (ref_seq_position - 1) - min( reference_genome_annotation.loc[gene_intergenic_id , 'chromStart'] , reference_genome_annotation.loc[gene_intergenic_id , 'chromEnd'] )\n",
    "\n",
    "        #store all information for SNP in relevant dictionaries\n",
    "        locus_tag_dict[SNP_index] = gene_intergenic_id\n",
    "        genomic_coord_dict[SNP_index] = gene_relative_coord\n",
    "\n",
    "        #use gene id to get category and names from mappings\n",
    "        if gene_intergenic_id in gene_categories_map.keys():\n",
    "            gene_categ_dict[SNP_index] = gene_categories_map[gene_intergenic_id]\n",
    "        else:\n",
    "            gene_categ_dict[SNP_index] = 'None'\n",
    "\n",
    "        if gene_intergenic_id in gene_names_map.keys():\n",
    "            gene_name_dict[SNP_index] = gene_names_map[gene_intergenic_id] \n",
    "        else:\n",
    "            gene_name_dict[SNP_index] = 'None'\n",
    "            \n",
    "    #H37Rv RefSeq position was not found in dictionary so Reference position must be intergenic    \n",
    "    elif len(ReferencePosition_Gene_mapping[ref_seq_position]) == 0:\n",
    "\n",
    "        #store all information for SNP in relevant dictionaries\n",
    "        locus_tag_dict[SNP_index] = 'intergenic'\n",
    "        genomic_coord_dict[SNP_index] = np.nan\n",
    "        gene_categ_dict[SNP_index] = 'None'\n",
    "        gene_name_dict[SNP_index] = 'None'\n",
    "\n",
    "    #keep track of progress of loop\n",
    "    num_SNPs = float( np.shape(SNP_annotation_DF)[0] )\n",
    "    if SNP_index % np.ceil(0.20 * num_SNPs) == 0:\n",
    "        print float(SNP_index) / num_SNPs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Update SNP annotation DF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 163,
   "metadata": {},
   "outputs": [],
   "source": [
    "#append columns for info from annotation\n",
    "SNP_annotation_DF['gene_id'] = pd.Series(locus_tag_dict)\n",
    "SNP_annotation_DF['gene_pos'] = pd.Series(genomic_coord_dict)\n",
    "SNP_annotation_DF['gene_category'] = pd.Series(gene_categ_dict)\n",
    "SNP_annotation_DF['gene_name'] = pd.Series(gene_name_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 164,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(835979, 7)"
      ]
     },
     "execution_count": 164,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape( SNP_annotation_DF )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 165,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>44</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>43.0</td>\n",
       "      <td>Essential</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>48</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>47.0</td>\n",
       "      <td>Essential</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>64</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>63.0</td>\n",
       "      <td>Essential</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>67</td>\n",
       "      <td>G</td>\n",
       "      <td>[A, T]</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>66.0</td>\n",
       "      <td>Essential</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>69</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>Rv0001</td>\n",
       "      <td>68.0</td>\n",
       "      <td>Essential</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   pos ref     alt gene_id  gene_pos gene_category gene_name\n",
       "0   44   C     [T]  Rv0001      43.0     Essential      dnaA\n",
       "1   48   C     [T]  Rv0001      47.0     Essential      dnaA\n",
       "2   64   G     [C]  Rv0001      63.0     Essential      dnaA\n",
       "3   67   G  [A, T]  Rv0001      66.0     Essential      dnaA\n",
       "4   69   C     [T]  Rv0001      68.0     Essential      dnaA"
      ]
     },
     "execution_count": 165,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 166,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>835974</th>\n",
       "      <td>4411494</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835975</th>\n",
       "      <td>4411499</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835976</th>\n",
       "      <td>4411500</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835977</th>\n",
       "      <td>4411501</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>835978</th>\n",
       "      <td>4411502</td>\n",
       "      <td>C</td>\n",
       "      <td>[G]</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            pos ref  alt     gene_id  gene_pos gene_category gene_name\n",
       "835974  4411494   C  [G]  intergenic       NaN          None      None\n",
       "835975  4411499   A  [G]  intergenic       NaN          None      None\n",
       "835976  4411500   C  [T]  intergenic       NaN          None      None\n",
       "835977  4411501   C  [T]  intergenic       NaN          None      None\n",
       "835978  4411502   C  [G]  intergenic       NaN          None      None"
      ]
     },
     "execution_count": 166,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 167,
   "metadata": {},
   "outputs": [],
   "source": [
    "#update SNP annotation file (rows of Genotype Matrix) with extra SNP information\n",
    "SNP_annotation_DF.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/rolling_DB_scrape/Genotypes_Filtered/genotypes_SNP_annotation_with_gene_info.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How to investigate SNPs that appeared within a particular gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 168,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>160931</th>\n",
       "      <td>759814</td>\n",
       "      <td>A</td>\n",
       "      <td>[G]</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>7.0</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160932</th>\n",
       "      <td>759818</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>11.0</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160933</th>\n",
       "      <td>759836</td>\n",
       "      <td>C</td>\n",
       "      <td>[T]</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>29.0</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160934</th>\n",
       "      <td>759841</td>\n",
       "      <td>G</td>\n",
       "      <td>[C]</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>34.0</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>160935</th>\n",
       "      <td>759851</td>\n",
       "      <td>G</td>\n",
       "      <td>[T]</td>\n",
       "      <td>Rv0667</td>\n",
       "      <td>44.0</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>rpoB</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos ref  alt gene_id  gene_pos          gene_category gene_name\n",
       "160931  759814   A  [G]  Rv0667       7.0  Antibiotic Resistance      rpoB\n",
       "160932  759818   C  [T]  Rv0667      11.0  Antibiotic Resistance      rpoB\n",
       "160933  759836   C  [T]  Rv0667      29.0  Antibiotic Resistance      rpoB\n",
       "160934  759841   G  [C]  Rv0667      34.0  Antibiotic Resistance      rpoB\n",
       "160935  759851   G  [T]  Rv0667      44.0  Antibiotic Resistance      rpoB"
      ]
     },
     "execution_count": 168,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SNP_annotation_DF[['Rv0667' in gene_list for gene_list in SNP_annotation_DF.gene_id]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:root] *",
   "language": "python",
   "name": "conda-root-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
