{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>.container { width:100% !important; }<|/style>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from IPython.core.display import display, HTML\n",
    "display(HTML(\"<style>.container { width:100% !important; }<|/style>\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib.patches import Rectangle\n",
    "\n",
    "import Bio\n",
    "from Bio.Alphabet import IUPAC\n",
    "from Bio.Blast.Applications import NcbiblastnCommandline\n",
    "from Bio.Blast import NCBIXML\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio import SeqIO\n",
    "from Bio.Graphics import GenomeDiagram\n",
    "from Bio.SeqUtils import GC\n",
    "from Bio.Align.Applications import MuscleCommandline\n",
    "from StringIO import StringIO\n",
    "from Bio import AlignIO\n",
    "from Bio.Align import AlignInfo\n",
    "from Bio.Seq import MutableSeq\n",
    "\n",
    "import itertools\n",
    "import sys\n",
    "import re\n",
    "\n",
    "from collections import Counter\n",
    "\n",
    "#for exporting to Adobe Illustrator\n",
    "mpl.rcParams['pdf.fonttype'] = 42\n",
    "mpl.rcParams['ps.fonttype'] = 42"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [1] Load reference genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load reference genome and reference annotation\n",
    "reference_genome = '/home/rv76/Farhat_Lab/Reference_Seqs/H37Rv/h37rv.fasta'\n",
    "for reference_genome in SeqIO.parse(reference_genome, \"fasta\"):\n",
    "    reference_genome.seq.alphabet = IUPAC.unambiguous_dna\n",
    "\n",
    "reference_genome_annotation = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/inhost_TB_dynamics_project/H37Rv/h37rv_genome_summary.txt', '\\t').set_index('name')\n",
    "\n",
    "# convert H37Rv sequence to a string\n",
    "string = str(reference_genome.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "4411532"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(reference_genome.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [2] Find short sequence repeat patterns (2-6bp short sequences with at least 3 repeats) in H37Rv genome "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get lists to store columns of what will become dataframe\n",
    "start_index_list = []\n",
    "SSR_motif_list = []\n",
    "num_repeats_list = []\n",
    "end_index_list = []\n",
    "\n",
    "# iterate through different lengths for the simple sequence repeat motifs\n",
    "for SSR_length in [2, 3, 4, 5, 6]:\n",
    "\n",
    "    # get all combinations of bases for the length of this short sequence\n",
    "    seq_combo_tuples = list(itertools.permutations(['A','C','G','T']*(SSR_length-1), SSR_length))\n",
    "    seq_combo_strings = [''.join(x) for x in seq_combo_tuples]\n",
    "    seq_combo_strings = list(set(seq_combo_strings))\n",
    "\n",
    "    # iterate through each short sequence of this length\n",
    "    for SS_i in seq_combo_strings:\n",
    "\n",
    "        num_repeats = 3 # start with 3 repeats & increase\n",
    "\n",
    "        # iterate through number of repeats for this simple sequence and find matches in ref genome\n",
    "        pattern = SS_i*num_repeats\n",
    "        start_indices_match = [match.start() for match in re.finditer(pattern, string)]\n",
    "\n",
    "        # update lists\n",
    "        start_index_list = start_index_list + list( np.array(start_indices_match) + 1 )\n",
    "        SSR_motif_list = SSR_motif_list + [SS_i]*len(start_indices_match)\n",
    "        num_repeats_list = num_repeats_list + [num_repeats]*len(start_indices_match)\n",
    "        end_index_list = end_index_list + list( np.array(start_indices_match) + num_repeats*SSR_length )\n",
    "\n",
    "        # keep increasing number of repeats until we don't find anymore in ref genome\n",
    "        while len(start_indices_match) > 0:\n",
    "            num_repeats += 1\n",
    "\n",
    "            # iterate through number of repeats for this simple sequence and find matches in ref genome\n",
    "            pattern = SS_i*num_repeats\n",
    "            start_indices_match = [match.start() for match in re.finditer(pattern, string)]\n",
    "\n",
    "            # update lists\n",
    "            start_index_list = start_index_list + list( np.array(start_indices_match) + 1 )\n",
    "            SSR_motif_list = SSR_motif_list + [SS_i]*len(start_indices_match)\n",
    "            num_repeats_list = num_repeats_list + [num_repeats]*len(start_indices_match)\n",
    "            end_index_list = end_index_list + list( np.array(start_indices_match) + num_repeats*SSR_length )\n",
    "        \n",
    "# concatenate lists into a DataFrame\n",
    "SSR_H37Rv_df = pd.DataFrame()\n",
    "\n",
    "SSR_H37Rv_df.loc[:,'H37Rv_start'] = start_index_list\n",
    "SSR_H37Rv_df.loc[:,'H37Rv_end'] = end_index_list\n",
    "SSR_H37Rv_df.loc[:,'motif'] = SSR_motif_list\n",
    "SSR_H37Rv_df.loc[:,'repeats'] = num_repeats_list\n",
    "\n",
    "# sort by Ref Pos, then by # repeats, drop duplicates and keep only row with largest num of repeats\n",
    "SSR_H37Rv_df.sort_values(by = ['H37Rv_start','repeats'], ascending = [True,False], inplace = True)\n",
    "SSR_H37Rv_df.drop_duplicates(subset = ['H37Rv_start','motif'], inplace = True, keep = 'first')\n",
    "\n",
    "# reset index\n",
    "SSR_H37Rv_df.reset_index(inplace = True, drop = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>561</td>\n",
       "      <td>569</td>\n",
       "      <td>CGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562</td>\n",
       "      <td>570</td>\n",
       "      <td>GCC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>643</td>\n",
       "      <td>648</td>\n",
       "      <td>AC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1121</td>\n",
       "      <td>1126</td>\n",
       "      <td>GC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1408</td>\n",
       "      <td>1413</td>\n",
       "      <td>CA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   H37Rv_start  H37Rv_end motif  repeats\n",
       "0          561        569   CGC        3\n",
       "1          562        570   GCC        3\n",
       "2          643        648    AC        3\n",
       "3         1121       1126    GC        3\n",
       "4         1408       1413    CA        3"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(18316, 4)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(SSR_H37Rv_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Counter({2: 10989, 3: 7161, 4: 97, 5: 16, 6: 53})"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# count the number of motifs with >= 3 repeats found in H37Rv of each length: 2, 3, 4 ...\n",
    "Counter(SSR_H37Rv_df.motif.apply(len))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>842</th>\n",
       "      <td>215864</td>\n",
       "      <td>215878</td>\n",
       "      <td>TCGGT</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4977</th>\n",
       "      <td>1188073</td>\n",
       "      <td>1188087</td>\n",
       "      <td>GCGGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4978</th>\n",
       "      <td>1188074</td>\n",
       "      <td>1188088</td>\n",
       "      <td>CGGCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4979</th>\n",
       "      <td>1188075</td>\n",
       "      <td>1188089</td>\n",
       "      <td>GGCGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4980</th>\n",
       "      <td>1188076</td>\n",
       "      <td>1188090</td>\n",
       "      <td>GCGCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4981</th>\n",
       "      <td>1188077</td>\n",
       "      <td>1188091</td>\n",
       "      <td>CGCGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5907</th>\n",
       "      <td>1398960</td>\n",
       "      <td>1398974</td>\n",
       "      <td>CGTCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7826</th>\n",
       "      <td>1869363</td>\n",
       "      <td>1869377</td>\n",
       "      <td>CGGCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7827</th>\n",
       "      <td>1869364</td>\n",
       "      <td>1869378</td>\n",
       "      <td>GGCGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7828</th>\n",
       "      <td>1869365</td>\n",
       "      <td>1869379</td>\n",
       "      <td>GCGCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8015</th>\n",
       "      <td>1914715</td>\n",
       "      <td>1914729</td>\n",
       "      <td>GGCGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8494</th>\n",
       "      <td>2038612</td>\n",
       "      <td>2038626</td>\n",
       "      <td>GGCGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12635</th>\n",
       "      <td>3022140</td>\n",
       "      <td>3022154</td>\n",
       "      <td>CGCCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13002</th>\n",
       "      <td>3112528</td>\n",
       "      <td>3112542</td>\n",
       "      <td>ACCCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13003</th>\n",
       "      <td>3112529</td>\n",
       "      <td>3112543</td>\n",
       "      <td>CCCGA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15808</th>\n",
       "      <td>3810470</td>\n",
       "      <td>3810484</td>\n",
       "      <td>CCCGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       H37Rv_start  H37Rv_end  motif  repeats\n",
       "842         215864     215878  TCGGT        3\n",
       "4977       1188073    1188087  GCGGC        3\n",
       "4978       1188074    1188088  CGGCG        3\n",
       "4979       1188075    1188089  GGCGC        3\n",
       "4980       1188076    1188090  GCGCG        3\n",
       "4981       1188077    1188091  CGCGG        3\n",
       "5907       1398960    1398974  CGTCG        3\n",
       "7826       1869363    1869377  CGGCG        3\n",
       "7827       1869364    1869378  GGCGC        3\n",
       "7828       1869365    1869379  GCGCG        3\n",
       "8015       1914715    1914729  GGCGC        3\n",
       "8494       2038612    2038626  GGCGG        3\n",
       "12635      3022140    3022154  CGCCG        3\n",
       "13002      3112528    3112542  ACCCG        3\n",
       "13003      3112529    3112543  CCCGA        3\n",
       "15808      3810470    3810484  CCCGG        3"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df[SSR_H37Rv_df.motif.apply(len) == 5]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### save dataframe for downstream analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "SSR_H37Rv_df.to_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/CSV_files/simple_sequence_repeat_regions/SSR_H37Rv_2-6bp_at_least_3_repeats.pkl')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [3] Analyze coverage of the H37Rv genome "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "SSR_H37Rv_df = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/CSV_files/simple_sequence_repeat_regions/SSR_H37Rv_2-6bp_at_least_3_repeats.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>561</td>\n",
       "      <td>569</td>\n",
       "      <td>CGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>562</td>\n",
       "      <td>570</td>\n",
       "      <td>GCC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>643</td>\n",
       "      <td>648</td>\n",
       "      <td>AC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1121</td>\n",
       "      <td>1126</td>\n",
       "      <td>GC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1408</td>\n",
       "      <td>1413</td>\n",
       "      <td>CA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   H37Rv_start  H37Rv_end motif  repeats\n",
       "0          561        569   CGC        3\n",
       "1          562        570   GCC        3\n",
       "2          643        648    AC        3\n",
       "3         1121       1126    GC        3\n",
       "4         1408       1413    CA        3"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "H37Rv_SSR_positions_list = []\n",
    "for SSR_i_start, SSR_i_end in zip(SSR_H37Rv_df.H37Rv_start, SSR_H37Rv_df.H37Rv_end):\n",
    "    \n",
    "    SSR_i_H37Rv_coords = list(np.arange(SSR_i_start, SSR_i_end + 1))\n",
    "    \n",
    "    H37Rv_SSR_positions_list = H37Rv_SSR_positions_list + SSR_i_H37Rv_coords\n",
    "    \n",
    "H37Rv_SSR_positions_list = list(set(H37Rv_SSR_positions_list)) # remove duplicate positions from overlapping SSR regions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** use this number when describing how much of the genome is covered by SSR regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "99665"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(H37Rv_SSR_positions_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2.259192498206972"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "100.0 * (float(len(H37Rv_SSR_positions_list)) / float(len(reference_genome.seq)))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Start at -1bp relative to the SSR track since that's where variants are usually called"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "SSR_H37Rv_df.loc[:, 'H37Rv_start'] = SSR_H37Rv_df.H37Rv_start - 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>560</td>\n",
       "      <td>569</td>\n",
       "      <td>CGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>561</td>\n",
       "      <td>570</td>\n",
       "      <td>GCC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>642</td>\n",
       "      <td>648</td>\n",
       "      <td>AC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1120</td>\n",
       "      <td>1126</td>\n",
       "      <td>GC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1407</td>\n",
       "      <td>1413</td>\n",
       "      <td>CA</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   H37Rv_start  H37Rv_end motif  repeats\n",
       "0          560        569   CGC        3\n",
       "1          561        570   GCC        3\n",
       "2          642        648    AC        3\n",
       "3         1120       1126    GC        3\n",
       "4         1407       1413    CA        3"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### save dataframe as excel sheet for publication"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "SSR_H37Rv_df.to_excel('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/CSV_files/simple_sequence_repeat_regions/SSR_H37Rv_2-6bp_at_least_3_repeats.xlsx')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note:** use this list when sorting INDEL variants that occur within an SSR region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "H37Rv_SSR_positions_list = []\n",
    "for SSR_i_start, SSR_i_end in zip(SSR_H37Rv_df.H37Rv_start, SSR_H37Rv_df.H37Rv_end):\n",
    "    \n",
    "    SSR_i_H37Rv_coords = list(np.arange(SSR_i_start, SSR_i_end + 1))\n",
    "    \n",
    "    H37Rv_SSR_positions_list = H37Rv_SSR_positions_list + SSR_i_H37Rv_coords\n",
    "    \n",
    "H37Rv_SSR_positions_list = list(set(H37Rv_SSR_positions_list)) # remove duplicate positions from overlapping SSR regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112673"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(H37Rv_SSR_positions_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [4] Load in Homoplasy Scores from *TopDis (spacer length = 4)* for INDELs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#load in INDELs\n",
    "all_homoplasic_INDELs = pd.read_pickle('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/CSV_files/homoplasies_detected_in_global_lineages/INDEL homoplasies/homoplasy_count_across_lineages_spacer_4.pkl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>ins_del</th>\n",
       "      <th>INDEL_type</th>\n",
       "      <th>codon_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>4C</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>total</th>\n",
       "      <th>num_isolates</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47371</th>\n",
       "      <td>4408100</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>103</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>41.0</td>\n",
       "      <td>27.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>388</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47365</th>\n",
       "      <td>4408087</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>116</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>23.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47313</th>\n",
       "      <td>4407851</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>21.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>182</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25151</th>\n",
       "      <td>2288850</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>pncA</td>\n",
       "      <td>Rv2043c</td>\n",
       "      <td>392</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47315</th>\n",
       "      <td>4407851</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>52</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos ref  alt          gene_category gene_name  gene_id gene_pos  \\\n",
       "47371  4408100  GC    G  Antibiotic Resistance       gid  Rv3919c      103   \n",
       "47365  4408087  CG    C  Antibiotic Resistance       gid  Rv3919c      116   \n",
       "47313  4407851  GC    G  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "25151  2288850   A  ACC  Antibiotic Resistance      pncA  Rv2043c      392   \n",
       "47315  4407851   G   GC  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "\n",
       "      ins_del  INDEL_type codon_pos      ...          1     2     3    4A  \\\n",
       "47371     del  frameshift        35      ...       41.0  27.0  51.0  43.0   \n",
       "47365     del  frameshift        39      ...       23.0  11.0  35.0  33.0   \n",
       "47313     del  frameshift       118      ...       21.0   9.0  40.0  20.0   \n",
       "25151     ins  frameshift       131      ...        1.0  14.0   6.0   3.0   \n",
       "47315     ins  frameshift       118      ...        5.0   1.0  12.0   5.0   \n",
       "\n",
       "         4B    4C    5    6  total  num_isolates  \n",
       "47371  11.0  29.0  0.0  0.0  202.0           388  \n",
       "47365  10.0  18.0  0.0  0.0  130.0           212  \n",
       "47313   6.0  19.0  0.0  0.0  115.0           182  \n",
       "25151   6.0   2.0  0.0  0.0   32.0            48  \n",
       "47315   7.0   1.0  0.0  0.0   31.0            52  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46306, 22)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(all_homoplasic_INDELs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note - SNPs & INDELs with homoplasy score $ = 0$ were dropped from DataFrames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [5] Separate INDELs into those occcurring in HT regions, SSR regions and those not in either"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### HT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "homopolymeric_regions_from_Luca = pd.read_csv('/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/CSV_files/comparison_with_homopolymeric_regions/homopolymer regions from Luca/list_homopolymers_7bpmore_pure.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "create a set of all Reference Positions that occur within a Homopolymeric Tract"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_HT_ref_postions = []\n",
    "for HT_i_start, HT_i_end in zip(homopolymeric_regions_from_Luca.chromStart, homopolymeric_regions_from_Luca.chromEnd):\n",
    "    \n",
    "    HT_i_ref_positions = range(HT_i_start, HT_i_end+1)\n",
    "    all_HT_ref_postions = all_HT_ref_postions + HT_i_ref_positions\n",
    "    \n",
    "all_HT_ref_postions = set(all_HT_ref_postions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1160"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(all_HT_ref_postions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### SSR"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "set of all Reference Positions that occur within a SSR region (including positions at -1bp relative to SSR region)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "H37Rv_SSR_positions = set(H37Rv_SSR_positions_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "112673"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(H37Rv_SSR_positions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Annotate each INDEL based on whether it is in a SSR or HT or other"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "iterate through all INDELs and figure out which occur in SSR and those that don't"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "SSR_region_list = []\n",
    "for indel_i_pos in all_homoplasic_INDELs.pos:\n",
    "    \n",
    "    if indel_i_pos in all_HT_ref_postions:\n",
    "        SSR_region_list.append('HT')\n",
    "    \n",
    "    elif indel_i_pos in H37Rv_SSR_positions:\n",
    "        SSR_region_list.append('yes')\n",
    "        \n",
    "    else:\n",
    "        SSR_region_list.append('other')\n",
    "        \n",
    "all_homoplasic_INDELs.loc[:,'SSR_region'] = SSR_region_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>ins_del</th>\n",
       "      <th>INDEL_type</th>\n",
       "      <th>codon_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>4C</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>total</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>SSR_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>47371</th>\n",
       "      <td>4408100</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>103</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>35</td>\n",
       "      <td>...</td>\n",
       "      <td>27.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>202.0</td>\n",
       "      <td>388</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47365</th>\n",
       "      <td>4408087</td>\n",
       "      <td>CG</td>\n",
       "      <td>C</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>116</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>39</td>\n",
       "      <td>...</td>\n",
       "      <td>11.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>33.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>130.0</td>\n",
       "      <td>212</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47313</th>\n",
       "      <td>4407851</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>182</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25151</th>\n",
       "      <td>2288850</td>\n",
       "      <td>A</td>\n",
       "      <td>ACC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>pncA</td>\n",
       "      <td>Rv2043c</td>\n",
       "      <td>392</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>131</td>\n",
       "      <td>...</td>\n",
       "      <td>14.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>48</td>\n",
       "      <td>other</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47315</th>\n",
       "      <td>4407851</td>\n",
       "      <td>G</td>\n",
       "      <td>GC</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>52</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos ref  alt          gene_category gene_name  gene_id gene_pos  \\\n",
       "47371  4408100  GC    G  Antibiotic Resistance       gid  Rv3919c      103   \n",
       "47365  4408087  CG    C  Antibiotic Resistance       gid  Rv3919c      116   \n",
       "47313  4407851  GC    G  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "25151  2288850   A  ACC  Antibiotic Resistance      pncA  Rv2043c      392   \n",
       "47315  4407851   G   GC  Antibiotic Resistance       gid  Rv3919c      352   \n",
       "\n",
       "      ins_del  INDEL_type codon_pos     ...         2     3    4A    4B    4C  \\\n",
       "47371     del  frameshift        35     ...      27.0  51.0  43.0  11.0  29.0   \n",
       "47365     del  frameshift        39     ...      11.0  35.0  33.0  10.0  18.0   \n",
       "47313     del  frameshift       118     ...       9.0  40.0  20.0   6.0  19.0   \n",
       "25151     ins  frameshift       131     ...      14.0   6.0   3.0   6.0   2.0   \n",
       "47315     ins  frameshift       118     ...       1.0  12.0   5.0   7.0   1.0   \n",
       "\n",
       "         5    6  total  num_isolates  SSR_region  \n",
       "47371  0.0  0.0  202.0           388       other  \n",
       "47365  0.0  0.0  130.0           212       other  \n",
       "47313  0.0  0.0  115.0           182         yes  \n",
       "25151  0.0  0.0   32.0            48       other  \n",
       "47315  0.0  0.0   31.0            52         yes  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6] Breakdown of INDELs by SSR region and type"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_homoplastic_INDEL_breakdown_stats(all_homoplasic_INDELs):\n",
    "    \n",
    "    print('(A1) INDELs in ncRNA')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.stop_codon == 'RNA'])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.stop_codon == 'RNA'].total >= 2)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.stop_codon == 'RNA'])[0])))\n",
    "    print('')\n",
    "\n",
    "    print('(A2) INDELs in intergenic')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.stop_codon == 'intergenic'])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.stop_codon == 'intergenic'].total >= 2)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.stop_codon == 'intergenic'])[0])))\n",
    "    print('')\n",
    "\n",
    "    print('(A3) INDELs in coding')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA')])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA')].total >= 2)) / float(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA')])[0])))\n",
    "    print('')\n",
    "\n",
    "    print('(B1) INDELs in coding & inframe')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA') & (all_homoplasic_INDELs.INDEL_type == 'inframe')])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA') & (all_homoplasic_INDELs.INDEL_type == 'inframe')].total >= 2)) / float(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA') & (all_homoplasic_INDELs.INDEL_type == 'inframe')])[0])))\n",
    "    print('')\n",
    "\n",
    "    print('(B2) INDELs in coding & frameshift')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')].total >= 2)) / float(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon != 'intergenic') & (all_homoplasic_INDELs.stop_codon != 'RNA') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0])))\n",
    "    print('')\n",
    "\n",
    "    print('(C1) INDELs in coding & frameshift & no change to STOP')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'no change') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0]))\n",
    "    try:\n",
    "        print(str(float(sum(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'no change') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')].total >= 2)) / float(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'no change') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0])))\n",
    "    except ZeroDivisionError:\n",
    "        print('ZeroDivisionError: float division by zero')\n",
    "    print('')\n",
    "\n",
    "    print('(C2) INDELs in coding & frameshift & premature STOP first 3/4 of gene')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'premature stop - first 3/4') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'premature stop - first 3/4') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')].total >= 2)) / float(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'premature stop - first 3/4') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0])))\n",
    "    print('')\n",
    "\n",
    "    print('(C3) INDELs in coding & frameshift & premature STOP last 1/4 of gene')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'premature stop - last 1/4') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'premature stop - last 1/4') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')].total >= 2)) / float(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'premature stop - last 1/4') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0])))\n",
    "    print('')\n",
    "\n",
    "    print('(C4) INDELs in coding & frameshift & elongation STOP appears after annotated STOP')\n",
    "    print(str(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'elongation - appears after annotated stop') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0]))\n",
    "    print(str(float(sum(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'elongation - appears after annotated stop') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')].total >= 2)) / float(np.shape(all_homoplasic_INDELs[(all_homoplasic_INDELs.stop_codon == 'elongation - appears after annotated stop') & (all_homoplasic_INDELs.INDEL_type == 'frameshift')])[0])))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6.1] all INDELs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46306, 23)"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(all_homoplasic_INDELs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.1143048417051786"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "float(sum(all_homoplasic_INDELs.total >= 2)) / float(np.shape(all_homoplasic_INDELs)[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6.2] INDELs *in* HT regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(330, 23)"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'HT'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.7515151515151515"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'HT'].total >= 2)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'HT'])[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(A1) INDELs in ncRNA\n",
      "3\n",
      "1.0\n",
      "\n",
      "(A2) INDELs in intergenic\n",
      "67\n",
      "0.761194029851\n",
      "\n",
      "(A3) INDELs in coding\n",
      "260\n",
      "0.746153846154\n",
      "\n",
      "(B1) INDELs in coding & inframe\n",
      "30\n",
      "0.566666666667\n",
      "\n",
      "(B2) INDELs in coding & frameshift\n",
      "230\n",
      "0.769565217391\n",
      "\n",
      "(C1) INDELs in coding & frameshift & no change to STOP\n",
      "0\n",
      "ZeroDivisionError: float division by zero\n",
      "\n",
      "(C2) INDELs in coding & frameshift & premature STOP first 3/4 of gene\n",
      "117\n",
      "0.777777777778\n",
      "\n",
      "(C3) INDELs in coding & frameshift & premature STOP last 1/4 of gene\n",
      "60\n",
      "0.783333333333\n",
      "\n",
      "(C4) INDELs in coding & frameshift & elongation STOP appears after annotated STOP\n",
      "53\n",
      "0.735849056604\n"
     ]
    }
   ],
   "source": [
    "get_homoplastic_INDEL_breakdown_stats(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'HT'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6.3] INDELs *in* SSR regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(2077, 23)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.2585459797785267"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes'].total >= 2)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes'])[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(A1) INDELs in ncRNA\n",
      "8\n",
      "0.0\n",
      "\n",
      "(A2) INDELs in intergenic\n",
      "246\n",
      "0.191056910569\n",
      "\n",
      "(A3) INDELs in coding\n",
      "1823\n",
      "0.268787712562\n",
      "\n",
      "(B1) INDELs in coding & inframe\n",
      "842\n",
      "0.407363420428\n",
      "\n",
      "(B2) INDELs in coding & frameshift\n",
      "981\n",
      "0.149847094801\n",
      "\n",
      "(C1) INDELs in coding & frameshift & no change to STOP\n",
      "2\n",
      "0.5\n",
      "\n",
      "(C2) INDELs in coding & frameshift & premature STOP first 3/4 of gene\n",
      "570\n",
      "0.145614035088\n",
      "\n",
      "(C3) INDELs in coding & frameshift & premature STOP last 1/4 of gene\n",
      "229\n",
      "0.157205240175\n",
      "\n",
      "(C4) INDELs in coding & frameshift & elongation STOP appears after annotated STOP\n",
      "180\n",
      "0.15\n"
     ]
    }
   ],
   "source": [
    "get_homoplastic_INDEL_breakdown_stats(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6.4] INDELs *not in* HT or SSR regions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(43899, 23)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'other'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.10269026629308185"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'other'].total >= 2)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'other'])[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(A1) INDELs in ncRNA\n",
      "216\n",
      "0.180555555556\n",
      "\n",
      "(A2) INDELs in intergenic\n",
      "7198\n",
      "0.12461794943\n",
      "\n",
      "(A3) INDELs in coding\n",
      "36485\n",
      "0.0979032479101\n",
      "\n",
      "(B1) INDELs in coding & inframe\n",
      "4806\n",
      "0.0801081980857\n",
      "\n",
      "(B2) INDELs in coding & frameshift\n",
      "31679\n",
      "0.100602923072\n",
      "\n",
      "(C1) INDELs in coding & frameshift & no change to STOP\n",
      "66\n",
      "0.181818181818\n",
      "\n",
      "(C2) INDELs in coding & frameshift & premature STOP first 3/4 of gene\n",
      "19143\n",
      "0.100558951053\n",
      "\n",
      "(C3) INDELs in coding & frameshift & premature STOP last 1/4 of gene\n",
      "7404\n",
      "0.0953538627769\n",
      "\n",
      "(C4) INDELs in coding & frameshift & elongation STOP appears after annotated STOP\n",
      "5066\n",
      "0.107382550336\n"
     ]
    }
   ],
   "source": [
    "get_homoplastic_INDEL_breakdown_stats(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'other'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6.5] $\\sum{Hs}$ for FS in HT regions as a proportion of $\\sum{Hs}$ for FS across genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(39844, 23)"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs[all_homoplasic_INDELs.INDEL_type == 'frameshift'].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "71351.0"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs[all_homoplasic_INDELs.INDEL_type == 'frameshift'].total.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(294, 23)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs[(all_homoplasic_INDELs.INDEL_type == 'frameshift')&(all_homoplasic_INDELs.SSR_region == 'HT')].shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10405.0"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_homoplasic_INDELs[(all_homoplasic_INDELs.INDEL_type == 'frameshift')&(all_homoplasic_INDELs.SSR_region == 'HT')].total.sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "14.582836960939579"
      ]
     },
     "execution_count": 55,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(float(10405) / float(71351))*100.0"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [6.6] proportion of INDELs with $Hs \\ge X$ for varying values of X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "Hs_thresholds = np.arange(1, 51)\n",
    "INDELs_HT_prop_list = []\n",
    "INDELs_SSR_prop_list = []\n",
    "INDELs_other_prop_list = []\n",
    "\n",
    "for Hs_i in Hs_thresholds:\n",
    "    \n",
    "    INDELs_HT_prop_list.append(float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'HT'].total >= Hs_i)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'HT'])[0]))\n",
    "    INDELs_SSR_prop_list.append(float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes'].total >= Hs_i)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes'])[0]))\n",
    "    INDELs_other_prop_list.append(float(sum(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'other'].total >= Hs_i)) / float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'other'])[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "plt.style.use('ggplot')\n",
    "plt.rcParams['lines.linewidth']=1.0\n",
    "plt.rcParams['axes.facecolor']='1.0'\n",
    "plt.rcParams['xtick.color']='black'\n",
    "plt.rcParams['axes.grid']=False\n",
    "plt.rcParams['axes.edgecolor']='black'\n",
    "plt.rcParams['grid.color']= '1.0'\n",
    "plt.rcParams.update({'font.size': 14})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAgwAAAF8CAYAAACjTONTAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4zLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvIxREBQAAIABJREFUeJzs3XlYVPX+B/D3mYU5Mwz7IsrikpKpQeF6jUyNcl/r5lKp6S0zzfpZaZq5lLl1zTbR0tJM6966LSplJW655BaBqbnliiiCqCyzL78/CBKFYYA5DMy8X8/jE3M4c/jg8yRvvsvnK9jtdjuIiIiIHJC5uwAiIiKq+xgYiIiIqFIMDERERFQpBgYiIiKqFAMDERERVYqBgYiIiCrFwEBERESVYmAgIiKiSjEwEBERUaUYGIiIiKhS9TowCILg7hKIiIi8gqSB4f3330e7du2gUqkwatQoh/cuXrwYERERCAgIwOjRo2E0GqUsjYiIiKpA0sDQqFEjTJ8+HaNHj3Z4348//oj58+dj8+bNOHPmDE6dOoWZM2fW6GtbrVY0jesIMaYdmsZ1hNVqrdHziIiIvJlQG6dVTp8+HZmZmVi1alW5nx8+fDiaNGmCuXPnAgA2b96MRx99FJcuXXL4XEEQUFH5TeM64qz6aciCgmHP0yHG8DZOH9xbo++DiIjIWyncXQAAHD58GAMGDCh9HR8fj+zsbFy5cgUhISFl7k1NTUVqamqlz7x4zQpZpB8En1DAPxIXLy1yed1ERETeok4EhsLCQgQEBJS+Lvm4oKDglsCQlJSEpKQkAMCCBQsqfGbDQDnOF56AXX0n7Pm/oGGgXILKiYiIvEOdCAxarRb5+fmlr0s+9vPzq/YzT/62G006tMWlgB5oZFqJk7/trnGdRERUt+h0OhQWFnLXnAMymQzBwcE1/juqE4GhdevWyMjIwCOPPAIAyMjIQIMGDW4ZXagKuVyO/Zt+QONn0vHVfzZALucIAxGRJ7l27RoAICwsjIHBAYPBgLy8vBr9TAUk3iVhsVhgMBhgtVphtVphMBhgsVhuuW/EiBH46KOPcOTIEVy9ehVz5sypdBumMwI0AZD5Xca5y+YaP4uIiOoWs9mMwMBAhoVKiKIIm81W4+dIGhjmzJkDtVqN+fPnY82aNVCr1ZgzZw7OnTsHrVaLc+fOAQB69uyJyZMno1u3bmjcuDEaN26M2bNn1/jriyoRMk02/rykr/GziIiIvFmtbKuUiqNtlSX8Xn4RjzZ8Fsuea1xLVRERUW3IyclBWFiYu8uoF1zxd1WvW0M7Q6a6hDOXOCVBRES1q0mTJre0AVi1ahUSExOh1WpL/8hkMqjV6tLXa9eudVPFjtWJRY9SkvlcRlZWzeduiIio7ntq4us4fq5QsufHxmjx4buv1vg5hYV/19ikSROsWLGitGVAXeXxgUHhk4tLeVwQQ0TkDY6fK8TOwnHSfYFzS6V7dh3n8VMSCsVVXC+Uw2Sut0s1iIiI3M7zRxjsgL/WhKwrFjSJULq7HCIi8iIDBw6EQvH3j1qTyYSEhAQ3VlR9nj/CYJMjwK8I5y7ztEoiIqpd3377La5du1b6Jzk52d0lVZvnjzDYFNBoC5CZc2vDKCIiInKOx48wKO1KaHyv4fxlBgYiIqLq8vzAACVUmjyc55QEERFRtXn8lIQPfKBQ5+I8pySIiDxebIxW0q2PsTFap+89c+bMLddGjRp1y1lJ5d1XF3l8YFAJPpCrL+P8SQYGIiJP54qmSlQ+j5+SUAkiZGI2zudwSoKIiKi6PD4wiDIVrMorMJjsKNSzRTQREVF1eHxgUMvUMNmNiA6Tc6cEERFRNXl8YNAo1DDYjYgKU3BagoiIqJo8PjCoFRqYYEJMuIIjDERERNXk8YFB81dgiAqXc2slERFRNXl8YNAqfWEWzH+NMHBKgoiIqDo8PjD4+hQHhqgwTkkQERFVl8cHBj8fLawyK6LD5DyAioiIas3OnTvRuXNnBAQEIDg4GPfccw/2798Pk8mEF154AVFRUdBqtWjatCn+7//+r/R9TZo0gVqthlarRUREBEaNGoXCwkI3fifFPL7To5/KDxajFdHhxbsk7HY7BEFwd1lERCSBp18dhz+v/inZ828Lug3LXq+89XR+fj769u2LpUuX4pFHHoHJZMKOHTugUqkwb948HDhwAPv27UPDhg1x9uxZ/Pzzz2Xev2HDBiQlJeHSpUvo0aMH5s2bhzfeeEOqb8spHh8Y/NX+sBVZoVXLIPoIuJJvQ2iA3N1lERGRBP68+icyHjgs3RfY5Nxtx48fBwAMGzYMAKBWq/Hggw8CAKZNm4ZBgwahUaNGAIpHFJo0aVLucyIiItCjRw+kp6fXrG4X8PgpCX/RHzaFHQAQEy7HOa5jICIiicXGxkIul2PkyJHYuHEjrl69Wvq5Tp064a233kJycjJ+//132O32Cp+TmZmJjRs3onnz5rVRtkMeHxj8RD/YFXbY7DZEhSmQyZ0SREQkMX9/f+zcuROCIODJJ59EWFgY+vfvj+zsbEydOhVTpkzB2rVr0a5dO0RGRuKTTz4p8/6BAwfCz88P0dHRCA8Px+zZs930nfzN4wODRq0BzIDeakB0mIK9GIiIqFbccccdWLVqFTIzM3Ho0CFkZWXh+eefh1wux/jx47Fr1y5cu3YNr7zyCkaPHo0//vij9L3ffvstCgoKsG3bNhw9ehS5ublu/E6KeXxgEEURMAF6qx7RnJIgIiI3aNmyJUaNGoVDhw6Vua5WqzF+/HgEBQXhyJEjt7zvvvvuw6hRo/Diiy/WVqkV8pLAYIfBakB0uAKZPE+CiIgkdvToUSxatAiZmZkAgPPnz+Pzzz9Hp06d8Pbbb2Pbtm3Q6/WwWCz45JNPUFBQgLvvvrvcZz3//PPYtGmT2xc+ekVgsBvtf40wKDjCQEREkvPz88PevXvRsWNH+Pr6olOnTmjTpg0WLVoEtVqNF154AREREQgNDcWSJUvw1VdfoVmzZuU+KywsDCNGjMDrr79ey99FWR6/rVKtVsNmsBUHhjA5MhkYiIg81m1Btzm99bHaz3dCZGQkvvjii3I/N3bsWIwdO7bC9545c+aWa0uXVt77QWoeHxhUKhVsBht0Fh3uCFXg0lUrLFY7FHI2byIi8jTONFWi6vH4KQm5XA6YgQJjAZQKAWEBcly8wnUMREREVeHxgQEAZGYB+YZ8AEBUGI+5JiIiqiqvCAxyqxzX/woMxcdcMzAQERFVhdcEhkJj8UlfUeEKnGe3RyIioirxmsBQYCwAUDzCwGOuiYiIqsYrAoPSrkShuQgAEB3Gbo9ERERV5R2BAQoUmYoDQ1QYuz0SERFVlVcEBh+7D4osxYEhpgG7PRIRUd3SpEkTpKamursMh7wjMMAHOosOABAWIEOBzga90ebmqoiIyBuNGjUK06dPd3cZVebxnR4BQCWooLfoAQAymYCoMAXO51gRG+UVeYmIyGs89dRTOH78uGTPj42NxYcffijZ82vKYrFAoZDmR7t3BAaZCgarofR1VFhxL4bYKKUbqyIiIlc7fvw4tm/f7u4yAAB//PEHxo0bh/T0dERGRmLevHm4dOkS1q5dC0EQ8Pbbb6Nbt27YsGEDACA9PR2TJk3C2bNn0bNnT3zyySfFJy4DSElJwfTp03HmzBm0atUKy5YtQ1xcHIDi6Yxx48Zh7dq1OHbsGIqKiiQJDV7xK7YoE6G3/R0YosPk3FpJRESSMZvN6NevHx588EFcvnwZ7733Hh599FHcd999ePTRRzF58mQUFhaWhgUA+OKLL/DDDz/g9OnTOHjwIFatWgUASEtLw+jRo/HBBx/gypUrGDt2LPr37w+j0Vj63s8//xzfffcdrl27JtkIg1cEBo1cDaP977/Y4mOuuVOCiIiksWfPHhQWFuLll1+Gj48Punfvjr59++Lzzz+v8D0TJ05Eo0aNEBwcjH79+iE9PR0AsHz5cowdOxYdO3aEXC7HyJEjoVKpsGfPnjLvjY6Ohlqtlux78orAoFbcGhh4zDUREUklKysL0dHRkMn+/jHbuHFjXLhwocL3RERElH6s0WhQWFjcofjs2bNYtGgRAgMDS/+cP38eWVlZpfdHR0dL8F2U5RWBQSPXwART6evocDZvIiIi6TRq1Ajnz5+Hzfb3jrxz584hMjISgiBU6VnR0dF45ZVXcO3atdI/Op0Ow4YNK72nqs+sDu8IDMqbAgObNxERkYQ6duwIX19fLFy4EGazGdu2bcOGDRswdOhQNGjQAKdOnXL6WU8++SSWLVuGvXv3wm63o6ioCN999x0KCgok/A5u5RWBQav0hVn4e0QhOkyB8zkW2O12N1ZFRESeysfHB+vXr8fGjRsRGhqKZ555BqtXr0bLli0xZswYHDlyBIGBgRg4cGClz2rXrh2WL1+OCRMmICgoCM2bNy9dEFmbvGJbpa+PFhbZ34EhQCuDTACuFdoQ5Cd3Y2VERORKsbGxdeb5rVu3LneLZ4sWLUoXNJY4c+ZMmdezZs0q87pnz57o2bNnuV/n5vdKxSsCg99NgQEoXvh4PsfKwEBE5EHqclOl+s4rpiT81H6wyWyw2f9efFK8tZILH4mIiJzhFYFBLaohs8qgt97UvImBgYiIyCleERhEUYTMIoPeqi+9VnKeBBEREVXOawKDYBbKBIaY8OLzJIiIiKhyXhQYUPYAqnA5zvM8CSIiIqdIGhjy8vIwaNAg+Pr6onHjxvjss8/Kvc9oNOLpp59GgwYNSntoO2qfWVWiKAImlDPCwCkJIqL6SqlU4tq1a+ypUwmDwVCmRXV1Sbqtcvz48fDx8UF2djbS09PRp08fxMfHo3Xr1mXue+edd/DLL7/g4MGDCAgIwJNPPolnn30WX3/9tUvqEEURdqMduhvXMIQqkHXFAqvVDrlc+paaRETkWoGBgdDpdMjJyamV1sj1lUwmQ3BwcI2fI1lgKCoqwldffYVDhw5Bq9UiMTER/fv3x6effor58+eXuff06dPo0aMHGjRoAAAYOnQoJk2a5LJa1Go17EZbmREGlY+AIK0M2VetaBTqFe0oiIg8jkajgUajcXcZXkGyn5THjx+HXC4v0xUrPj6+3K5XY8aMwXPPPYesrCwEBgZi7dq16NWrV7nPTU1NRWpqapVqEUURNkPZwACUtIhmYCAiIqqMZD8pCwsLERAQUOZaQEBAuYdlxMbGIiYmBpGRkZDL5bjzzjvx/vvvl/vcpKQkJCUlAQAWLFjgVC0VBYaocAUycyzoeIfKqecQERF5K8kWPWq1WuTn55e5lp+fDz8/v1vuHTduHAwGA65cuYKioiIMHjy4whGG6hBFERa95ZbAEMNjromIiJwiWWCIjY2FxWLBiRMnSq9lZGTcsuCx5PqoUaMQHBwMlUqFZ599Fvv27UNubq5LahFFEZYiS5ltlcBfzZu4U4KIiKhSkgUGX19fDB48GDNmzEBRURF27dqFdevW4fHHH7/l3vbt22P16tW4fv06zGYzkpOT0ahRI4SGhrqkFqVSCZvBhkJzUZnrbN5ERETkHEn7MCQnJ0Ov1yM8PBzDhg3D0qVL0bp1a+zYsQNarbb0vn//+98QRREtWrRAWFgYvv/+e3zzzTcuq0MQBMhtchQay66fiApj8yYiIiJnSLo9IDg4GN9+++0t1++9914UFhaWvg4JCcHatWulLAVKmwKFpsIy16LDFcjkeRJERESV8orW0ACgsCtRaCo7JRERJEdegRVGE7uEEREROeI1gcHHrkDRTWsY5HIBDYPluJDLaQkiIiJHvCYwKOGDIovuluvR4QpurSQiIqqE1wQGFXygLy8whHEdAxERUWW8JzAIqlsaNwHcKUFEROQMLwsMhluusxcDERFR5aoVGPR6PUwmk6trkZRarobBdmtgiApnt0ciIqLKOBUYXnzxRezbtw8A8N133yE4OBiBgYHYsGGDpMW5klqhhtFuLHPNbrfj+/X/w45fMzFr7hLY7dxeSUREVB6nAsPatWvRpk0bAMBrr72GNWvWYP369Zg2bZqkxbmSRn5rYJg9Lxmrtgeg0BqCBSkNMXtespuqIyIiqtuc6vSo0+mg0Whw5coVnDp1Cg899BAA4OzZs5IW50oahQYmmMtcO3DwHMyaPgAAsyYBBw4udUdpREREdZ5TgSE2NhZr167FyZMn8cADDwAAcnNzoVarJS3OlTRKX5hvCgzt4mKwJSUNJk0CBMtVtIuLcVN1REREdZtTgSE5ORnPPfccfHx88NFHHwEAfvzxRzz44IOSFudKvj4aWAQLbHYbZELxTMzMqc8ASMae9BXYmT8STz75lHuLJCIiqqMEez1e6ScIgtMLFadPn46P236KE/3+gK9Cc8vnn3k7F41CFZj+WKCryyQiIqr3HI4wnDp1qtIHNGvWzGXFSEmtVkNhlUNv1ZcbGMb288eA6dl4eVgAFHLBDRUSERHVXQ4DQ/Pmzcv8Fn/zb/SCIMBqrR89DERRhOyvwFCe+Nt8EBWmwPd79Oh/z62BgoiIyJs53FZps9lgtVphs9lgs9ng7+9f+nHJ5+oLURQht8oqDAwA8HQ/PyzbkF+LVREREdUPVer0KAj1d6heFEUIZhkM5bSHLvHwfRqknTDhzyxzhfcQERF5I685S6I4MAA6ByMMoo8Mjz+oxYcpBbVYGRERUd3ndYHB0ZQEADzVxw+rfyqE0VRvN48QERG5nMNFj48//niZaYiioiKMGDGizD2rV6+WpjIXE0URMFYeGFpEKRF/mw++2lGE4fdra6k6IiKiuq3SXRI3qk9nR9xMrVbDbrJXGhiA4sWPb32Zz8BARET0F4eBYebMmbVVh+REUYTNaHMqMPT9hwbPvZ+H30+ZcGczn1qojoiIqG7zqjUMNoMNekvlgUEhFzCmtx8+2MDFj0RERIC3BQa9FXpbxdsqbzSmtxb/3VaEAp1N4sqIiIjqPq8KDBad1akpCQCIDFXgvngRn20ulLgyIiKius/LAoPZ6cAAAGP7+eHDlAKnD7giIiLyVE4Hht27d5d+vGvXLkmKkVJxYLA4tYahxP13iygy2LHnD6OElREREdV9TgeGvn37ln7cq1cvSYqRklqthqnIVKURBplMwFN9/fDBei5+JCIi7+Z0YLhxWL4+DtGrVCqYi8wOW0OXZ+SDWmz4RYcr1+vPQVtERESu5rAPw41u7PhYHw+hkslkkFvk0JmLqvS+kAA5+v5Dg+FTfoLy2ja0i4vBzKnP1Mu/AyIioupyOjDUx1GFmymhhM6sq/r7rv6EzcfbArI7sCUlDUAyZk0b7/oCiYiI6iiv2SUBAD52JYosVQ8MF88cBGQiAMCkScCBg+dcXRoREVGd5nRg8IQheJWggq4agaF9fAyU+iMAAKX+ENrFxbi6NCIiojqtWlMS9XV6QiWoqrRLosTMqc8ASMYXe67DGNgCr07p7friiIiI6jCnRxhSUlJKP964caMkxUhNLRdhcLI19I0EQcCsaePx+zdDEdEwEh+ksPsjERF5F6cDwz333FP6cWJioiTFSE2UizDYqt+ESS4X8PHkUMz+5BpOXjC7sDIiIqK6rUqLHjMzM3Hs2DGpapGcWqaG0V6zro23RysxdXgAxryZC5utfk7NEBERVVWVAsO4cePQu3f9nb9XK9WwwAKbvWYnUD47yB8A8N43+a4oi4iIqM5zOjBcunQJu3btglarxfbt26WsSTJqUQ2lXQm9terrGG4klwv46KVQzF17HcczOTVBRESez+nAsGbNGgwaNAhjxozBypUrpaxJMqIoQglltXZK3Kx5pBIzRgRi9MJcWK2cmiAiIs/mdGBYuXIlnnjiCQwfPhzr169HUVHVWizXBaIoQmlTuCQwAMC4/n5QKQUs/h+nJoiIyLM5FRj27t0Lo9GIxMREhIaG4t5778UXX3whdW0up1aroXBhYJDJBKx4MQT//uI6jpw1ueSZREREdZFTgWHVqlUYOXJk6esRI0Zg1apVUtUkGVEUIbfJXRYYAKBpQyVeeyIIoxfmwsKpCSIi8lCVBgaDwYD//ve/GDFiROm1fv364dChQzh16pSkxbmaKIqQW+U1XvR4syf7aBHgK0PPcT+i79ApmDV3Sb3thklERFSeSltDWywWfP/992jcuHHpNR8fH2zbtg1arVbS4lxNFEXILDKXjjAAxZ0gW4nf4b0/uwBCS55oSUREHqfSEQatVotOnTrdcv3OO+9EeHi4JEVJRRRFCGbB5YEBAE4cPw4ISgA80ZKIiDyPVx1vLWVgaBcXAx9dWvELuxlBkW1d/jWIiIjcxenTKj2BKIoQTJAkMJScaHng4FI0aByPjX8m4tudRRiY6Ovyr0VERFTbvCowqNVq2Avt0FtcHxhKTrQskXbCiP6vXIbeaMew++vXWg8iIqKbeVVgEEURdqNdkhGGmyW0UOHHhQ3Q++Vs6Ix2jOntJ/nXJCIikopTaxjeeustpKenAwD27NmDmJgYNGvWDL/88oukxblacWCwQW9z7bbKirRu4oPNiyIwd+01vPs1u0ESEVH95VRgWLx4MZo2bQoAmDp1KiZNmoRXXnkFzz//vKTFuZooirAZbLUywlCieaQSW95qiOR1+Zj32bVa+7pERESu5NSUxPXr1xEQEICCggJkZGQgNTUVcrkcL7zwgtT1uZQoirDoLJKsYXCkcQMFtr4VgR6Ts7Fx8wH4FW5C+/gYzJz6DARBqNVaiIiIqsOpwBAdHY3du3fj8OHD6NKlC+RyOfLz8yGT1a9dmaIowqq31uoIQ4mGIQr0jP4Oi39OhF02DltS0sHmTkREVF849RP/zTffxMMPP4w33ngDr776KgAgJSUFHTt2dPi+vLw8DBo0CL6+vmjcuDE+++yzCu9NS0tDly5doNVq0aBBA7zzzjtV+DacI4oiLEVmtwQGADhy9E/YZRoAgFlzF7amFbqlDiIioqpyaoShd+/eyMrKKnPt4YcfRocOHRy+b/z48fDx8UF2djbS09PRp08fxMfHo3Xr1mXuy83NRc+ePbF48WI8/PDDMJlMyMzMrOK3Ujm1Wg2zzn2BoV1cDLakpMGkSYDceAa/FQ3Eiu8KMKa3llMTRERUpwn2ap6SZDQaodFoYLVay/18UVERgoKCcOjQIcTGxgIAHn/8cURGRmL+/Pll7p02bRrOnz+PTz/9tGrFC0KVDnk6f/48Oj3xD3Sa3Rlf3VP7x3Pb7XbMnpeMAwfPoV1cDIY8+iSGz8lBqyY+WPp8CPx969cUDxEReY8a9WFw9MP6+PHjkMvlpWEBAOLj47F9+/Zb7t2zZw/uvPNOdO7cGSdPnkTHjh2xZMkSxMTE3HJvamoqUlNTq1WvKIowFZpcflqls25u7gQAu99viBeW5qHDM1n4bHoYElqo3FIbERGRIzUKDI6G0QsLCxEQEFDmWslOi5tlZmYiLS0NmzZtwp133onJkydj2LBh2LVr1y33JiUlISkpCQCwYMGCKtUriiKMBUa3TUmUR62SIfn5UHyxtQh9pmbjlUcDkHNkDX79vXgUgjspiIioLpCs06NWq0V+ftlmRfn5+fDzu7XjoVqtxqBBg9C+fXsAwMyZMxEaGlq6ndNVRFGEId9QpwJDiUe6+SIh1gddJxxB9tXOsCv68JhsIiKqMxwGhujo6Ap/u61s7UBsbCwsFgtOnDiBFi1aAAAyMjJuWfAIAHFxcWW+TsnH1VxeUSGFQlHc6bEOBgaguMlTnM9/8JNiLICSY7KXurkqIiKiSgLDmjVrqv1gX19fDB48GDNmzMCKFSuQnp6OdevWYffu3bfc+8QTT+Chhx7CxIkT0bp1a7z++utITExEYGBgtb9+eQRBgA9U0NVy46aq6BAfhW1/7aSAzYDmLW53d0lERESOA8N9991Xo4cnJydj9OjRCA8PR0hICJYuXYrWrVtjx44d6NWrFwoLi/sQdO/eHXPnzkWfPn2g0+mQmJjosGdDTajlYp0dYQDKHpNtC+qCDacewPhMM1pEKd1dGhEReTGH2yoHDBiAdevWlb6eOXMmZs+eXfq6ffv22L9/v7QVOlDVbZUAEBkVCcsSAVn9z0IuyCWqzHU++r4Asz+5ho0LGqB1Ex93l0NERF7KYWDw9/cvs3AxODgYeXl5pa/9/PzK3fVQW6oTGJo3bw7TW3b80ft3+Co0ElXmWp9tLsRLy/KwYW4DbrskIiK3qFKnoJt/ONfH7X6iKEIlqOr0tMTNht+vxZLnQtBnajZ2H3ZPDwkiIvJuVQoM9TEg3EwURfgIPvUqMADAwERfrJwcisEzLmPLb/WrdiIiqv8cLno0m81YuXJl6ciC0WjExx9/XPp5i8UibXUSEEURVnvdat7krJ4dNPjvjHAMff0yusfsxfWsA2zuREREtcLhGoauXbtW+oNo69atLi/KWdVZw5CUlIT8iTos77IM8YFxElUmrX9NWYtVBzoAghI+ujRM7nuRzZ2IiEhSDkcYtm3bVktl1B61Wg2d3VQvRxhKXDp7EBDuAcDmTkREVDucag199epV7Nu3D3l5eQgJCUH79u0RFBQkdW2SEEURClsRdPU4MNx4TDbsZoRF3+XukoiIyMNVGhhef/11zJ07FxaLBaGhocjJyYFSqcTLL7+MmTNn1kaNLiWKIuRWOQxuOrHSFW5s7hQSdTd+OH0P9h81on1LbrkkIiJpOAwMX3zxBd577z2sWbMGAwYMgEKhgMViwbfffosJEyagZcuWGDJkSG3V6hLFgUFWr6ckbj4me8MvOgx8NRs/LoxAm6Zs7kRERK7ncFvl8uXL8dZbb+Ghhx6CQlGcLRQKBR5++GEsWrQIH374Ya0U6UqiKEJmrt+B4Wb9/qHBonHB6P1yNk5eMLu7HCIi8kAOA0N6ejp69+5d7ud69+6NjIwMSYqSkiiKECwC9HX4AKrqGNpdi1cfD0TPKdnIzKl/212JiKhucxgYjEYjgoODy/1cUFAQTCaTJEVJSRRFCGbBo0YYSjzZ1w9P9/NDzynZyLlmdXc5RETkQRyuYbDb7Th9+nSFvQ6q2gOhLlCr1YDJXq93STjy4pAAXC+yoffL2dj07wYI1Nb9A7aIiKjucxgYioqK0Lx58wqnrDETAAAgAElEQVSDQX3sLiiKIuxGwGCrv7skKvPaE4HI19lw18gMtFR8hU7xjdgNkoiIasThlITNZoPVaoXNZiv3j9Va/4a9iwODzePWMNxIEAQEXv8Psq7YkXrlSSxIaYTZ85LdXRYREdVjVTp8yhMUBwa7R65huNGvv5+DTRECADBr7sae9IturoiIiOozh1MS3bp1cziMLQgCNm/e7PKipCSKIqw6i8cHhhu7QcrMOThieQhXC6wI8uOaBiIiqjqHgeGxxx4r9/qFCxfw7rvvQqfTSVKUlERRhFVv9fjAcGM3yLZ3xqAgaCiSXryEjfMjEB7E0EBERFXjMDCMGTOmzOsrV65g3rx5WL58OYYMGYIZM2ZIWpwURFGERe/5Iww3d4O02+2Yueoaur9wCT8ubIDIUKeOESEiIgLg5BqG/Px8vPrqq2jevDmys7ORlpaGDz/8EFFRUVLX53JqtRoWncVjt1VWRBAEvPZEEB5/wBfd/u8STl9kR0giInKew8Cg1+sxb948NGvWDH/88Qd27tyJTz/9FLfddltt1edyoijCXGSu14dP1cSUYYGYONgf3SddwrHzDA1EROQch+PSTZs2hdVqxeTJk9GuXTtkZ2cjOzu7zD3du3eXtEBXE0URpiKTx09JODJhkD98RQFJL17Cd/MaIK4ZD6wiIiLHHAYGURQhCAKWLl1a7ucFQcCpU6ckKUwqoijCVODdgQEAnujlB7VKwL3jTyNOswH3t1WzuRMREVXIYWA4c+ZMLZVRe0RRhLHAAFi56O/onk9gLGyJX8wPY//3JwEkl1koSUREVMIrGzcZ8o1eP8IAAAcOnoNFVbwexSI2x9d7/dxcERER1VUOf82+9957Kx2i/vnnn11akNREUYSx0ADYFLDarZAL3tuT4MbmTkr9IWRrEjAp+QreHBsMuZxTE0RE9DeHgeFf//pXbdVRa9RqNQwGAwLlodBbDdAqfN1dktvc2NypXVwMJk7oiaGv52LQjMtY+0oY/DReNwBFREQVEOz18YzqvwiCUOUjtu12O+RyOW7/rjV2dN+GUFWIRNXVT2aLHc++ewX7jhqxbk4DRIdzrQcREXnhGgZBEODj4wNRJnIdQzmUCgFL/y8Ejz2gxT3PXsT+o0Z3l0RERHWAV/76KIoiVAwMFRIEAZP+GYDmjZTo90o2kp8LwcGdq3Dg4Dm0i4vh9ksiIi/kvYEBPgwMleh/jwZR4Q1w//OnoCu4C1afPtiSkgZuvyQi8j4VTkl06tSp9OPZs2fXSjG1RRRF+DAwOCWhhQptNV/C6hMJADBpEnDg4Dk3V0VERLWtwsBw/PhxGAzF5y0sWrSo1gqqDaIoQgml1x1AVV2Jd4dAqf8dAKAwnkK7uBg3V0RERLWtwimJAQMGIDY2Fk2aNIFer0eXLl3Kva++9WEAirdWKu1K6C0MDM4o2X65Oe0oMor6o9egzu4uiYiIalmFgWHlypXYuXMnzpw5g/3792PMmDG1WZekRFGE0qaAweadJ1ZWlSAImDVtPGYB2LBbh4dn5WDnuw3RuIFXLoEhIvJKDv/FT0xMRGJiIkwmE0aOHFlbNUlOFEXIbXKuYaiGfp01OH3Jgv6vZOPntxsiQOt1O3OJiLySU//ajx49Glu3bsXo0aPRo0cPjB49Glu2bJG6NsmIogiFVc4piWp6dpAf7osXMeT1yzBb6m3fLyIiqgKnAsOKFSswZMgQREREYPDgwWjYsCGGDx+O5cuXS12fJERRhMwi4whDNQmCgLeeCYZCLmDie1eq3G2TiIjqH6daQ8fGxuLLL79EfHx86bWDBw/ioYcewokTJyQt0JHqtIYGgGHDhkE5WESrNq3w8h0vSVCZdyjQ2XDf8xfxaJIWLzwS4O5yiIhIQk6NMFy5cgWtWrUqc+32229HXl6eJEVJTRRFCGaBIww15KeRYd2cBnj363x8u7PI3eUQEZGEnAoMiYmJmDRpEnQ6HQCgqKgIL730Ejp3rp/b69RqNQQzGBhcIDpcga9fC8e4t6/gqZfXoO/QKZg1dwmnKYiIPIxTgWHZsmU4ePAgAgIC0KBBAwQGBiIjIwMffPCB1PVJQhRFwAQYrNxW6QptY1XoGr0PH+/rgB9yx2HexlYYPfk/OHnBDL3RVnqf3W7HrLlLGCqIiOohpzbSN2zYENu3b0dmZiaysrLQqFEjREVFSV2bZERRhN1o5wiDCxVc3A/I2gEArKqmSPldxM4p2ci6YoFGlCEyRA5d/gWcudwRNkUfbE45CJ5JQURUf1RpE31UVBQ6dOhQr8MCwMAghXZxMfDRpQEAfHRpGN8lDSfWRKHw+8b4Y1UkPpkahiDLftgUoQAAsyYOS35OwOL/Xcfpi2Z3lk5ERE7wylZ9oijCprdCbzW6uxSPUdI++sDBpaVHYAPFO1lCA+QIDZCjVzvg95Q0mDQJUOp+Q49OZhw92wKd/3MRkSEKDLxXg4H3qPHlZyvw6+88SpuIqC5xaltlXVXdbZXvvvsufrm8B9ceLMDGLhskqIzKY7fbMXteMg4cLBsGrFY7dh024tudRVj1XTby9TJAJkKpS8eUvhc4bUFEVAd47QiDVW/llEQtKzmT4mZyuYAucSK6xIk4tn0hfjSOAwCYNXfh6735mGa2w0fJUQYiIndyOjBcv34dx44dQ2FhYZnr3bt3d3lRUlOr1bDoLNwlUQe1j4/B1r+mLRT6P2AMbok2oy9g1qhADO3mC5mMwYGIyB2cCgyrVq3C+PHjodVqodFoSq8LgoBTp05JVpxURFGERWfmCEMddOtaiJ7YnmHAKx9dw7//ex2vjw5C745qrmsgIqplTq1hiIyMxIoVK9CrV6/aqMlp1V3DsGHDBiR/kow/R5/D8d6HJaiMXM1ut2PDbj2mf3wVQX4y3OazFZfPZXBhJBFRLXFqW6XFYsGDDz4odS21RhRFmApNHGGoRwRBQP97NPjtw0YItu7F6gNt8UPuOMz/rhlmz0t2d3lERB7PqcAwZcoUzJkzBzabrfKb6wFRFGEoNMBkM8Fqt7q7HKoCuVyAOecXQKYCAFjUd+CjXS1wKou9HIiIpORUYFi8eDHmzJkDPz8/xMTElPlTH4miCKPBCLVcDT0XPtY7NzaJUurS0aKRHf+YcBET37uCS3kWN1dHROSZnFr0uGbNGqnrqFWiKMJgMEAtF6G36qFV+Lq7JKqC8ppE5V63Yf7n1xE3JgtP9fPDi4/4I1Ard3epREQeQ9LGTXl5eRgzZgx++uknhIaGYt68eRg+fHiF95tMJsTFxaGwsBCZmZmVPr+6ix5PnjyJnj17wj85GOsSv0a0pn63uqa/ncu24LXV1/DdXh3iQg5Ddm0HOsZHcmEkEVENOTUlYTabMXPmTDRr1gyiKKJZs2aYOXMmTCaTw/eNHz8ePj4+yM7Oxtq1azFu3DgcPlzxroQ333wT4eHhVfsOquHvEQY1Fz56mJgGCqx4KRSDb9+KLccaYtOVpzD3xwS8OOsjd5dGRFSvORUYJk+ejNTUVCxbtgwZGRlYtmwZtmzZgilTplT4nqKiInz11Vd4/fXXodVqkZiYiP79++PTTz8t9/7Tp09jzZo1mDp1avW+kypgYPB8Z/88ArsiEABgUzbEkt334rE3cvDLEQOP1SYiqgan1jB8+eWXyMjIQEhICADg9ttvR0JCAuLj47F48eJy33P8+HHI5XLExsaWXouPj8f27dvLvf/ZZ5/F3LlzoVarHdaSmpqK1NRUZ8quEAOD52sXF4Mtf3WM9NGl4dkeOWh4+1CMnJeLID8Znhngh0e6+mLBoqW3nG1BRES3ciowVPQbmaPf1AoLCxEQEFDmWkBAAAoKCm6595tvvoHFYsGgQYOwbds2h7UkJSUhKSkJALBgwYJKKi+fKIrQ6/VQy9XQMTB4pPIWRgqCgImD/bFxnx5Lvs3Hs29nwaBrC5uyD7akpAFI5kFXREQVcCow/POf/0S/fv0wc+ZMxMTE4OzZs5gzZw4eeeSRCt+j1WqRn59f5lp+fj78/PzKXCsqKsLkyZPx/fffV6P86lEoFBAEAaJMBb2FgcETVXTQlUwmoE8nDfp00qDbPz/GDnPxIlyTJgEHDi6t7TKJiOoNp9YwLFy4EElJSRg/fjzatm2LCRMmoFu3bg5/w4+NjYXFYsGJEydKr2VkZKB169Zl7jtx4gTOnDmDe++9FxERERg8eDAuXryIiIgInDlzpnrflRNEUYQSSk5JeLH77g4o7ecAuxmaBh3dWxARUR0m6bbKoUOHQhAErFixAunp6ejduzd2795dJjRYLBbk5uaWvt69ezcmTJiAtLQ0hIWFQS6veC99dbdVAkBYWBgeXj8E9zTqjMcaV7zVkzyX3W7H7HnJOHDwHKKbtUbKyW6YMyYQI3v4Vf5mIiIvU+GUxM8//4wuXboAALZs2VLhAxwdb52cnIzRo0cjPDwcISEhWLp0KVq3bo0dO3agV69eKCwshEKhQEREROl7goODIZPJylyTgiiKUNgVnJLwYjdPWzx3zoSeU7KhM9gxboC/GysjIqp7KhxhaNOmDQ4dOgQAaNq0aflvdvPx1jUZYWjRogX6rRyAJuFNMDF2gosro/rq9EUzHnwpG2P7+eHFIQGVv4GIyEtUOMJQEhaA4h4JnkYURchtCu6SoDKaNlRi6+II9Jh8CYUGG2aOCORWSyIiOLnoccCAAeVeHzx4sEuLqU2iKEJukXHRI90iKkyBLW81xPpdOkz+8CobPRERwcltlVu3bi33emU9E+oyURQhY2CgCjQIkiP13xHoPTUb49+5grCi/+LX39ngiYi8l8PAMGPGDADFh0KVfFzi1KlTaNy4sXSVSUwURQgWgYGBKhTsL8dPCyMQP+I3XLjSCTYFGzwRkfdyGBjOnz8PALDZbKUfA8WLDaOjozFr1ixJi5OSWq2GYBZgsBrcXQrVYf6+MrRUfI3ziicBFDd42pb2uZurIiKqfQ4Dw8qVK2Gz2dC5c2eMGDECKpWqtuqSnCiKEMzgCANVqmN8I2z/61wKmfkiftMNQs8plzB+gD96d1RDLuf0BBF5vkoXPcpkMkyaNMmjwgJQHBhgZGCgys2c+gwm972InqFLMa1HGrK/icVjSVrM/ewaWo68gEVfXMeV6xbMmrsEfYdOway5S7hQkog8jlOLHrt06YI9e/agU6dOUtdTa0RRhN1k57ZKqlR551I89oAWjz2gxb6jRiR/m4/GQy7DZOgIm6IPNnOdAxF5IKcCQ+PGjdGrVy8MGDAA0dHRZVaIv/baa5IVJyVRFGHT2zjCQDXSoaUKHV4Ow4W0JdhqGQUAMGsSsGrXEXTcp0P3u9RQ+XDKgojqP6cCg16vx8CBAwEAmZmZpdfr89YyURRhMzAwkGvcc1cYdv21zkGpP4SWzWWYu/Y6Hp+bi57t1Rh0rwYPthOx6O1lOHCQ2zOJqP6R9PApqdWkNfSMGTOgF/X4se0WHOzxq4srI29z40FWN4aBi1csWL9bh2936rA9vQBmYxHsikAodWmY0vcipy2IqN5wOjCcOHECn3/+OS5cuIDIyEgMGzYMLVq0kLo+h2oSGObNm4ecohykdP4Rx3sfdnFlRLfq8cir2Jw3pvR1C81ebP/kYYQHVXwiKxFRXeFUa+gNGzagbdu2OHr0KIKDg3Hs2DG0a9cO69evl7o+yYiiCIvewikJqjX/uCsCPro0AIBC/wd8Axqi1RMX8MSCHOw/anRzdUREjjm1hmHatGlYt24dunXrVnpt27ZtmDBhAvr37y9ZcVISRREmnQkmmwlWuxVygb/lkbRmTn0GQDIOHFxaOm2Rl2/Dyh8KMWxODsIDZRg/0A9/7FmD9ENnuM6BiOoUp6YkgoKCkJOTA4Xi73xhsVgQGhqKa9euSVqgIzWZkli5ciV+/vlnbH1oJw71TIdW4evi6oicZ7Xa8f1ePSYtPozTeYGAoITceAZP3XcG774+kqGBiNzOqSmJu+66C4sWLSpz7a233sJdd90lSVG1QRRF6PV6aORqTkuQ28nlAvp11uB2+f8AQQkAsKqaYOW+tggffB69plzCjJVXkbJHh+w8Nokiotrn1JTE0qVL0a9fP7zzzjuIjo7GuXPnoNVq6/0aBoPBADUDA9Uh7eJisOWv7Zk+ujS82Pcinn56LPYfM2H/USPe/yYfOzMKYDAlATIVUjeegsn6Aea++rS7SyciD+f0LgmLxYI9e/YgKysLjRo1QseOHaFUKqWuz6GaTEls3LgR77zzDgpe0uPj9stxu3+si6sjqrqKtmfeqM+QKfjxyrjS1wrBiD6dgzDwHg36/kONID+uxyEi13NqhAEo/uF84x+ZzKnZjDpLrVaXjjAYbDyxkuqG8tpQ36x9fAy23jAK8WyPHNzZ+VF8s1OHie9fQYeWKgxM1ODkr1/g2LETXDxJRC7hVGA4ePAgBg4cCKPRiMjISGRmZkIURXzzzTeIj4+XukZJlExJhMmDOCVB9Up5uy0EQcBjD2hRpLfhpwN6zF6WjkOXugJCEjb9kIk83Sq8O+cJd5dORPWYU1MS7dq1w7BhwzBp0qTSaYDFixdj7dq1+PVX93VJrMmURHp6OkaNGoXY5FZ4uvlT6B7e1bXFEblR36FT8EPu39MWPoIOtzcJwMBEDQYm+iKumZIjDkRUJU7NKxw/fhzPP/986T8wgiDgueeew4kTJyQtTkplFj1aOMJAnqVdXExpkygfXRpe7LoV7z8XggK9HQ/PvIzbR1zAS8vyMGbKZ+gzhLstiKhyTk1J9O7dG+vXr8egQYNKr23YsAF9+vSRrDCplWyrVCu4S4I8T0XTFve0EfHm2CBk/GnC/83fjp2n7oZd1hk/bSrClwf34v57WiAyVI7IUAUalfw3RMZDs4jIucBgtVoxdOhQtG3bFtHR0Th//jx+/fVXDBgwACNGjCi9b/Xq1ZIV6mqlIwwykYGBPI6jxZOCIOCu5ir4Fm6GXVY8bWGX+0Jt3Inmka1wIdeCQ6f1yMy1ICvXirOXjLBYewCCAqnfHQGQzEOziLyQU4GhTZs2aNOmTenrVq1aoUePHpIVVRtunJIwWLlLgrzPzT0f+na3YOJg/1vuu3Ebp0XdCsk7/HHXziL0/YcGCjlHGoi8hdceb200GuHv74/paTMRoPTHxNgJLq6OqG5zpucDAMyauwQLUxrCpEmAUvcb+rQDLqETsnIteLq/P0b30iLEX+bUs4io/nI6MGzduhWffvpp6fHWjz32GLp37y51fQ7VJDDY7XbI5XIsOPJvWO1WvHzHSy6ujsgzVBQsfj1uRPK6AqzbpUNjv7M4fKoQFvUd8NGlYXLfi5y2IPIwTu2SWLFiBYYMGYKIiAgMHjwYDRs2xPDhw7F8+XKp65OMIAhQqVRQ2pVcw0DkQMl6iJT/LMCsaeNLRw7axqrw0Uuh+GNVJPTXL8CivgMAYNIk4Is9Idjymx75RbbS59jtdp6BQVSPObWGYeHChdi0aVOZJk1DhgzBQw89hCeffFKy4qQmiiIUNgUDA1ENhAXK8UinK1iQkgazJgFy42kERYdj1qpryPjThMYNFGjfUoXLZ37BpvRmsKj7YEtKGrh4kqh+cSowXLlyBa1atSpz7fbbb0deXp4kRdUWURQht8oYGIhqqKJtnGaLHYdOm7D/mAlvbFOUGYX4dHcG7thaiPa3q9C0oaJ0ipFrIYjqJqcCQ2JiIiZNmoQFCxZAo9GgqKgIU6dORefOnaWuT1KiKEJmYWAgqqmKtnEqFQLubqHC3S1UyDp4CgtT9DBpEqAwHMdtt/ngf9t1mPLhVRhMdrS/XYWinAzs/uN2WMQ+2FzBKARDBZF7OBUYli1bhmHDhiEgIADBwcHIy8tD586d8fnnn0tdn6REUYTcIuPhU0S1oKJRCADIyrVg/zEjXppfCIuYAAAwaxLw5rYi/JCZ9XcTqVA59v2yDRt/i4VF5NQGUW2qNDDY7Xbo9Xqkpqbi0qVLpcdbR0VF1UZ9klKr1RAsMujtHGEgkpqjZlKNQhUYEKrAb/+4iIV/9YZQ6tIw9oE8DB32OC7kWHAh14oLuRb8ekoNi9gCQPHUxo8H1mGm3c5RBiKJVRoYBEHAnXfeiYKCAkRFRXlEUCghiiIEM6ATGBiI6oIKRyFaqkrvUeWcxMKUIpg0CZCbzuGU5X4kPJWFZwb4Y/j9vvBVO7X5i4iqyKkpibvvvhvHjx9Hy5Ytpa6nVomiCMEE6BUMDER1gaNRiBI3h4oZL4/Dlt+MSF6Xj+kfX8XjD2oxrr8Wn65a7nCdA9dCEFWNU4Gha9eu6NmzJ0aNGoXo6Ogy/1ONHj1asuKkJooi7CZAzxEGonqjvFCR1FaNpLZqnLlkxtL1Bbh7zGnoDYmwywOQuvFPHM/5H4b9s+xheZ9/+R2+PtASFlXFCyyJ6G9OBYZdu3ahadOm2L59e5nrgiDU/8BgtEOv4KJHIk/QJEKJBU8F4+DmRdhkLu4RY1Hdhq3H5NBtLChz795jYbComgAoXmC5L2NZbZdLVK84FRi2bt0qdR1uIYoi1qWsQ07vHMx++zXMeO5VDkkSeYCO8Y2w/YaDtcZ2v4hZ07qWuWfW3P9hYUoeTJoECNbrOGkfjOuFNgRouQaCqDwOz5LQ6XSYM2cODh06hISEBEydOhUqlaqi22tdTc6SAIC7OybgVNNTUA3zhXACeEYxFjOfn+HCConIHZxZn3DjPQltGuOq/xBszzBg3ZxwNG2odFPlRHWXw8AwevRo7N+/H7169cL333+Pbt264b333qvN+hyqaWBoGtcUuf+4CnVvXwDAPza3x7p3v3VVeURUz7z/TT4W/Oc6vpgZhn+0Et1dDlGd4nDsbePGjfjpp5+wcOFCbNy4ESkpKbVVV61oGNIQyCoOHEIukNAswc0VEZE7TRjkjw8mhWDwjMv479ZCd5dDVKc4HGHw9/dHfn5+6euSLo91RU1HGKZNm4a9GXth7mBFwe1FSBuyj2sYiAgHT5kwcHo2nujlh+mPBfDfBSJUsujRYrFg69atpT+Ub34NAN27d5e2Qgmp1Wp0TuiMKa+8jDt/uBs6qx6+Co27yyIiN4tr5oPd7zfEoBmXcTzTjMaW/yH98Bn2ayCv5nCEoUmTJg7/xxAEAadOnZKkMGfUdIThzTffxOXLl/Hmm29i0K6H8cxtT+OBiCQXVkhE9ZneaEOHJ/bg6EV/2OVa+OjSMLnvRfZrIK/kcIThzJkztVSGe4iiCL2+uGlTt7Cu2JbzMwMDEZVSq2SIsa7DH/JxAIrPrtj522o3V0XkHl694VgURRgMxU2buobfh22Xt1fyDiLyNu3jY+CjSwMAyMyXsLdgKF76IA9XrlvdXBlR7fLqwKBWq0sDQ5uA1rhquopM3QU3V0VEdcnMqc9gct+L6Bm6FNN6/IqjnzaFTm9Hqycu4I0111Cot7m7RKJa4VSnR0914wiDTJChS3gXbMvZjscaD3dzZURUV5R3dsWS50Pw/MP+mPXJNbQceQEvD/NH9uG1+O3QWS6MJI/FwGD4+xyJrmFdkJq9mYGBiCrVIkqJta+E4bcTRgydnoE/c+4HZCps4UFW5KG8ekqivMCwI2cnrHbOTRKRc+5uoUIL4StAVtw236RJwM70XDdXReR6DAw3BIYIdQQixAikX81wY1VEVN+0iyu7MHJfwVAsTymAzVb9bd9EdY3XB4aSbZUluobfh2053C1BRM67eWHkjveb4OMfCnD/i5dw7LzZ3eURuYTDxk11XU0bNx06dAhDhgzB4cOHS69tvbwdbx17Gxvu/cYVJRKRl7Ja7UheX4A5n17Ds4P9MXlIAHyUXAhJ9ZdXL3q8cVtliU4hHXDo+mEUmAvgp/RzU2VEVN/J5QKeHeSPAZ01mPDuFbQfl4X2Idtw6exB7qSgeknSKYm8vDwMGjQIvr6+aNy4MT777LNy73vzzTfRpk0b+Pn5oWnTpnjzzTelLKvUzWsYAEAtV6NdcFvsyN1VKzUQkWeLaaDAujnhiPXdi1X72+OH3HF4Y1MiOo78GXPWXMPKjQX46YAeh8+YcLXAiplvLEHfoVMwa+6SGo2gErmapCMM48ePh4+PD7Kzs5Geno4+ffogPj4erVu3LnOf3W7H6tWrERcXhz///BMPPvggoqOjMXToUCnLKzcwAEC3v7o+9m7YU9KvT0TeQRAE6LP3ArIEAIBdHgDD9QwYTHHY8bsBF3KtuJBrxakLOpgsPQFBjh83FeDbw7vRs+sdiApVIDJUjkahCjQKkWH5h8uRxp4PVMskW8NQVFSEoKAgHDp0CLGxsQCAxx9/HJGRkZg/f77D906cOBF2ux3vvfeew/tquoahqKgI4eHhKCoqKnP9SP4fGLXvX9iXxFEGInKNWXOXYGFKQ5g0CRUeYtV36BT8kDuu9PWd2s145JGHkJVrxYVcC7JyrTh6Jh+FRhUgCBCsRYhtUID772mBRiFyRIUp0ChUjkYhcqxe9TEOHjnNUEEuI9kIw/HjxyGXy0vDAgDEx8dj+3bHOxDsdjt27NiBsWPHSlVaKZVKBYPBALvdXuZ/pjv8WkJn0eF00Rk09W0ieR1E5PlmTn0GQDIOHFxa+kP8Zu3iYrAlJa00VAzobsDU4YFl7uk79J3SUGGX+0Jt3Inmka1wIdeCw2f0uJBrwe8n8nBV9yAgyPDjpuvYfmoTXvxXIjq0VCEsUF78Xrsds+cl48DBcwwV5BTJAkNhYSECAgLKXAsICEBBQYHD982aNQs2mw1PPPFEuZ9PTU1FamqqS2pUKBSQyWQwm83w8fEpvS4IArqGd8G2y9vRtGkTl3wtIvJu5bWYvll1QkXf7hZMHOxf5p6+Q9/GD/qSUBGAC9kn8e7X+ThwzIhgPznat/RBXuYBbDvUHBaxj8PulAwWVEKywKDVapGfn1/mWn5+Pvz8Kt558P7772P16tXYsWMHVKaikhQAACAASURBVCpVufckJSUhKan4COoFCxbUuM6SdQw3BgYA6BbeFesvpOCJpiNr/DWIiJwhVagY3v0SZk2LgM1mx/FMM/YfNWHmHissYisAxd0p392eg9902YgKLZ7WKFkzse7rL/DxtkiYNY6DBXk+yQJDbGwsLBYLTpw4gRYtWgAAMjIyblnwWOLjjz/G/Pnz8fPPPyMqKkqqsm5RsrXS379sQu8Sdi+mZEyD2WaGUqastXqIiBypSaiQyQS0jPFByxgf/HngHBammGHSJECp/x0P33sdfXq1Kl2AuT3DgMxcK/YdjINZEwSgOFj8b+8uPHDIgIQWPlCrijfacRTCO0jauGno0KEQBAErVqxAeno6evfujd27d98SGtauXYsXXngBW7duxR133OH082u66BEAYmJisHPnTsTExNzyufu39cAbd76OTiEdavQ1iIjqGmd/yM+auwQLUhrCrEmA3HgGdzUDBN9YHD5rRstoJdq39EHWyV344TclLOpWFS7opPpP0sCQl5eH0aNHY9OmTQgJCcH8+fMxfPhw7NixA7169UJhYSEAoGnTpsjMzCwzDfHYY49h2bJljot3QWCIjY1FSkpKmcWZJV4/PBdKmRIv3/FSjb4GEVF9VVGwMJhsSD9pwv5jJixcsQcXjX//G9refx12fDkRCjlHGTyJV7eGBoC4uDisWbMGcXFxt3xuZ84uvPHHPGzsklKjr0FE5Mlu3DIqN2UhPEQDsywYfTtpMPBeDZISRKiUAqct6jmvbg0NVNy8CQDaB7fDsYITuGq6iiCfoFqujIiofihvzcS5y1as26XDoi+uY+S8HET6XsCx861gVXHxZH3FwFDOiZUlVHIVOoV0wI6cnegf2a+WKyMiqh/KW4jZuIECEwf7Y+Jgf1y+asUDT2yFVdUVQPHiyY92nUDQV9fR4Q4V7rqteAElF0/WbQwMDkYYAKBr2H3YlvMzAwMRUTWFB8kxqEMRTvy11VOpP4J2d5hxItOCzzYX4Y9zxQsooTuBjFOtOQpRR3l9YCjvxMobdQvvimV/fnhLN0giInJeedMWJf+m6o02pP9pwpNTz8OqSgRQPArxnz378PBpE1o3UfLf3zrA6wNDZSMMzbW3wQ77/7d33/FRVPv/x1+zLSGFBAgRogEu0qSIJCBfmoCgFBVQr2LBCoiKiu1SrkoofkH4qhcU4YrlBxcFuSqC96pIkdAhIUAogRASklDSIIFkU7ec3x8LQxaEBNhNKJ/n47EP2Nn3zpw5hN1PzpyZ4aA1maaBTaqwZUIIcf242PUjavgY6NTSl792PMX006MQppID1A4PYcA7WVjMGoO6+jGoix8dmluYPG2OHLaoBlIwVFAwaJrmOiyRvVYKBiGE8KILXXBqx8EyflpfxPAPT3Aks4DCwjtxmu9jtRy2qFI3/GmVI0aMICIi4qI3u1p29D+8v24KYZtvIqJxBONHvScVrRBCVIMef53KhpOP68/DfPbz2Tt30TvSV7/ypPCOG753KxphANjy42bSfNLZ3CuWOfbPmTRzchW1TgghRHk9ImpiKdoOgLl4N60bOJjxYz43P3KYwZOyWbTaykmrgwlTPuP+x8YwYcpnV/yLpXCRQxIXOa3yjISDCdDQ9XdnU9i+ensVtEwIIcS5LjR5Muekg/9uLuK7NYUMnZ6BrawryhjEql/24XTOYdK759+kS1waKRgqMcIQ0TiCLSmxqMagHXY9F0IIUfUuNHmybrCR5/oF8ly/QPo++imrcocBYK9xG9OimxCdk8Gdt/lwZ3MfOrSw0PAmI5M+qHjyZGWuDVGVmep0w89h+OCDDzh58iQffPDBBTNKKSbNnMzqU39wtFUG2wfEEGQJuqLtCiGE8I7yl6q2FG3ntb453PvA08TsKyU2sYyY/aWcyi+iuKQEZQzCWHaMrs0LuLt7x/PW9cfarWxIDMRhCbtgzhsZc1E8Y+4/clVN6JQRhkqMMGiaRtTr44liPH+LH8u43e8yO/LTKmqhEEKIS3GhwxZ3t6uhZ3o/Molo29MAOCxhHM6Mo6Ts/F9AD2dacViaXzTnjYzNry1z1tem/n/yGdDZj/p1qv/ruvpbUM0qUzCUN7FVFL3X9uGnI8t48JaBXmyZEEKIy3Gxaz6c0bVdHTadvuaDpWg7T9ydyYTnz79nkDEzU782xIVy3siYi3bS404HG3aX8s5XJ7mtoZlBXfwY2KUG38z/oloOW9zwhyTmzZtHdHQ08+bNq/R74k/u4vHNQ1jR/Tdu8bv5irYvhBCi6lV2vsDVMIehzKb4Y0cxyzYWsXDFCYpKHSiDH8ayIwyMyOL14fdxc4iJ+rWNWMya1+ZC3PAFw3fffcfSpUv57rvvLul9Mw98yprstfzYZTFGzXhFbRBCCCEq477BY/n9xIv685ssKTRs3IJjJxxk5TmoHWhAs+eRladQpmAsRdsZfX+GR+ZCyHUYKnFa5Z95penLOHEy++DnXmiVEEIIcb4ObcP161BYirYzvOs+Ns0KI3VRONZfGrLtn2H8xbASZQoGXPfk2LYr3SPbloLhEucwnGHUjHwW8QmzD85h18ndXmiZEEII4S5q3MuMvj+DviFzGH1/hn75bACjUaN+HRO9Imu4FRXtb2/gkW3f8JMeK7pb5cWE+93C5NYTeSnuFVZ2X46fqUbFbxJCCCEuU2UmdF7onhxXvO0bfQ7D1q1bee2119i6detlvV8pxYtxI6llqcUHt//vFbVFCCGEuFrd8CMMl3tI4gxN05jedio919xD1uYscvcelxtUCSGEuO5IwXCFBQNAkDmI9ikRLK37H+gFW5NiYSZEvT7eQ60UQgghqpdMevRAwQCQlZCpl1/OphCXEnfF6xRCCCGuFlIwXOZpleeKaByBIen0k2I43PkYhwpTr3i9QgghxNXghi4YlFL84x//IC8vjwkTJlzRBMrxo97jJdMIOq3uwEuZwxkeOZR+6+5nbvKXOJTDg60WQgghqt4NfZbEhAkTmDJlCjabDbPZzN///ncmTJjgsfalWFN4fefbOJSDGXd8RNPAJh5btxBCCFGVbuhJj7GxsdhsNgBsNhuxsbEeXX/jgMYs7fID/+/QfB7YMIiXb32RE78eZ2fKTjmTQgghxDXlhi4YOnTowKpVqygrKwOgTZs2Ht+GQTMwtPFz9L6pF4P+8zBH6x+Dv8iZFEIIIa4tN3TBEBUVBbhGGo4dO4bZbPbathr6NyB8UxhHex0DXGdSxER7dkRDCCGE8JYbeg5DeampqURGRrJv3z5CQ0M9ss5zTZwxiTn2z3E2BU6Bv78///if/2Ng2AA5NCGEEOKqJgVDOa+++ipms5mPP/7YY+ssTynFpJmT2Z6ynYjGEfQZ0ocxu9+hrk8IH7R5nyYyKVIIIcRVSgqGcjIzM2nVqhU7d+4kPDzcY+u9GLvTzpeHvmbGgU94uuEQXm82Sm5iJYQQ4qojBcM5xo0bx4kTJ5g7d65H11uRjOIMovZOIi5vB20OtSRvT66cSSGEEOKqIQXDOXJzc2nWrBmbN2+madOmHl13ZQz75wv8XOcXsICWA0NKHuejVz6s8nYIIYQQ5d3QV3r8M7Vr1+aNN97Qz6CoajkJ2WBx/V3VhUU3fU+v6D58nDiDxPwD1dImIYQQ4oY+rfJCRo0aRZMmTYiPj6dt27ZVuu2IxhFsTYrF2RQMSTDc9Bx9evbhl4zfGLz5CfxN/vSv349j0Uc5vD+dyMaRcthCCCGE18khiQuYOXMmq1ev5ueff/bK+i/k3DMpyhcDSil2noxn3M/vsN1vp2sk4iQ8UNKf2cNm4WP0qdK2CiGEuHFIwXABJSUlNGvWjMWLF9OpUyevbONyDXxtEJt7nb3oU+DJALQQA91Du3HvTffQ+6Ze1LbUumDhIYQQQlwqmcNwAb6+vkRFRfH3v//da0XJ5Sp/K21DEjyd9ySbe6+nd2gvfsv8nTtXdeaOf7dndsDnbO4Vy2z750yaObl6Gy2EEOKaJiMMF2G322nZsiWzZ8+md+/eXtvOpbrYYQuAEkcJ977fl/13JOnL6h4L4f0HJtGtbldCfOpUR7OFEEJcw6RgqMDixYv56KOP2Lp16zU1pF/+MtRaKnQN7IJfKz82Hd9CI/+GdK97F3fV7cbqRavYlbxLDlsIIYS4KCkYKuB0OomIiKB58+ZYrVY6dOhAVFTUVf/FeqFRCJvTRlzeDtbmrOPb+IVkWrLAANpxeKhsEJ+OmInJICfPCCGEcCcFQyU88cQTfPfddyilsFgsjBs3jgkTJnh9u9527uRJv4IaGGuZuLN2B7qEdKZrSGfaBLfGgEEmUAohxA1OfpWshJMnT+qFSVlZGTExMdXcIs8495oPz5me5pVHXmHTic1sPL6JV3e8TkZxJrXzg0kLSEf1gi2psThmOpj0+kS3dVU0r0IIIcS1TUYYKmHChAlMnTqVsrIyNE3jlltuYcmSJbRv397r2/amynzJZ5fk8OCHD5PUJllfpjk1bvYPo6F/Axr5NaKhfwO2rd3GypLVqAagJcMIwzApKoQQ4joiBUMlKKWYOHEisbGxREZG0qBBA9577z369OnDlClTCAsL83obqlP5CZSGJBhhGs7zw58ntTCVtKJ0UgvTWLhpEbmhefp7NKeGyWgiwORPgCkAf5M/eTl5ZDmywR84CXeq9jzS46/UsdQhxCeEuj51qG2uzcw5n7AjZccFi4rKFB5SnAghhGdJwXCZ8vPzmTp1KnPnzuWNN97gzTffZPr06cTGxl4zEyMrqzJfvucWFS+ZRjD21TFY7YUUOgqx2q288n+vsKf9fv09Nx+qz9097ian9DjHS49zouwERwuOUaaVgQaUQR1Vh6ZhTfTCI8AUwO5du4m374Y6oGVBb+1unrr/qXIZf+bO+4L5Zd+gmpxtT9Tr4y95v4QQQrhIwXCFDh06xOjRo1m5ciVWqxWHw3FdTYysrMstKs79Ej93ImabmJZMfmMSVrsVq91VeHz6/SzSmh3WM3UyaxPZNkJ/3Wq3cjj3CDYfm54x2ozUCaiDr9GXGkZffAw+5GTlkOHMBD8gD7o6O/Ni/xdo5N+IBn7h+Bp9K71vUnwIIa53UjB4SKdOndiyZYv+/M4772Tjxo2YTDKv9AxPFRWXmtEOwvPmZxg1/DWKnSWUOEoodZTyxj/eZG/k2RGPm46E0iqiFamFqRwtPkZtSy0a+TciLzWXxLIDqNMjGncbevJYv0fxNfria/DFx+jLgsUL+MG+BNVQ5nAIIa5PUjB4SPmJkUajkdDQUGw2GwMHDuThhx+mV69emM1mfS7E9XbYwlM89dv8lRYnDuXgWHEGaYVpvP3laFJapurvC8moQ6fI/6HEUUKJo5QSZwkJqQkU1iw6u3In+Jh8CDD543/6MEledh4ZjkzXHI486Kg68GjPRwjxCSHEp45rLoelDh/P/scVz+EQQghPk4LBQ8pPjDxTDKSlpbFkyRJ+/PFH9u3bR3h4OAkJCdjt9hvysMXVprJfvJcz6vGi6QXGvDra7TDJax+NchvRCEutR4+7eujzN3JKT5BhzcCGzXWXlzKoq0JoGtbUbfKoaw7HLteIRyb00/rw4sMjCDIHEWwOItgSjK/BVw6jCCE8SgqGKnL06FHuvfdeEhIS9GVhYWG89dZbREZG0q5dO2rWrPmnhYd8iFevqhjROOPP5nBMfH2CW+Ex68fPSG96RM/UygmiafNmnLKdIq/sJKdsp7Db7TicDteVVoqhkWpAh2YdqGUJJsgcTLA5iDVr1rC6bA3qZtflw5/VnmLyK5OwGCxe2X+ZCyLEtU0KhipU/rCFyWSif//+hIeHExcXx65duwgPD8disZCQkIDD4cBsNvP2228zZcoUt/VIUXFtqso5HPe/PoCYnnH68+bxTRj57Ei3omLZpmUcr5+rZ8xlJpQPWAxmgs3BBFlchUVmeiaHHKlQEzgJbbU2dInsgkkzYdKMmAxm1m5aS4wzFhUCWjbcZejKgF4DMBtMGDUTJs3Ekl9/YoVjJSoMtKMwwHA/z//1OYwGI2bNjEkz8cW3X7LY/j2qEWiH4GnjEP42/G1MpzNGzYRRMzD10w+q7NRbKXSEcJGCoQpd7IvebreTkJDAkCFD2L17t/4eg8FArVq1aNCggf7Yv38/a9aswW63YzabGT16NO+//36ltyWuXlU5mvFnmfGj3qPQUcQp20lOlp3ipO0kY+aM48DtB/X3NToQztODnsGh7NiVHZvTzuIViznaKEPP3HSkLr0798au7NiVA4eys27HerdrddTMDaRl05bYnDYcyoHNaSPl2CGKA4r1jLnETFBQEHana1sOZafUXoZTc7oCTjBhws/ih8lgxKSZMWlGCgqsFKgCsAClUJcQwm8Kx2gwni50TKSlp5KujrjmlBRAU+1W7mhxByaDSc+YDCZidsQS79gFtV33XOmidWbQ3QPxMfhQw+iLr9GXfy/9nv/Yf0GFg5YGQwyPM3rY3/SJsWdGba6mkZqqLpiuxYxwJwXDVab8KITFYmHs2LGMHDmS9PR00tLSSE9P55NPPiE1NVV/j6ZpBAcHU69ePerXr0/9+vVJSUkhJiYGh8OByWRiyJAhvP766wQFBVGzZk1q1qyJ0WisVFFRmeJDCpSriyc/ML1x5sqVZM49bHNndATfTv8Gm9OuFzHPRw1lR5ezhXebmJZMf2saduXArmzYnXbemx3F/nYH9EyTPY0Z9dRrruJF2bA7XYXOFz99QVq5Q0D1DofSu2vv05NeSyh1lrJ1Xwz5tQv0jE+xhZrBQXrGruwYnUbsyg5GwAaBBFI3OASTZsJocI3WZGVlk62yoQZQBA24heZ/ae4qhE6Psuzdv5dEZxIEAaegpdaCyDaRbn0UtzuOBLW/8pmT0EZrTeeITqdHe4x6m9ZuWsdWRwyq7pnRo24M6j1QL6hMmpElv/zEcscKffSov7EvTwx83G17C5ct4lfHctchsCrOlB/NMmkmzJqZL779gsX2H8qNZj3J28Pfxny6YDSeGUHTTLz/yf9eNUVMdRZDXj3nLzc3l6FDh7JixQpCQkKYOnUqTzzxxHk5pRRjx47lyy+/BGDo0KFMmzbthvzCiYqKAjjvizc0NFS/FPWpU6fOKypeffVVMjMzycjIICMjgylTpuBwOADX6MWyZcuIi4vj1KlTnDp1ioKCAoxGIzab61oFv/32G/Pnz6dNmzYEBAQQGBhIQEAAAQEBbNq0iTVr1uBwOFixYgVJSUmMGDECPz8//P398fPz47PPPmPGjBnYbDZWrVoFcN6ETk8VHlWZuZTc1UTTtPO+aC8nAzB+1HswE7avPvvhU52Zc++B0qFhB4LMQW6ZzmGdiU/arWfuqtuNyNoRbplewT05kHRAz/Txv4fBDR45b3sZpmPMSTpbxDxsepCoO84pdLZMYs6Js5lhpueIeuxsxqEcDHxjEDE9t7sWmKHJxkbMHv8ZdqerQHEoB298+CbZHbJdGT/w2+HH0z2HYHeeLnSUg10r4qHV6RUHQcm+Ytp2u92tPRuPboDbLiETDAWJ+dT3recaEXK6RobKnGWk56ejGrliKhT2Hd1PWG4YdqddL662aDGo0xe8VTfDxuObKU0tc9veNlMcql71ZNbkriVrf44+UmVXdpKDU1ABpzN/gUUl/+aX6N/1fXecHj2zKRs0Av4Cm1Uss5fN/dPPEtVIuTLOWOYu/ZoAH//TRZcJs2biVH4++TfnuzKlsSz84d+E1qnrtp7sEznkheW5MiWxfP/9EsLrhbvWcfrw3qG0Q6TXPezKFMSyYuFqIm5rd7rwdBWV23ZuIz5oN/SCrUmxMJNK/V+viFcLhpEjR2KxWMjKymLnzp3cd999tG3bllatWrnl5s6dy9KlS4mPj0fTNO655x4aN27Miy++6M3mXZU0TavwzIkLFRUhISG0bt0agJSUFLei4rXXXnNbr1KKvn37smLFCv15aGgoQ4cOxWq1YrVaKSgowGq1kpyc7FZ8rFixgsOHD1NYWEhRURFFRUVkZGToxUdZWRnvv/8+n3zyCT4+PlgsFnx8fDh16hQ5OTkopfjtt99YsGABzZs3x2KxYDab9fkbu3btwul0snz5clauXEmPHj30181mM6tXr2bVqlU4HA5+//134uPjefTRRzGbzZhMJkwmE9999x2LFy/W25uens4LL7yAwWDQH3PnzuWrr77CbrezcuVKcnNzefPNNzEajZhMJv3P6dOn8/HHH2Oz2Vi5ciVOp5OJEye6fWhcbYWOp0eF1Ekn9mQbqrbzgj+XVZV577V3WftwNPt/TKRFvea89+O7Xst4al1GzUiHRu3Zsi8Gw20mnPvsdA7rxK0Bt7rluod0Y9e+3XqmV3AP+tS71y2zx28XB/bN1jP9fPvwdKMhbplDvsl8eomZ+3z78nKT8z9zy34vccs94vsgUe2i3DITN050yzzpO5io/zkns2Uin+ZUT+Yp38eJ6npOZsZEPj18NjPc9wWiBrtn4JwRLQ06ro7kp5lL3DIPjnqILb22uZ4Y4I7o1iyY9i+3Qm/YxOHs7Hx61MsHGmy5mU/HfeK2nlenvkbe/5w+bOcLN8WEMKl3VLlCx0HUyii44/QbAkHtddClWyds+mE7B6uyVkEzV8TZ1FWAe4LXCobCwkJ+/PFH9uzZQ0BAAF27dmXAgAEsWLCADz74wC07f/583nrrLW655RYA3nrrLb744osbsmCojMstKs5dR6dOnYiOjtaLin79+jFw4MA/XV/54mPkyJHnbb/8oZQzkzXffvttSktL9cewYcPIznb99qSUok6dOowcORKbzUZZWRllZWXs3r0bp9P1ReF0Ojlx4gQ1atTAZrNRUFCAzWYjMTFRL2AcDgcxMTH4+Phgt9ux2+3YbDZiYmKw2+2Aq8j56aefSEhIwOl04nA4cDqdJCcn6xmbzcZXX33Fzz//7DrDwOHQ/ywoKHDLTZ48mcmTJ6NpGkajEYPBgNPp1DO//vor06ZNw8/PTy9ONE2jsLAQq9WqZ2bMmEFwcLD+uqZp5OXlkZubq2dmz55NaGioWyYrK4usrCy98Pryyy8JDw9H0zQ9d/jwYdLT0/XMv/71L2699VZ9HZqmkZycTHJysp5ZtGgRt912m9t6DAYDCQkJ7Nu3Ty/ilixZwu23367/HGmaRnx8PHv27NEzy5Yto127dm6Z7du3Ex8fr2f++9//0qFDB7c2xcTEEBcXp2eWL19Op06d3DKbNm1i69atOJ1OsgyZ3HXXXXTr1k1/HWD9+vVs3LhRz/Ts2ZMePXq4ZaKjo1m/fj0Oh4NsYxa9e/fm7rvvdiuaNE1j9erVrFu7Ts/de++93HPPPW77tmLFCtatOZvp27cvffv21TMAm5ZvIm/DcVQN0Iphc7fNzNRmuv0/2rx8M3nr3TOfGj91W8/WFVvJW3cc5QtaCcR0j2GOzxy3fYtdFUte9AlUDYVWrLGt5za+CvzKbT1xf8SR98cJlJ9CK9LY3ms784Lnnfd/f/ua7eStvnju3MyO3juYX2u+2/Z2RO9wZU63aWfvnSyos8Ats3PtTvJWnc3E3xPPwtCFbpld63eRt+Lstnb32c3i+ovdMrs37Cbv97OZPX338P3N37v92+7dtJe8306An4Iijb399rKkgXshAJCyNZnCnHxMDc3Y02wcSklh2U/L3DKHtqa4ZdJT0lj36zq3n5HMtccoPFqAqZErk3Uwg4NdktzanRWdQeGRs5njSTnkdMl2y5z6PY/C5LOZwqQCgiPcR9jKfiqmsGkBltssKDsUHM4/b78uh9cKhgMHDmA0GmnWrJm+rG3btqxdu/a87N69e2nbtq1bbu/evd5q2g3BE0XFpeQuNOpRXq9evYiJidELj/79+3Pfffe5ZQ4ePEhiYqKeeeyxx3j3Xfff1gICAtwKmOHDh1+0gLFYLIwaNarCzN/+9rc/7bNzc+PGjSMqKgqn06kXIAMHDtRHawC6du3KokWLUErpuSFDhvDHH3/omXbt2vH111/rGaUUw4YNY926dXqmRYsWfPbZZ26ZM4efwFV4NWjQgA8//NA1LHo6N3r0aNLS0vRM3bp1GT16tL4OpRTjx4/X5wAppQgMDOTZZ5/VM2f+nDp1qlsRZzAY6Nu3r74eQC8EzmQcDgddu3Z1y5z5kj+TKSkpoU2bNm6ZNWvWuGXy8/MJDw/XM0opcnNz3TI5OTkEBga67UtmZqZbJiMjQ5/zdOZx5MgRt8IzPT2d4uKzEy7PrC8tLc0td+jQIXJzc93aXX4UzuFwkJSUROvWrd3adODAAZxFTigCBSQmJtK8eXO3n7XExES3zP79+2natKnbevbu3Yuz0AmFrsyePXv0PjqT2bVrF85Ch57ZuXMnoaGhbnO+tm/f7paJi4ujVq1a5/74uwq4CnLnZrZt20ZQUJDb9rZt24bT6gCrKxMTE0NAQIBbJiYmxi2zZcsWfH193fZt8+bNbplNmzZhNBrd1rNx40a3zIYNG9xeV0q5ZQDWrVunF/3l5aTlULTDCmYNbIqc2jl88803F81k18pm3rx5bj8jmSmZFMUVgEWDMkVmrUz++c9/urUrMzmTom0F+noyamUwa9Yst8yxpGMUxZ7NHAs+xowZM9zacyZTXEMDo4ZvJ9/z9uuyKC9Zt26duummm9yWzZ07V3Xv3v28rMFgUPv27dOfHzhwQAHK6XSel125cqUaM2aMGjNmjPJi84UXOJ1OFRUVpfr376+ioqL+9N/3astUNhcVFaUsFosClMViUVFRUddF5mpsk+y/7L/sv2f27VJ57Rt3+/btqkaNGm7LPvzwQ3X//fefl61Zs6baunWr/nzbtm0qICCgwm1IwSCuFldboVPVBdP1mrka2yT7L/vvqX27VF47rbKwsJBatWqxd+9emjZtCsDTTz9NWFjYeXMYOnfuzHPPPcfwgeOmEQAAB6xJREFU4cMB+Prrr5k7d67bzZz+zPV4WqUQQghxVfJI2XEBgwcPVo899piyWq1qw4YNqmbNmmrPnj3n5ebMmaNatGihjhw5oo4ePapatmyp5syZU+H6cR2ekoc85CEPechDHpf5qCyvnlY5e/Zsnn/+eUJDQ6lTpw5z5syhVatWrF+/nn79+ukzxkeMGEFKSgpt2rQBYNiwYYwYMaLC9UdGRrJt2zZv7oIop3379tLfVUT6uupIX1ct6e+q4+m+9mrBULt2bZYuXXre8m7duunFArgOLUyfPp3p06d7szlCCCGEuEyG6m6AEEIIIa5+xgkVnax/lYuMjKw4JDxG+rvqSF9XHenrqiX9XXU82dfX9M2nhBBCCFE15JCEEEIIISokBYMQQgghKnRNFgy5ubk8+OCD+Pv707BhQxYuXFjdTbpuzJo1i/bt2+Pj48Ozzz7r9trq1atp0aIFfn5+9OzZU79Xgbg8paWlDB06lIYNGxIYGEi7du347bff9Nelvz1vyJAh1K9fn5o1a9KsWTO+/PJL/TXpb+9ISkrC19eXIUPO3i1z4cKFNGzYEH9/fwYNGqTfcE1cvh49euDr60tAQAABAQFu9ynxVH9fkwVD+dtmf/vtt7z00ktysyoPCQsL49133+X55593W378+HEeeughJk+eTG5uLu3bt2fw4MHV1Mrrg91uJzw8nLVr13Lq1CkmT57Mo48+SmpqqvS3l4wbN47U1FTy8/P5+eefeffdd4mLi5P+9qKRI0fSoUMH/fnevXsZMWIECxYsICsrCz8/P15++eVqbOH1Y9asWVitVqxWK4mJiYCH+/sSLtx4VbBarcpsNqvExER92ZAhQ9SYMWOqsVXXn3feeUc988wz+vPPP/9cderUSX9utVqVr6+v203DxJVr06aN+uGHH6S/q8D+/ftVvXr11OLFi6W/vWTRokXqkUceUVFRUerJJ59USik1btw49fjjj+uZgwcPKrPZrPLz86urmdeF7t27qy+++OK85Z7s72tuhOFCt82WEQbvOvcW5P7+/tx6663S7x6UlZXFgQMHaNWqlfS3F7388sv4+fnRokUL6tevT//+/aW/vSA/P5/x48fz0UcfuS0/t69vvfVWLBYLBw4cqOomXnfGjRtHSEgIXbp0ITo6GvBsf19zBYPVaiUoKMhtWVBQEAUFBdXUohuD9Lt32Ww2nnzySZ555hlatGgh/e1Fs2fPpqCggPXr1/PQQw/h4+Mj/e0F7733HkOHDiU8PNxtufS1d0ybNo2UlBSOHj3KCy+8wAMPPEBycrJH+/uaKxgCAgLIz893W5afn09gYGA1tejGIP3uPU6nk6eeegqLxcKsWbMA6W9vMxqNdO3alSNHjjBnzhzpbw/buXMnq1at4o033jjvNelr7+jYsSOBgYH4+PjwzDPP0KVLF3799VeP9rdX7yXhDc2aNcNut5OUlKTfNjs+Pp5WrVpVc8uub61atWL+/Pn688LCQpKTk6Xfr5BSiqFDh5KVlcWvv/6K2WwGpL+rit1u1/tV+ttzoqOjSU1NpUGDBoBrVMHhcJCQkEDfvn2Jj4/XsykpKZSWlrodZhZXTtM0lFK0atXKc/19BXMsqk1lb5stLp3NZlPFxcVq7NixasiQIaq4uFjZbDaVnZ2tatasqX744QdVXFysRo8erTp27Fjdzb3mjRgxQnXs2FEVFBS4LZf+9rysrCy1aNEiVVBQoOx2u1q+fLny8/NTS5culf72sMLCQpWRkaE/3nrrLfXwww+r7OxstWfPHhUYGKjWrVunrFarevLJJ9XgwYOru8nXtLy8PLV8+XL98/qbb75Rfn5+av/+/R7t72uyYDhx4oQaOHCg8vPzU+Hh4erbb7+t7iZdN6Kios67V3pUVJRSSqmVK1eq5s2bK19fX9W9e3d16NCham3rtS41NVUBysfHR/n7++uPb775Rikl/e1p2dnZ6q677lJBQUEqMDBQtW7dWs2dO1d/Xfrbe8qfJaGUUt9++60KDw9Xfn5+asCAAerEiRPV2LprX3Z2tmrfvr0KCAhQQUFBqmPHjmrFihX6657qb7mXhBBCCCEqdM1NehRCCCFE1ZOCQQghhBAVkoJBCCGEEBWSgkEIIYQQFZKCQQhxxRo1asSqVavcls2bN4+uXbtWU4uEEJ4mBYMQQgghKiQFgxDC66ZNm8bNN99MYGAgzZs3Z/Xq1dXdJCHEJbrmLg0thLi2JCYmMmvWLGJjYwkLCyM1NRWHw1HdzRJCXCIpGIQQHjFo0CBMprMfKWVlZURERGA0GiktLSUhIYG6devSqFGj6mukEOKyySEJIYRHLF26lJMnT+qP2bNnA9CkSRNmzJjBhAkTCA0N5bHHHuPYsWPV3FohxKWSgkEI4XVPPPEEGzZsIC0tDU3TGDNmTHU3SQhxiaRgEEJ4VWJiIn/88QelpaX4+vpSo0YNjEZjdTdLCHGJpGAQQnhVaWkpY8eOJSQkhHr16pGdnc2UKVOqu1lCiEskd6sUQgghRIVkhEEIIYQQFZKCQQghhBAVkoJBCCGEEBWSgkEIIYQQFZKCQQghhBAVkoJBCCGEEBWSgkEIIYQQFZKCQQghhBAVkoJBCCGEEBX6/+8e5naKgFQCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 540x396 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots()\n",
    "\n",
    "ax.plot(Hs_thresholds, INDELs_other_prop_list, color = 'xkcd:black', linewidth = 1.0)\n",
    "ax.scatter(Hs_thresholds, INDELs_other_prop_list, color = 'xkcd:black', linewidth = 0.5, s = 15, edgecolor = 'k')\n",
    "\n",
    "ax.plot(Hs_thresholds, INDELs_SSR_prop_list, color = 'xkcd:green', linewidth = 1.0)\n",
    "ax.scatter(Hs_thresholds, INDELs_SSR_prop_list, color = 'xkcd:green', linewidth = 0.5, s = 15, edgecolor = 'k')\n",
    "\n",
    "ax.plot(Hs_thresholds, INDELs_HT_prop_list, color = 'xkcd:blue', linewidth = 1.0)\n",
    "ax.scatter(Hs_thresholds, INDELs_HT_prop_list, color = 'xkcd:blue', linewidth = 0.5, s = 15, edgecolor = 'k')\n",
    "\n",
    "# ax.set_title('Distribution of Homoplasy\\nScores for 330 INDELs', fontsize = 10, color = 'k', pad = -10)\n",
    "ax.set_xlabel('Hs', fontsize = 12, color = 'k', labelpad = 1)\n",
    "ax.set_ylabel('Proportion of INDELs >= Hs' , fontsize = 12, color = 'k', labelpad = 1)\n",
    "\n",
    "ax.spines['right'].set_visible(False)\n",
    "ax.spines['top'].set_visible(False)\n",
    "ax.grid(False)\n",
    "\n",
    "ax.tick_params(labelcolor = 'k')\n",
    "ax.set_xlim(0, 51)\n",
    "ax.set_ylim(-0.01, 1.015)\n",
    "\n",
    "ax.tick_params(axis='y', which='major', labelsize=12 , labelcolor = 'k')\n",
    "ax.tick_params(axis='x', which='major', labelsize=12 , labelcolor = 'k')\n",
    "\n",
    "#create legend\n",
    "HT = 'xkcd:blue'\n",
    "SSR = 'xkcd:green'\n",
    "other = 'xkcd:black'\n",
    "\n",
    "handles = [Rectangle((0,0),1,1,color=c,ec=\"k\") for c in [HT,SSR,other]]\n",
    "labels = [\"HT\",\"SSR\",\"other\"]\n",
    "ax.legend(handles, labels, fontsize=12)\n",
    "\n",
    "\n",
    "fig = plt.gcf()\n",
    "fig.set_size_inches(7.5, 5.5)\n",
    "fig.tight_layout()\n",
    "\n",
    "file_name = '/n/data1/hms/dbmi/farhat/Roger/homoplasy_project/Figures/for publication/HT_SSR_other_proportion_INDELs_Hs.pdf'\n",
    "## plt.savefig(file_name, bbox_inches='tight', dpi = 300 , transparent = True) # incorporated into manhattan panel\n",
    "\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'43899.0'"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'other'])[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'2077.0'"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes'])[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'330.0'"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "str(float(np.shape(all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'HT'])[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## [7] check INDELs in SSR regions with highest homoplasy score"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "####################################################################################################################################################################################"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_SSR_homoplasic_INDELs = all_homoplasic_INDELs[all_homoplasic_INDELs.SSR_region == 'yes']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>gene_category</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_pos</th>\n",
       "      <th>ins_del</th>\n",
       "      <th>INDEL_type</th>\n",
       "      <th>codon_pos</th>\n",
       "      <th>...</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4A</th>\n",
       "      <th>4B</th>\n",
       "      <th>4C</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>total</th>\n",
       "      <th>num_isolates</th>\n",
       "      <th>SSR_region</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>752</th>\n",
       "      <td>55540</td>\n",
       "      <td>G</td>\n",
       "      <td>GCCGCCGCCT</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>ponA1</td>\n",
       "      <td>Rv0050</td>\n",
       "      <td>1878</td>\n",
       "      <td>ins</td>\n",
       "      <td>inframe</td>\n",
       "      <td>626</td>\n",
       "      <td>...</td>\n",
       "      <td>272.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>397.0</td>\n",
       "      <td>2238</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34894</th>\n",
       "      <td>3247864</td>\n",
       "      <td>C</td>\n",
       "      <td>CTAGG</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>ppsA</td>\n",
       "      <td>Rv2931</td>\n",
       "      <td>2420</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>807</td>\n",
       "      <td>...</td>\n",
       "      <td>90.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>66.0</td>\n",
       "      <td>79.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>324.0</td>\n",
       "      <td>469</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17487</th>\n",
       "      <td>1625332</td>\n",
       "      <td>C</td>\n",
       "      <td>CGGT</td>\n",
       "      <td>Essential</td>\n",
       "      <td>opcA</td>\n",
       "      <td>Rv1446c</td>\n",
       "      <td>34</td>\n",
       "      <td>ins</td>\n",
       "      <td>inframe</td>\n",
       "      <td>12</td>\n",
       "      <td>...</td>\n",
       "      <td>184.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>186.0</td>\n",
       "      <td>5157</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>756</th>\n",
       "      <td>55543</td>\n",
       "      <td>G</td>\n",
       "      <td>GCCGCCT</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>ponA1</td>\n",
       "      <td>Rv0050</td>\n",
       "      <td>1881</td>\n",
       "      <td>ins</td>\n",
       "      <td>inframe</td>\n",
       "      <td>627</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>149.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>153.0</td>\n",
       "      <td>1322</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41519</th>\n",
       "      <td>3969423</td>\n",
       "      <td>GGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PPE61</td>\n",
       "      <td>Rv3532</td>\n",
       "      <td>81</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>21.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>151.0</td>\n",
       "      <td>1289</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>40873</th>\n",
       "      <td>3895244</td>\n",
       "      <td>G</td>\n",
       "      <td>GGCC</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PPE60</td>\n",
       "      <td>Rv3478</td>\n",
       "      <td>819</td>\n",
       "      <td>ins</td>\n",
       "      <td>inframe</td>\n",
       "      <td>273</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>133.0</td>\n",
       "      <td>963</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37345</th>\n",
       "      <td>3528971</td>\n",
       "      <td>TCGC</td>\n",
       "      <td>T</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PPE53</td>\n",
       "      <td>Rv3159c</td>\n",
       "      <td>193</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>65</td>\n",
       "      <td>...</td>\n",
       "      <td>40.0</td>\n",
       "      <td>32.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>435</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30979</th>\n",
       "      <td>2873093</td>\n",
       "      <td>G</td>\n",
       "      <td>GCCA</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Rv2553c</td>\n",
       "      <td>Rv2553c</td>\n",
       "      <td>173</td>\n",
       "      <td>ins</td>\n",
       "      <td>inframe</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>120.0</td>\n",
       "      <td>732</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>47313</th>\n",
       "      <td>4407851</td>\n",
       "      <td>GC</td>\n",
       "      <td>G</td>\n",
       "      <td>Antibiotic Resistance</td>\n",
       "      <td>gid</td>\n",
       "      <td>Rv3919c</td>\n",
       "      <td>352</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>118</td>\n",
       "      <td>...</td>\n",
       "      <td>9.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>115.0</td>\n",
       "      <td>182</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10908</th>\n",
       "      <td>968965</td>\n",
       "      <td>GCGC</td>\n",
       "      <td>G</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PE_PGRS15</td>\n",
       "      <td>Rv0872c</td>\n",
       "      <td>1280</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>427</td>\n",
       "      <td>...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>440</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>748</th>\n",
       "      <td>55533</td>\n",
       "      <td>TGCC</td>\n",
       "      <td>T</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>ponA1</td>\n",
       "      <td>Rv0050</td>\n",
       "      <td>1871</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>624</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>388</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6427</th>\n",
       "      <td>552495</td>\n",
       "      <td>G</td>\n",
       "      <td>GCTA</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv0461</td>\n",
       "      <td>Rv0461</td>\n",
       "      <td>470</td>\n",
       "      <td>ins</td>\n",
       "      <td>inframe</td>\n",
       "      <td>157</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>97.0</td>\n",
       "      <td>1641</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44680</th>\n",
       "      <td>4279825</td>\n",
       "      <td>AGCC</td>\n",
       "      <td>A</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv3814c</td>\n",
       "      <td>Rv3814c</td>\n",
       "      <td>191</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>64</td>\n",
       "      <td>...</td>\n",
       "      <td>16.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>262</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9336</th>\n",
       "      <td>830868</td>\n",
       "      <td>G</td>\n",
       "      <td>GGC</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv0739</td>\n",
       "      <td>Rv0739</td>\n",
       "      <td>14</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>50.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>62.0</td>\n",
       "      <td>12668</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41588</th>\n",
       "      <td>3972244</td>\n",
       "      <td>TGCC</td>\n",
       "      <td>T</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PPE62</td>\n",
       "      <td>Rv3533c</td>\n",
       "      <td>210</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>70</td>\n",
       "      <td>...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>132</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8002</th>\n",
       "      <td>699978</td>\n",
       "      <td>CCG</td>\n",
       "      <td>C</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv0603</td>\n",
       "      <td>Rv0603</td>\n",
       "      <td>123</td>\n",
       "      <td>del</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>41</td>\n",
       "      <td>...</td>\n",
       "      <td>12.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>239</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>44224</th>\n",
       "      <td>4227026</td>\n",
       "      <td>GCGA</td>\n",
       "      <td>G</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv3780</td>\n",
       "      <td>Rv3780</td>\n",
       "      <td>38</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>13</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>52.0</td>\n",
       "      <td>382</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41941</th>\n",
       "      <td>4029034</td>\n",
       "      <td>GGGC</td>\n",
       "      <td>G</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv3587c</td>\n",
       "      <td>Rv3587c</td>\n",
       "      <td>729</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>243</td>\n",
       "      <td>...</td>\n",
       "      <td>13.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>241</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33672</th>\n",
       "      <td>3107859</td>\n",
       "      <td>C</td>\n",
       "      <td>CGGT</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>Rv2799</td>\n",
       "      <td>Rv2799</td>\n",
       "      <td>92</td>\n",
       "      <td>ins</td>\n",
       "      <td>inframe</td>\n",
       "      <td>31</td>\n",
       "      <td>...</td>\n",
       "      <td>15.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>48.0</td>\n",
       "      <td>119</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41740</th>\n",
       "      <td>3998059</td>\n",
       "      <td>AGGC</td>\n",
       "      <td>A</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PPE64</td>\n",
       "      <td>Rv3558</td>\n",
       "      <td>80</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>27</td>\n",
       "      <td>...</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>5139</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30981</th>\n",
       "      <td>2873093</td>\n",
       "      <td>GCCA</td>\n",
       "      <td>G</td>\n",
       "      <td>Essential</td>\n",
       "      <td>Rv2553c</td>\n",
       "      <td>Rv2553c</td>\n",
       "      <td>173</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>58</td>\n",
       "      <td>...</td>\n",
       "      <td>19.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>44.0</td>\n",
       "      <td>403</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14792</th>\n",
       "      <td>1340167</td>\n",
       "      <td>GGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>PE/PPE</td>\n",
       "      <td>PPE18</td>\n",
       "      <td>Rv1196</td>\n",
       "      <td>819</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>273</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>381</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>45928</th>\n",
       "      <td>4348820</td>\n",
       "      <td>G</td>\n",
       "      <td>GGC</td>\n",
       "      <td>None</td>\n",
       "      <td>eccCa1_eccCb1</td>\n",
       "      <td>Rv3870_Rv3871</td>\n",
       "      <td>None</td>\n",
       "      <td>ins</td>\n",
       "      <td>frameshift</td>\n",
       "      <td>None</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>199</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34403</th>\n",
       "      <td>3186588</td>\n",
       "      <td>ACGG</td>\n",
       "      <td>A</td>\n",
       "      <td>Antigen</td>\n",
       "      <td>dipZ</td>\n",
       "      <td>Rv2874</td>\n",
       "      <td>1742</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>581</td>\n",
       "      <td>...</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>37.0</td>\n",
       "      <td>1147</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>43733</th>\n",
       "      <td>4194000</td>\n",
       "      <td>CACG</td>\n",
       "      <td>C</td>\n",
       "      <td>Non-Essential</td>\n",
       "      <td>ctpJ</td>\n",
       "      <td>Rv3743c</td>\n",
       "      <td>1374</td>\n",
       "      <td>del</td>\n",
       "      <td>inframe</td>\n",
       "      <td>458</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>72</td>\n",
       "      <td>yes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25 rows × 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           pos   ref         alt          gene_category      gene_name  \\\n",
       "752      55540     G  GCCGCCGCCT          Non-Essential          ponA1   \n",
       "34894  3247864     C       CTAGG          Non-Essential           ppsA   \n",
       "17487  1625332     C        CGGT              Essential           opcA   \n",
       "756      55543     G     GCCGCCT          Non-Essential          ponA1   \n",
       "41519  3969423  GGCC           G                 PE/PPE          PPE61   \n",
       "40873  3895244     G        GGCC                 PE/PPE          PPE60   \n",
       "37345  3528971  TCGC           T                 PE/PPE          PPE53   \n",
       "30979  2873093     G        GCCA              Essential        Rv2553c   \n",
       "47313  4407851    GC           G  Antibiotic Resistance            gid   \n",
       "10908   968965  GCGC           G                 PE/PPE      PE_PGRS15   \n",
       "748      55533  TGCC           T          Non-Essential          ponA1   \n",
       "6427    552495     G        GCTA          Non-Essential         Rv0461   \n",
       "44680  4279825  AGCC           A          Non-Essential        Rv3814c   \n",
       "9336    830868     G         GGC          Non-Essential         Rv0739   \n",
       "41588  3972244  TGCC           T                 PE/PPE          PPE62   \n",
       "8002    699978   CCG           C          Non-Essential         Rv0603   \n",
       "44224  4227026  GCGA           G          Non-Essential         Rv3780   \n",
       "41941  4029034  GGGC           G          Non-Essential        Rv3587c   \n",
       "33672  3107859     C        CGGT          Non-Essential         Rv2799   \n",
       "41740  3998059  AGGC           A                 PE/PPE          PPE64   \n",
       "30981  2873093  GCCA           G              Essential        Rv2553c   \n",
       "14792  1340167  GGCC           G                 PE/PPE          PPE18   \n",
       "45928  4348820     G         GGC                   None  eccCa1_eccCb1   \n",
       "34403  3186588  ACGG           A                Antigen           dipZ   \n",
       "43733  4194000  CACG           C          Non-Essential           ctpJ   \n",
       "\n",
       "             gene_id gene_pos ins_del  INDEL_type codon_pos     ...      \\\n",
       "752           Rv0050     1878     ins     inframe       626     ...       \n",
       "34894         Rv2931     2420     ins  frameshift       807     ...       \n",
       "17487        Rv1446c       34     ins     inframe        12     ...       \n",
       "756           Rv0050     1881     ins     inframe       627     ...       \n",
       "41519         Rv3532       81     del     inframe        27     ...       \n",
       "40873         Rv3478      819     ins     inframe       273     ...       \n",
       "37345        Rv3159c      193     del     inframe        65     ...       \n",
       "30979        Rv2553c      173     ins     inframe        58     ...       \n",
       "47313        Rv3919c      352     del  frameshift       118     ...       \n",
       "10908        Rv0872c     1280     del     inframe       427     ...       \n",
       "748           Rv0050     1871     del     inframe       624     ...       \n",
       "6427          Rv0461      470     ins     inframe       157     ...       \n",
       "44680        Rv3814c      191     del     inframe        64     ...       \n",
       "9336          Rv0739       14     ins  frameshift         5     ...       \n",
       "41588        Rv3533c      210     del     inframe        70     ...       \n",
       "8002          Rv0603      123     del  frameshift        41     ...       \n",
       "44224         Rv3780       38     del     inframe        13     ...       \n",
       "41941        Rv3587c      729     del     inframe       243     ...       \n",
       "33672         Rv2799       92     ins     inframe        31     ...       \n",
       "41740         Rv3558       80     del     inframe        27     ...       \n",
       "30981        Rv2553c      173     del     inframe        58     ...       \n",
       "14792         Rv1196      819     del     inframe       273     ...       \n",
       "45928  Rv3870_Rv3871     None     ins  frameshift      None     ...       \n",
       "34403         Rv2874     1742     del     inframe       581     ...       \n",
       "43733        Rv3743c     1374     del     inframe       458     ...       \n",
       "\n",
       "           2      3    4A    4B    4C    5    6  total  num_isolates  \\\n",
       "752    272.0    0.0   0.0   0.0   0.0  6.0  5.0  397.0          2238   \n",
       "34894   90.0   12.0  66.0  79.0  38.0  1.0  3.0  324.0           469   \n",
       "17487  184.0    2.0   0.0   0.0   0.0  0.0  0.0  186.0          5157   \n",
       "756      2.0  149.0   0.0   0.0   0.0  0.0  0.0  153.0          1322   \n",
       "41519   25.0   21.0  22.0  17.0  37.0  2.0  0.0  151.0          1289   \n",
       "40873    1.0    0.0   0.0  16.0   3.0  0.0  0.0  133.0           963   \n",
       "37345   40.0   32.0  22.0   2.0   3.0  2.0  0.0  128.0           435   \n",
       "30979   20.0   12.0  31.0  16.0  10.0  0.0  0.0  120.0           732   \n",
       "47313    9.0   40.0  20.0   6.0  19.0  0.0  0.0  115.0           182   \n",
       "10908   25.0   12.0  12.0  14.0  19.0  1.0  0.0  100.0           440   \n",
       "748      0.0    0.0  43.0  18.0  36.0  0.0  0.0   97.0           388   \n",
       "6427     0.0    0.0  97.0   0.0   0.0  0.0  0.0   97.0          1641   \n",
       "44680   16.0   16.0   9.0  14.0   5.0  1.0  1.0   70.0           262   \n",
       "9336    50.0    3.0   1.0   0.0   1.0  1.0  1.0   62.0         12668   \n",
       "41588   15.0    4.0  11.0   9.0   5.0  0.0  0.0   53.0           132   \n",
       "8002    12.0    8.0   9.0   8.0   4.0  0.0  1.0   53.0           239   \n",
       "44224    6.0    2.0  13.0  12.0  10.0  0.0  0.0   52.0           382   \n",
       "41941   13.0    4.0  15.0   2.0   5.0  0.0  0.0   49.0           241   \n",
       "33672   15.0    3.0   3.0   5.0  17.0  0.0  0.0   48.0           119   \n",
       "41740    5.0    1.0   3.0  18.0   7.0  0.0  0.0   46.0          5139   \n",
       "30981   19.0    3.0   9.0   3.0   5.0  0.0  0.0   44.0           403   \n",
       "14792    7.0   10.0   6.0  10.0   7.0  1.0  1.0   43.0           381   \n",
       "45928    0.0   38.0   0.0   0.0   0.0  0.0  0.0   38.0           199   \n",
       "34403    7.0    2.0   9.0   7.0   3.0  0.0  0.0   37.0          1147   \n",
       "43733    4.0    6.0   5.0   9.0   5.0  0.0  0.0   34.0            72   \n",
       "\n",
       "       SSR_region  \n",
       "752           yes  \n",
       "34894         yes  \n",
       "17487         yes  \n",
       "756           yes  \n",
       "41519         yes  \n",
       "40873         yes  \n",
       "37345         yes  \n",
       "30979         yes  \n",
       "47313         yes  \n",
       "10908         yes  \n",
       "748           yes  \n",
       "6427          yes  \n",
       "44680         yes  \n",
       "9336          yes  \n",
       "41588         yes  \n",
       "8002          yes  \n",
       "44224         yes  \n",
       "41941         yes  \n",
       "33672         yes  \n",
       "41740         yes  \n",
       "30981         yes  \n",
       "14792         yes  \n",
       "45928         yes  \n",
       "34403         yes  \n",
       "43733         yes  \n",
       "\n",
       "[25 rows x 23 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_SSR_homoplasic_INDELs.sort_values(by='total', ascending = False).head(n = 25)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "pos                    4348820\n",
       "ref                          G\n",
       "alt                        GGC\n",
       "gene_category             None\n",
       "gene_name        eccCa1_eccCb1\n",
       "gene_id          Rv3870_Rv3871\n",
       "gene_pos                  None\n",
       "ins_del                    ins\n",
       "INDEL_type          frameshift\n",
       "codon_pos                 None\n",
       "stop_codon          intergenic\n",
       "key              G_4348820_GGC\n",
       "1                            0\n",
       "2                            0\n",
       "3                           38\n",
       "4A                           0\n",
       "4B                           0\n",
       "4C                           0\n",
       "5                            0\n",
       "6                            0\n",
       "total                       38\n",
       "num_isolates               199\n",
       "SSR_region                 yes\n",
       "Name: 45928, dtype: object"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_SSR_homoplasic_INDELs.loc[45928,:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*ponA1*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>236</th>\n",
       "      <td>55533</td>\n",
       "      <td>55554</td>\n",
       "      <td>GCC</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>237</th>\n",
       "      <td>55533</td>\n",
       "      <td>55551</td>\n",
       "      <td>GCCGCC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>238</th>\n",
       "      <td>55534</td>\n",
       "      <td>55555</td>\n",
       "      <td>CCG</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>239</th>\n",
       "      <td>55534</td>\n",
       "      <td>55552</td>\n",
       "      <td>CCGCCG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>240</th>\n",
       "      <td>55535</td>\n",
       "      <td>55553</td>\n",
       "      <td>CGC</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>241</th>\n",
       "      <td>55535</td>\n",
       "      <td>55553</td>\n",
       "      <td>CGCCGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     H37Rv_start  H37Rv_end   motif  repeats\n",
       "236        55533      55554     GCC        7\n",
       "237        55533      55551  GCCGCC        3\n",
       "238        55534      55555     CCG        7\n",
       "239        55534      55552  CCGCCG        3\n",
       "240        55535      55553     CGC        6\n",
       "241        55535      55553  CGCCGC        3"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df[(SSR_H37Rv_df.H37Rv_start <= 55540) & (SSR_H37Rv_df.H37Rv_end >= 55533)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*gid*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18303</th>\n",
       "      <td>4407846</td>\n",
       "      <td>4407852</td>\n",
       "      <td>GC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       H37Rv_start  H37Rv_end motif  repeats\n",
       "18303      4407846    4407852    GC        3"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df[(SSR_H37Rv_df.H37Rv_start <= 4407851) & (SSR_H37Rv_df.H37Rv_end >= 4407851)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*PPE60*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16191</th>\n",
       "      <td>3895236</td>\n",
       "      <td>3895245</td>\n",
       "      <td>CGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16192</th>\n",
       "      <td>3895237</td>\n",
       "      <td>3895246</td>\n",
       "      <td>GGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       H37Rv_start  H37Rv_end motif  repeats\n",
       "16191      3895236    3895245   CGG        3\n",
       "16192      3895237    3895246   GGC        3"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df[(SSR_H37Rv_df.H37Rv_start <= 3895244) & (SSR_H37Rv_df.H37Rv_end >= 3895244)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*PPE18*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5655</th>\n",
       "      <td>1340159</td>\n",
       "      <td>1340168</td>\n",
       "      <td>CGG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5656</th>\n",
       "      <td>1340160</td>\n",
       "      <td>1340169</td>\n",
       "      <td>GGC</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      H37Rv_start  H37Rv_end motif  repeats\n",
       "5655      1340159    1340168   CGG        3\n",
       "5656      1340160    1340169   GGC        3"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df[(SSR_H37Rv_df.H37Rv_start <= 1340167) & (SSR_H37Rv_df.H37Rv_end >= 1340167)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "*ppsA*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style>\n",
       "    .dataframe thead tr:only-child th {\n",
       "        text-align: right;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>H37Rv_start</th>\n",
       "      <th>H37Rv_end</th>\n",
       "      <th>motif</th>\n",
       "      <th>repeats</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>13542</th>\n",
       "      <td>3247858</td>\n",
       "      <td>3247866</td>\n",
       "      <td>GC</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13543</th>\n",
       "      <td>3247859</td>\n",
       "      <td>3247865</td>\n",
       "      <td>CG</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       H37Rv_start  H37Rv_end motif  repeats\n",
       "13542      3247858    3247866    GC        4\n",
       "13543      3247859    3247865    CG        3"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "SSR_H37Rv_df[(SSR_H37Rv_df.H37Rv_start <= 3247864) & (SSR_H37Rv_df.H37Rv_end >= 3247864)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
